{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Keyring is skipped due to an exception: not all arguments converted during string formatting\n"
     ]
    }
   ],
   "source": [
    "from riptide import *\n",
    "from cobra.flux_analysis.variability import find_essential_genes, find_essential_reactions\n",
    "from cobra.medium.minimal_medium import minimal_medium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "iCdR703 = cobra.io.load_json_model('/home/mjenior/Desktop/Jenior_Consortia_2022/data/iCdR703.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCdG703</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f7e0561cd50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1243</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1313</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*biomass - 1.0*biomass_reverse_01e59</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCdG703 at 0x7f7e0561cd50>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iCdR703"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/mjenior/Desktop/R20291_nuc_genes.fasta', 'w') as outFasta:\n",
    "    for x in iCdR703.genes:\n",
    "        entry = '>' + x.id + '|' + x.name.replace(' ','_') + '\\n'\n",
    "        outFasta.write(entry)\n",
    "        outFasta.write(x.annotation['ccds'] + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse RefSeq gene ID lookup table\n",
    "refseq_dict = {}\n",
    "with open('/home/mjenior/Desktop/active_projects/Tamayo/c-di-AMP/CdR20291_genes.tsv', 'r') as gene_catalog:\n",
    "    header = gene_catalog.readline()\n",
    "    for line in gene_catalog:\n",
    "        line = line.split()\n",
    "        refseq_dict[line[1]] = line[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes missing in CD196: 0\n"
     ]
    }
   ],
   "source": [
    "cd196_missing = ['CDR20291_0183','CDR20291_1419','CDR20291_1744','CDR20291_1745','CDR20291_1746',\n",
    "                 'CDR20291_1747','CDR20291_1748','CDR20291_1749','CDR20291_1750','CDR20291_1751',\n",
    "                 'CDR20291_1752','CDR20291_1753','CDR20291_1754','CDR20291_1755','CDR20291_1756',\n",
    "                 'CDR20291_1757','CDR20291_1758','CDR20291_1759','CDR20291_1760','CDR20291_1761',\n",
    "                 'CDR20291_1762','CDR20291_1763','CDR20291_1764','CDR20291_1765','CDR20291_1766',\n",
    "                 'CDR20291_1767','CDR20291_1768','CDR20291_1769','CDR20291_1770','CDR20291_1771',\n",
    "                 'CDR20291_1772','CDR20291_1773','CDR20291_1774','CDR20291_1775','CDR20291_1776',\n",
    "                 'CDR20291_1777','CDR20291_1778','CDR20291_1779','CDR20291_1780','CDR20291_1781',\n",
    "                 'CDR20291_1782','CDR20291_1783','CDR20291_1784','CDR20291_1785','CDR20291_1786',\n",
    "                 'CDR20291_1787','CDR20291_1788','CDR20291_1809']\n",
    "\n",
    "check_missing = []\n",
    "for x in cd196_missing:\n",
    "    try:\n",
    "        check_missing.append(refseq_dict[x])\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "check_missing = set(check_missing)\n",
    "included_genes = []\n",
    "for x in iCdR703.genes:\n",
    "    if x.id in check_missing:\n",
    "        included_genes.append(x.id)\n",
    "\n",
    "print('Genes missing in CD196:', len(included_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in transcription\n",
    "solo = {}\n",
    "with open('/home/mjenior/Desktop/repos/Smith_etal_Enterococcus/data/solo_reps_norm.tsv', 'r') as transcription:\n",
    "    for line in transcription:\n",
    "        line = line.split()\n",
    "        abunds = [float(x) for x in line[1:]]\n",
    "        try:\n",
    "            patric = refseq_dict[line[0]]\n",
    "            solo[patric] = abunds\n",
    "        except:\n",
    "            continue\n",
    "        \n",
    "coculture = {}\n",
    "with open('/home/mjenior/Desktop/repos/Smith_etal_Enterococcus/data/coculture_reps_norm.tsv', 'r') as transcription:\n",
    "    for line in transcription:\n",
    "        line = line.split()\n",
    "        abund = [float(x) for x in line[1:]]\n",
    "        try:\n",
    "            patric = refseq_dict[line[0]]\n",
    "            coculture[patric] = abunds\n",
    "        except:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2879.0"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# proline racemase\n",
    "numpy.median(solo['645463.3.peg.3289'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "67.0"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpy.median(coculture['645463.3.peg.3289'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "medium = ['cpd00001_e', # water\n",
    "'cpd00082_e', # D-fructose\n",
    "'cpd00035_e', # alanine\n",
    "'cpd00041_e', # aspartic acid\n",
    "'cpd00023_e', # glutamic acid\n",
    "'cpd00119_e', # histidine\n",
    "'cpd00107_e', # leucine\n",
    "'cpd00060_e', # methionine\n",
    "'cpd00161_e', # threonine\n",
    "'cpd00069_e', # tyrosine\n",
    "'cpd00084_e', # cysteine\n",
    "'cpd00033_e', # glycine\n",
    "'cpd00322_e', # isoleucine\n",
    "'cpd00066_e', # phenylalanine\n",
    "'cpd00054_e', # serine\n",
    "'cpd00065_e', # tryptophan\n",
    "'cpd00156_e', # valine\n",
    "'cpd00220_e', # riboflavin\n",
    "'cpd00644_e', # pantothentate(B5)\n",
    "'cpd00393_e', # folate(B9)\n",
    "'cpd00133_e', # niacin(PP)\n",
    "'cpd00263_e', # Pyridoxine\n",
    "'cpd00104_e', # Biotin\n",
    "'cpd00149_e', # Cobalt\n",
    "'cpd00971_e', # sodium\n",
    "'cpd00099_e', # Chloride\n",
    "'cpd00205_e', # potassium\n",
    "'cpd00009_e', # phosphate\n",
    "'cpd00063_e', # calcium\n",
    "'cpd00254_e', # magnesium\n",
    "'cpd10515_e', # Fe2+\n",
    "'cpd00030_e', # Mn2+\n",
    "'cpd00242_e', # Carbonate\n",
    "'cpd00226_e', # hypoxanthine\n",
    "'cpd01242_e', # thyminose\n",
    "'cpd00307_e', # cytosine\n",
    "'cpd00092_e', # uracil\n",
    "'cpd00117_e', # D-Alanine\n",
    "'cpd00067_e', # H+\n",
    "'cpd00567_e', # D-Proline\n",
    "'cpd00132_e', # L-Asparagine\n",
    "'cpd00210_e', # Taurine\n",
    "'cpd00320_e', # D-Aspartate\n",
    "'cpd03279_e', # Deoxyinosine\n",
    "'cpd00246_e', # Inosine\n",
    "'cpd00311_e', # Guanosine\n",
    "'cpd00367_e', # Cytidine\n",
    "'cpd00277_e', # Deoxyguanosine\n",
    "'cpd00182_e', # Adenosine\n",
    "'cpd00654_e', # Deoxycytidine\n",
    "'cpd00412_e', # Deoxyuridine\n",
    "'cpd00438_e', # Deoxyadenosine\n",
    "'cpd00274_e', # Citrulline\n",
    "'cpd00186_e', # D-Glutamate\n",
    "'cpd00637_e', # D-Methionine\n",
    "'cpd00105_e', # D-Ribose\n",
    "'cpd00305_e', # Thiamin\n",
    "'cpd00309_e', # Xanthine\n",
    "'cpd00098_e', # Choline\n",
    "'cpd00053_e', # L-Glutamine\n",
    "'cpd00039_e', # Lysine\n",
    "'cpd00064_e', # Ornithine\n",
    "'cpd00051_e', # D-arginine\n",
    "'cpd00207_e'] # Guanine\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "medium_exch = set(['EX_' + x for x in medium])\n",
    "for rxn in iCdR703.exchanges:\n",
    "    rxn.lower_bound = 0.\n",
    "    if rxn.id in medium_exch:\n",
    "        rxn.lower_bound = -1000."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "iCdR703.reactions.rxn00933_c.bounds = (1.,1000.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Running max fit RIPTiDe for objective fraction range: 0.35 to 0.95 with intervals of 0.05 \n",
      "\n",
      "Testing minimum objective fractions...\n",
      "Fraction = 0.35 | Rho = 0.1598 ; p = 0.0071\n",
      "Fraction = 0.4 | Rho = 0.1959 ; p = 0.0009\n",
      "Fraction = 0.45 | Rho = 0.1987 ; p = 0.0007\n",
      "Fraction = 0.5 | Rho = 0.1665 ; p = 0.0044\n",
      "Fraction = 0.55 | Rho = 0.1999 ; p = 0.0006\n",
      "Fraction = 0.6 | Rho = 0.1675 ; p = 0.0039\n",
      "Fraction = 0.65 | Rho = 0.1759 ; p = 0.0022\n",
      "Fraction = 0.7 | Rho = 0.1782 ; p = 0.0018\n",
      "Fraction = 0.75 | Rho = 0.173 ; p = 0.0022\n",
      "Fraction = 0.8 | Rho = 0.1324 ; p = 0.0185\n",
      "Fraction = 0.85 | Rho = 0.1324 ; p = 0.0182\n",
      "Fraction = 0.9 | Rho = 0.1371 ; p = 0.0138\n",
      "Testing local objective fractions to 0.55...\n",
      "Fraction = 0.525 | Rho = 0.2095 ; p = 0.0003\n",
      "Fraction = 0.575 | Rho = 0.1894 ; p = 0.0011\n",
      "\n",
      "Context-specific metabolism fit with 0.525 of optimal objective flux\n",
      "\n",
      "Reactions pruned to 291 from 1313 (77.84% change)\n",
      "Metabolites pruned to 290 from 1243 (76.67% change)\n",
      "Flux through the objective DECREASED to ~76.7 from ~125.53 (38.9% change)\n",
      "Context-specific metabolism correlates with transcriptome (r=0.21, p<0.001 *)\n",
      "\n",
      "Maxfit RIPTiDe completed in, 6 minutes and 44 seconds \n",
      "\n"
     ]
    }
   ],
   "source": [
    "iCdR703_solo = riptide.maxfit_contextualize(model=iCdR703, transcriptome=solo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Running max fit RIPTiDe for objective fraction range: 0.35 to 0.95 with intervals of 0.05 \n",
      "\n",
      "Testing minimum objective fractions...\n",
      "Fraction = 0.35 | Rho = 0.0053 ; p = 0.929\n",
      "Fraction = 0.4 | Rho = -0.04 ; p = 0.498\n",
      "Fraction = 0.45 | Rho = -0.0186 ; p = 0.7516\n",
      "Fraction = 0.5 | Rho = 0.0332 ; p = 0.5724\n",
      "Fraction = 0.55 | Rho = 0.0364 ; p = 0.5318\n",
      "Fraction = 0.6 | Rho = 0.0205 ; p = 0.7271\n",
      "Fraction = 0.65 | Rho = 0.0283 ; p = 0.6299\n",
      "Fraction = 0.7 | Rho = 0.0224 ; p = 0.7028\n",
      "Fraction = 0.75 | Rho = 0.0229 ; p = 0.6976\n",
      "Fraction = 0.8 | Rho = -0.0159 ; p = 0.7855\n",
      "Fraction = 0.85 | Rho = 0.0096 ; p = 0.8676\n",
      "Fraction = 0.9 | Rho = -0.016 ; p = 0.7791\n",
      "Testing local objective fractions to 0.55...\n",
      "Fraction = 0.525 | Rho = 0.028 ; p = 0.6312\n",
      "Fraction = 0.575 | Rho = 0.0371 ; p = 0.5246\n",
      "\n",
      "Context-specific metabolism fit with 0.575 of optimal objective flux\n",
      "\n",
      "Reactions pruned to 297 from 1313 (77.38% change)\n",
      "Metabolites pruned to 298 from 1243 (76.03% change)\n",
      "Flux through the objective DECREASED to ~98.87 from ~125.53 (21.24% change)\n",
      "Context-specific metabolism does not correlate with transcriptome (r=0.037, n.s.)\n",
      "\n",
      "Maxfit RIPTiDe completed in, 4 minutes and 55 seconds \n",
      "\n"
     ]
    }
   ],
   "source": [
    "iCdR703_coculture = riptide.maxfit_contextualize(model=iCdR703, transcriptome=coculture)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "riptide.save_output(riptide_obj=iCdR703_solo, path='/home/mjenior/Desktop/repos/Smith_etal_Enterococcus/data/solo_bhi_reps2')\n",
    "riptide.save_output(riptide_obj=iCdR703_coculture, path='/home/mjenior/Desktop/repos/Smith_etal_Enterococcus/data/coculture_bhi_reps2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "iCdR703_solo = cobra.io.read_sbml_model('/home/mjenior/Desktop/repos/Smith_etal_Enterococcus/data/solo_bhi_reps/model.sbml')\n",
    "iCdR703_coculture = cobra.io.read_sbml_model('/home/mjenior/Desktop/repos/Smith_etal_Enterococcus/data/coculture_bhi_reps/model.sbml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Essential substrates\n",
    "\n",
    "import pandas\n",
    "from cobra.medium import minimal_medium\n",
    "\n",
    "# Non-exhaustive\n",
    "def find_essential_metabolites(model, fraction=0.01):\n",
    "\n",
    "    min_growth = model.slim_optimize() * fraction\n",
    "    warnings.filterwarnings('ignore')\n",
    "    ser = minimal_medium(model, min_objective_value=min_growth, minimize_components=True)\n",
    "    warnings.filterwarnings('default')\n",
    "\n",
    "    rxns = list(ser.index)\n",
    "    fluxes = list(ser.values)\n",
    "    cpds = []\n",
    "    for x in rxns: cpds.append(model.reactions.get_by_id(x).reactants[0].name)        \n",
    "    \n",
    "    media = {'Reaction': rxns, 'Substrate': cpds, 'Units': fluxes}    \n",
    "    media = pandas.DataFrame(media)\n",
    "    print('Minimal media components: ' + str(len(media.index)))\n",
    "    \n",
    "    return(media)\n",
    "\n",
    "def growth_impact(model, essential, set_media=False):\n",
    "    min_media = set(list(essential['Reaction']))\n",
    "    \n",
    "    if set_media == True:\n",
    "        for rxn in model.exchanges:\n",
    "            rxn.lower_bound = 0.\n",
    "            if rxn.id in min_media: rxn.lower_bound = -1000.\n",
    "    \n",
    "    rxns = []\n",
    "    cpds = []\n",
    "    demand = []\n",
    "    opt_growth = model.slim_optimize()\n",
    "    for rxn in model.exchanges:\n",
    "\n",
    "        old_bound = rxn.lower_bound\n",
    "        rxn.lower_bound = old_bound * 1.0\n",
    "        sub_growth = model.slim_optimize()\n",
    "        current = 100. - ((sub_growth / opt_growth) * 100.)\n",
    "        rxn.lower_bound = old_bound\n",
    "        \n",
    "        if current > 1.0:\n",
    "            rxns.append(rxn.id)\n",
    "            cpds.append(rxn.reactants[0].name)\n",
    "            demand.append(current)\n",
    "        \n",
    "    impact = {'Reaction': rxns, 'Substrate': cpds, 'Demand': demand}\n",
    "    impact = pandas.DataFrame(impact)\n",
    "    \n",
    "    return impact\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimal media components: 37\n"
     ]
    }
   ],
   "source": [
    "solo_min_media = find_essential_metabolites(iCdR703_solo.model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Substrate</th>\n",
       "      <th>Units</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EX_cpd00082_e</td>\n",
       "      <td>D-Fructose</td>\n",
       "      <td>24.633114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EX_cpd00023_e</td>\n",
       "      <td>L-Glutamate</td>\n",
       "      <td>7.843940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EX_cpd00971_e</td>\n",
       "      <td>Na+</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EX_cpd00051_e</td>\n",
       "      <td>L-Arginine</td>\n",
       "      <td>2.577065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EX_cpd00064_e</td>\n",
       "      <td>Ornithine</td>\n",
       "      <td>3.206612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EX_cpd10515_e</td>\n",
       "      <td>Fe2+</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EX_cpd00009_e</td>\n",
       "      <td>Phosphate</td>\n",
       "      <td>23.050080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EX_cpd00060_e</td>\n",
       "      <td>L-Methionine</td>\n",
       "      <td>2.174398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>EX_cpd00035_e</td>\n",
       "      <td>L-Alanine</td>\n",
       "      <td>335.710032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EX_cpd00220_e</td>\n",
       "      <td>Riboflavin</td>\n",
       "      <td>0.044293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EX_cpd00117_e</td>\n",
       "      <td>D-Alanine</td>\n",
       "      <td>42.279966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EX_cpd00254_e</td>\n",
       "      <td>Mg</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EX_cpd00205_e</td>\n",
       "      <td>K+</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EX_cpd00305_e</td>\n",
       "      <td>Thiamin</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EX_cpd00065_e</td>\n",
       "      <td>L-Tryptophan</td>\n",
       "      <td>0.515413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EX_cpd00149_e</td>\n",
       "      <td>Co2+</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EX_cpd00099_e</td>\n",
       "      <td>Cl-</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EX_cpd00104_e</td>\n",
       "      <td>Biotin</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EX_cpd00063_e</td>\n",
       "      <td>Ca2+</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>EX_cpd00156_e</td>\n",
       "      <td>L-Valine</td>\n",
       "      <td>5.282982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>EX_cpd00393_e</td>\n",
       "      <td>Folate</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>EX_cpd00322_e</td>\n",
       "      <td>L-Isoleucine</td>\n",
       "      <td>8.021114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>EX_cpd00030_e</td>\n",
       "      <td>Mn2+</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>EX_cpd00084_e</td>\n",
       "      <td>L-Cysteine</td>\n",
       "      <td>1.042906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>EX_cpd00644_e</td>\n",
       "      <td>Pantothenate</td>\n",
       "      <td>0.084560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>EX_cpd00263_e</td>\n",
       "      <td>Pyridoxol</td>\n",
       "      <td>0.040267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>EX_cpd00107_e</td>\n",
       "      <td>L-Leucine</td>\n",
       "      <td>7.328527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>EX_cpd00069_e</td>\n",
       "      <td>L-Tyrosine</td>\n",
       "      <td>3.350184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>EX_cpd00066_e</td>\n",
       "      <td>L-Phenylalanine</td>\n",
       "      <td>3.527357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>EX_cpd00119_e</td>\n",
       "      <td>L-Histidine</td>\n",
       "      <td>1.095252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>EX_cpd00054_e</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>8.317073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>EX_cpd00041_e</td>\n",
       "      <td>L-Aspartate</td>\n",
       "      <td>166.422163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>EX_cpd00133_e</td>\n",
       "      <td>Nicotinamide</td>\n",
       "      <td>0.201333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>EX_cpd00311_e</td>\n",
       "      <td>Guanosine</td>\n",
       "      <td>355.237597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>EX_cpd00654_e</td>\n",
       "      <td>Deoxycytidine</td>\n",
       "      <td>1.268399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>EX_cpd00412_e</td>\n",
       "      <td>Deoxyuridine</td>\n",
       "      <td>528.654817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>EX_cpd00186_e</td>\n",
       "      <td>D-Glutamate</td>\n",
       "      <td>24.159981</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Reaction        Substrate       Units\n",
       "0   EX_cpd00082_e       D-Fructose   24.633114\n",
       "1   EX_cpd00023_e      L-Glutamate    7.843940\n",
       "2   EX_cpd00971_e              Na+    0.040267\n",
       "3   EX_cpd00051_e       L-Arginine    2.577065\n",
       "4   EX_cpd00064_e        Ornithine    3.206612\n",
       "5   EX_cpd10515_e             Fe2+    0.040267\n",
       "6   EX_cpd00009_e        Phosphate   23.050080\n",
       "7   EX_cpd00060_e     L-Methionine    2.174398\n",
       "8   EX_cpd00035_e        L-Alanine  335.710032\n",
       "9   EX_cpd00220_e       Riboflavin    0.044293\n",
       "10  EX_cpd00117_e        D-Alanine   42.279966\n",
       "11  EX_cpd00254_e               Mg    0.040267\n",
       "12  EX_cpd00205_e               K+    0.040267\n",
       "13  EX_cpd00305_e          Thiamin    0.040267\n",
       "14  EX_cpd00065_e     L-Tryptophan    0.515413\n",
       "15  EX_cpd00149_e             Co2+    0.040267\n",
       "16  EX_cpd00099_e              Cl-    0.040267\n",
       "17  EX_cpd00104_e           Biotin    0.040267\n",
       "18  EX_cpd00063_e             Ca2+    0.040267\n",
       "19  EX_cpd00156_e         L-Valine    5.282982\n",
       "20  EX_cpd00393_e           Folate    0.040267\n",
       "21  EX_cpd00322_e     L-Isoleucine    8.021114\n",
       "22  EX_cpd00030_e             Mn2+    0.040267\n",
       "23  EX_cpd00084_e       L-Cysteine    1.042906\n",
       "24  EX_cpd00644_e     Pantothenate    0.084560\n",
       "25  EX_cpd00263_e        Pyridoxol    0.040267\n",
       "26  EX_cpd00107_e        L-Leucine    7.328527\n",
       "27  EX_cpd00069_e       L-Tyrosine    3.350184\n",
       "28  EX_cpd00066_e  L-Phenylalanine    3.527357\n",
       "29  EX_cpd00119_e      L-Histidine    1.095252\n",
       "30  EX_cpd00054_e         L-Serine    8.317073\n",
       "31  EX_cpd00041_e      L-Aspartate  166.422163\n",
       "32  EX_cpd00133_e     Nicotinamide    0.201333\n",
       "33  EX_cpd00311_e        Guanosine  355.237597\n",
       "34  EX_cpd00654_e    Deoxycytidine    1.268399\n",
       "35  EX_cpd00412_e     Deoxyuridine  528.654817\n",
       "36  EX_cpd00186_e      D-Glutamate   24.159981"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solo_min_media"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimal media components: 42\n"
     ]
    }
   ],
   "source": [
    "coculture_min_media = find_essential_metabolites(iCdR703_coculture.model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Substrate</th>\n",
       "      <th>Units</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EX_cpd00082_e</td>\n",
       "      <td>D-Fructose</td>\n",
       "      <td>0.014212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EX_cpd00023_e</td>\n",
       "      <td>L-Glutamate</td>\n",
       "      <td>11.074325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EX_cpd00971_e</td>\n",
       "      <td>Na+</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EX_cpd00051_e</td>\n",
       "      <td>L-Arginine</td>\n",
       "      <td>3.638382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EX_cpd00064_e</td>\n",
       "      <td>Ornithine</td>\n",
       "      <td>369.606675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EX_cpd10515_e</td>\n",
       "      <td>Fe2+</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EX_cpd00009_e</td>\n",
       "      <td>Phosphate</td>\n",
       "      <td>32.542838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EX_cpd00060_e</td>\n",
       "      <td>L-Methionine</td>\n",
       "      <td>3.069885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>EX_cpd00035_e</td>\n",
       "      <td>L-Alanine</td>\n",
       "      <td>286.862934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EX_cpd00220_e</td>\n",
       "      <td>Riboflavin</td>\n",
       "      <td>0.062535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EX_cpd00117_e</td>\n",
       "      <td>D-Alanine</td>\n",
       "      <td>59.692205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EX_cpd00254_e</td>\n",
       "      <td>Mg</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EX_cpd00205_e</td>\n",
       "      <td>K+</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EX_cpd00305_e</td>\n",
       "      <td>Thiamin</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EX_cpd00065_e</td>\n",
       "      <td>L-Tryptophan</td>\n",
       "      <td>0.727676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EX_cpd00149_e</td>\n",
       "      <td>Co2+</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EX_cpd00099_e</td>\n",
       "      <td>Cl-</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EX_cpd00104_e</td>\n",
       "      <td>Biotin</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EX_cpd00063_e</td>\n",
       "      <td>Ca2+</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>EX_cpd00156_e</td>\n",
       "      <td>L-Valine</td>\n",
       "      <td>7.458683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>EX_cpd00039_e</td>\n",
       "      <td>L-Lysine</td>\n",
       "      <td>27.447044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>EX_cpd00393_e</td>\n",
       "      <td>Folate</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>EX_cpd00322_e</td>\n",
       "      <td>L-Isoleucine</td>\n",
       "      <td>267.626770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>EX_cpd00030_e</td>\n",
       "      <td>Mn2+</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>EX_cpd00084_e</td>\n",
       "      <td>L-Cysteine</td>\n",
       "      <td>1.472408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>EX_cpd00644_e</td>\n",
       "      <td>Pantothenate</td>\n",
       "      <td>0.119384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>EX_cpd00263_e</td>\n",
       "      <td>Pyridoxol</td>\n",
       "      <td>0.056850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>EX_cpd00107_e</td>\n",
       "      <td>L-Leucine</td>\n",
       "      <td>10.346649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>EX_cpd00069_e</td>\n",
       "      <td>L-Tyrosine</td>\n",
       "      <td>4.729897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>EX_cpd00066_e</td>\n",
       "      <td>L-Phenylalanine</td>\n",
       "      <td>4.980035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>EX_cpd00119_e</td>\n",
       "      <td>L-Histidine</td>\n",
       "      <td>1.546312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>EX_cpd00132_e</td>\n",
       "      <td>L-Asparagine</td>\n",
       "      <td>7.322244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>EX_cpd00161_e</td>\n",
       "      <td>L-Threonine</td>\n",
       "      <td>5.639492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>EX_cpd00041_e</td>\n",
       "      <td>L-Aspartate</td>\n",
       "      <td>24.894492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>EX_cpd00133_e</td>\n",
       "      <td>Nicotinamide</td>\n",
       "      <td>0.284249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>EX_cpd00246_e</td>\n",
       "      <td>Inosine</td>\n",
       "      <td>310.647774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>EX_cpd00311_e</td>\n",
       "      <td>Guanosine</td>\n",
       "      <td>2.119416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>EX_cpd00277_e</td>\n",
       "      <td>Deoxyguanosine</td>\n",
       "      <td>0.875486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>EX_cpd00182_e</td>\n",
       "      <td>Adenosine</td>\n",
       "      <td>3.622903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>EX_cpd00654_e</td>\n",
       "      <td>Deoxycytidine</td>\n",
       "      <td>1.790766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>EX_cpd00412_e</td>\n",
       "      <td>Deoxyuridine</td>\n",
       "      <td>476.721836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>EX_cpd00186_e</td>\n",
       "      <td>D-Glutamate</td>\n",
       "      <td>34.109831</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Reaction        Substrate       Units\n",
       "0   EX_cpd00082_e       D-Fructose    0.014212\n",
       "1   EX_cpd00023_e      L-Glutamate   11.074325\n",
       "2   EX_cpd00971_e              Na+    0.056850\n",
       "3   EX_cpd00051_e       L-Arginine    3.638382\n",
       "4   EX_cpd00064_e        Ornithine  369.606675\n",
       "5   EX_cpd10515_e             Fe2+    0.056850\n",
       "6   EX_cpd00009_e        Phosphate   32.542838\n",
       "7   EX_cpd00060_e     L-Methionine    3.069885\n",
       "8   EX_cpd00035_e        L-Alanine  286.862934\n",
       "9   EX_cpd00220_e       Riboflavin    0.062535\n",
       "10  EX_cpd00117_e        D-Alanine   59.692205\n",
       "11  EX_cpd00254_e               Mg    0.056850\n",
       "12  EX_cpd00205_e               K+    0.056850\n",
       "13  EX_cpd00305_e          Thiamin    0.056850\n",
       "14  EX_cpd00065_e     L-Tryptophan    0.727676\n",
       "15  EX_cpd00149_e             Co2+    0.056850\n",
       "16  EX_cpd00099_e              Cl-    0.056850\n",
       "17  EX_cpd00104_e           Biotin    0.056850\n",
       "18  EX_cpd00063_e             Ca2+    0.056850\n",
       "19  EX_cpd00156_e         L-Valine    7.458683\n",
       "20  EX_cpd00039_e         L-Lysine   27.447044\n",
       "21  EX_cpd00393_e           Folate    0.056850\n",
       "22  EX_cpd00322_e     L-Isoleucine  267.626770\n",
       "23  EX_cpd00030_e             Mn2+    0.056850\n",
       "24  EX_cpd00084_e       L-Cysteine    1.472408\n",
       "25  EX_cpd00644_e     Pantothenate    0.119384\n",
       "26  EX_cpd00263_e        Pyridoxol    0.056850\n",
       "27  EX_cpd00107_e        L-Leucine   10.346649\n",
       "28  EX_cpd00069_e       L-Tyrosine    4.729897\n",
       "29  EX_cpd00066_e  L-Phenylalanine    4.980035\n",
       "30  EX_cpd00119_e      L-Histidine    1.546312\n",
       "31  EX_cpd00132_e     L-Asparagine    7.322244\n",
       "32  EX_cpd00161_e      L-Threonine    5.639492\n",
       "33  EX_cpd00041_e      L-Aspartate   24.894492\n",
       "34  EX_cpd00133_e     Nicotinamide    0.284249\n",
       "35  EX_cpd00246_e          Inosine  310.647774\n",
       "36  EX_cpd00311_e        Guanosine    2.119416\n",
       "37  EX_cpd00277_e   Deoxyguanosine    0.875486\n",
       "38  EX_cpd00182_e        Adenosine    3.622903\n",
       "39  EX_cpd00654_e    Deoxycytidine    1.790766\n",
       "40  EX_cpd00412_e     Deoxyuridine  476.721836\n",
       "41  EX_cpd00186_e      D-Glutamate   34.109831"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coculture_min_media"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shared: 36\n",
      "\n",
      "Solo only: 1\n",
      "Coculture only: 6\n"
     ]
    }
   ],
   "source": [
    "solo_substrates = set(solo_min_media['Substrate'])\n",
    "coculture_substrates = set(coculture_min_media['Substrate'])\n",
    "\n",
    "shared_substrates = solo_substrates.intersection(coculture_substrates)\n",
    "solo_only_substrates = solo_substrates.difference(coculture_substrates)\n",
    "coculture_only_substrates = coculture_substrates.difference(solo_substrates)\n",
    "\n",
    "print('Shared: ' + str(len(shared_substrates)) + '\\n')\n",
    "print('Solo only: ' + str(len(solo_only_substrates)))\n",
    "print('Coculture only: ' + str(len(coculture_only_substrates)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'L-Serine'}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solo_only_substrates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Adenosine',\n",
       " 'Deoxyguanosine',\n",
       " 'Inosine',\n",
       " 'L-Asparagine',\n",
       " 'L-Lysine',\n",
       " 'L-Threonine'}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coculture_only_substrates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCdG703_riptide</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f7defed37d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>290</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>291</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>0.0 + 1.0*biomass - 1.0*biomass_reverse_01e59</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCdG703_riptide at 0x7f7defed37d0>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iCdR703_solo.model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "215"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(iCdR703_solo.model.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCdG703_riptide</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f7de570d390</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>298</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>297</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>0.0 + 1.0*biomass - 1.0*biomass_reverse_01e59</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCdG703_riptide at 0x7f7de570d390>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iCdR703_coculture.model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "218"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(iCdR703_coculture.model.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo reactions: 25\n",
      "rxn01069_c O-phospho-L-homoserine phosphate-lyase (adding water;L-threonine-forming)\n",
      "EX_cpd00207_e Guanine exchange\n",
      "rxn08686_c H2O transport via diffusion\n",
      "rxn01859_c Deoxyadenosine:orthophosphate ribosyltransferase\n",
      "EX_cpd00001_e H2O exchange\n",
      "rxn12188_c 2-deoxyglucose-6-phosphatase\n",
      "rxn01138_c Adenosine:phosphate alpha-D-ribosyltransferase\n",
      "EX_cpd03279_e Deoxyinosine exchange\n",
      "rxn05317_c Transport of deoxyadenosine, extracellular\n",
      "rxn01300_c ATP:L-homoserine O-phosphotransferase\n",
      "EX_cpd00438_e Deoxyadenosine exchange\n",
      "rxn10036_c b-ketoacyl synthetase (n-C16:1)\n",
      "rxn10770_c Choline dehydrogenase\n",
      "rxn00223_c D-glucose aldose-ketose-isomerase\n",
      "rxn01302_c L-homoserine:NADP+ oxidoreductase\n",
      "rxn01985_c Deoxyinosine:orthophosphate ribosyltransferase\n",
      "rxn00313_c meso-2,6-diaminoheptanedioate carboxy-lyase (L-lysine-forming)\n",
      "rxn12191_c Choline monooxygenase (NADP)\n",
      "rxn00340_c L-aspartate:ammonia ligase (AMP-forming)\n",
      "rxn05203_c Guanine symport\n",
      "rxn09687_c Deoxyinosine transport in via proton symport\n",
      "rxn05234_c 2'-Deoxyguanosine 5'-triphosphate:oxydized-thioredoxin 2'-oxidoreductase\n",
      "rxn10663_c Fatty acid oxidation (n-C16:1)\n",
      "rxn00558_c D-glucose-6-phosphate aldose-ketose-isomerase\n",
      "rxn01446_c Deoxyguanosine:orthophosphate ribosyltransferase\n",
      "\n",
      "Coculture-only reactions: 31\n",
      "rxn00022_c maltose glucohydrolase\n",
      "rxn01780_c Ornithine aminomutase\n",
      "EX_cpd19585_e 2-Methylbutyrate exchange\n",
      "rxn01575_c L-isoleucine aminotransferase\n",
      "EX_cpd00367_e Cytidine exchange\n",
      "rxn10144_c ATP phosphohydrolase (polar-amino-acid-importing)\n",
      "EX_cpd00307_e Cytosine exchange\n",
      "rxn05466_c Ammonia diffusion\n",
      "rxn27496_c 2,4-diaminopentanoate dehydrogenase\n",
      "rxn13251_c Fumarate dependent DHORD\n",
      "rxn19080_c 2-amino-4-oxopentanoate thiolase\n",
      "EX_cpd00161_e L-Threonine exchange\n",
      "rxn05198_c Transport of cytidine, extracellular\n",
      "rxn05527_c Cytosine hydrogen symport\n",
      "EX_cpd00013_e Ammonia exchange\n",
      "rxn05151_c ATP phosphohydrolase (polar-amino-acid-importing)\n",
      "rxn05318_c Adenosine hydrogen symport\n",
      "EX_cpd00182_e Adenosine exchange\n",
      "rxn01359_c (S)-dihydroorotate:NADP+ oxidoreductase\n",
      "rxn08168_c L-aspartate:fumarate oxidoreductase (deaminating)\n",
      "rxn11676_c Glycine reductase\n",
      "rxn05169_c ATP phosphohydrolase (polar-amino-acid-importing)\n",
      "rxn05227_c Aspartate oxidase\n",
      "ID003_c 2-Methylbutyrate diffusion\n",
      "rxn02269_c (S)-3-methyl-2-oxopentanoate:NAD+ 2-oxidoreductase (CoA-methylbutanoylating)\n",
      "EX_cpd00132_e L-Asparagine exchange\n",
      "EX_cpd00039_e L-Lysine exchange\n",
      "rxn01132_c Maltose:phosphate 1-beta-D-glucosyltransferase\n",
      "rxn02380_c beta-D-Glucose 6-phosphate ketol-isomerase\n",
      "rxn01648_c Cytidine:orthophosphate alpha-D-ribosyltransferase\n",
      "R01174_3_c Phosphate butyryltransferase\n"
     ]
    }
   ],
   "source": [
    "# Distinct reactions - BHI\n",
    "rxns1 = set([x.id for x in iCdR703_solo.model.reactions])\n",
    "rxns2 = set([x.id for x in iCdR703_coculture.model.reactions])\n",
    "\n",
    "solo_only_rxns = rxns1.difference(rxns2)\n",
    "print('Solo reactions:', len(solo_only_rxns))\n",
    "for x in solo_only_rxns: print(x, iCdR703.reactions.get_by_id(x).name)\n",
    "\n",
    "coculture_only_rxns = rxns2.difference(rxns1)\n",
    "print('\\nCoculture-only reactions:', len(coculture_only_rxns))\n",
    "for x in coculture_only_rxns: print(x, iCdR703.reactions.get_by_id(x).name)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shared exchange reactions: 47\n",
      "EX_cpd00082_e D-Fructose exchange\n",
      "EX_cpd00041_e L-Aspartate exchange\n",
      "EX_cpd00065_e L-Tryptophan exchange\n",
      "EX_cpd00305_e Thiamin exchange\n",
      "EX_cpd00060_e L-Methionine exchange\n",
      "EX_cpd00254_e Mg exchange\n",
      "EX_cpd00311_e Guanosine exchange\n",
      "EX_cpd00277_e Deoxyguanosine exchange\n",
      "EX_cpd00263_e Pyridoxol exchange\n",
      "EX_cpd00122_e N-Acetyl-D-glucosamine exchange\n",
      "EX_cpd00069_e L-Tyrosine exchange\n",
      "EX_cpd00205_e K+ exchange\n",
      "EX_cpd00339_e 5-Aminopentanoate exchange\n",
      "EX_cpd00064_e Ornithine exchange\n",
      "EX_cpd00412_e Deoxyuridine exchange\n",
      "EX_cpd00009_e Phosphate exchange\n",
      "EX_cpd00023_e L-Glutamate exchange\n",
      "EX_cpd00107_e L-Leucine exchange\n",
      "EX_cpd00063_e Ca2+ exchange\n",
      "EX_cpd00099_e Cl- exchange\n",
      "EX_cpd10515_e Fe2+ exchange\n",
      "EX_cpd00066_e L-Phenylalanine exchange\n",
      "EX_cpd00067_e H+ exchange\n",
      "EX_cpd00220_e Riboflavin exchange\n",
      "EX_cpd00084_e L-Cysteine exchange\n",
      "EX_cpd00654_e Deoxycytidine exchange\n",
      "EX_cpd00047_e Formate exchange\n",
      "EX_cpd00644_e Pantothenate exchange\n",
      "EX_cpd00156_e L-Valine exchange\n",
      "EX_cpd00051_e L-Arginine exchange\n",
      "EX_cpd00322_e L-Isoleucine exchange\n",
      "EX_cpd00054_e L-Serine exchange\n",
      "EX_cpd00226_e Hypoxanthine exchange\n",
      "EX_cpd00393_e Folate exchange\n",
      "EX_cpd00246_e Inosine exchange\n",
      "EX_cpd00104_e Biotin exchange\n",
      "EX_cpd00092_e Uracil exchange\n",
      "EX_cpd00133_e Nicotinamide exchange\n",
      "EX_cpd00971_e Na+ exchange\n",
      "EX_cpd00030_e Mn2+ exchange\n",
      "EX_cpd00119_e L-Histidine exchange\n",
      "EX_cpd00186_e D-Glutamate exchange\n",
      "EX_cpd00117_e D-Alanine exchange\n",
      "EX_cpd00149_e Co2+ exchange\n",
      "EX_cpd00011_e CO2 exchange\n",
      "EX_cpd00029_e Acetate exchange\n",
      "EX_cpd00035_e L-Alanine exchange\n"
     ]
    }
   ],
   "source": [
    "# Shared exchanges\n",
    "rxns1 = set([x.id for x in iCdR703_solo.model.exchanges])\n",
    "rxns2 = set([x.id for x in iCdR703_coculture.model.exchanges])\n",
    "shared_exch = rxns1.intersection(rxns2)\n",
    "\n",
    "print('Shared exchange reactions:', len(shared_exch))\n",
    "for x in shared_exch: print(x, iCdR703.reactions.get_by_id(x).name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo culture-only metabolites: 13\n",
      "cpd00098_c Choline\n",
      "cpd00001_e H2O\n",
      "cpd15237_c hexadecenoate\n",
      "cpd00447_c Betaine aldehyde\n",
      "cpd00809_c O-Phospho-L-homoserine\n",
      "cpd00438_e Deoxyadenosine\n",
      "cpd03279_c Deoxyinosine\n",
      "cpd00438_c Deoxyadenosine\n",
      "cpd00227_c L-Homoserine\n",
      "cpd03279_e Deoxyinosine\n",
      "cpd00082_c D-Fructose\n",
      "cpd00207_c Guanine\n",
      "cpd00207_e Guanine\n",
      "\n",
      "Coculture-only metabolites: 21\n",
      "cpd03470_c Iminoaspartate\n",
      "cpd00039_e L-Lysine\n",
      "cpd00367_c Cytidine\n",
      "cpd00182_e Adenosine\n",
      "cpd00760_c 2-Methylbutyryl-CoA\n",
      "cpd00106_c Fumarate\n",
      "cpd24394_c 2-amino-4-ketopentanoate\n",
      "cpd02448_c 2,4-diaminopentanoate\n",
      "cpd19585_e 2-Methylbutyrate\n",
      "cpd00307_e Cytosine\n",
      "cpd00179_c Maltose\n",
      "cpd00508_c (3S)-3-Methyl-2-oxopentanoate\n",
      "cpd00036_c Succinate\n",
      "cpd00282_c S-Dihydroorotate\n",
      "cpd00013_e Ammonia\n",
      "cpd00161_e L-Threonine\n",
      "cpd00307_c Cytosine\n",
      "cpd00132_e L-Asparagine\n",
      "cpd00367_e Cytidine\n",
      "cpd00247_c Orotate\n",
      "cpd19585_c 2-Methylbutyrate\n"
     ]
    }
   ],
   "source": [
    "# Distinct metabolites - BHI\n",
    "cpds1 = set([x.id for x in iCdR703_solo.model.metabolites])\n",
    "cpds2 = set([x.id for x in iCdR703_coculture.model.metabolites])\n",
    "\n",
    "solo_only_cpds = cpds1.difference(cpds2)\n",
    "print('Solo culture-only metabolites:', len(solo_only_cpds))\n",
    "for x in solo_only_cpds: print(x, iCdR703.metabolites.get_by_id(x).name)\n",
    "\n",
    "coculture_only_cpds = cpds2.difference(cpds1)\n",
    "print('\\nCoculture-only metabolites:', len(coculture_only_cpds))\n",
    "for x in coculture_only_cpds: print(x, iCdR703.metabolites.get_by_id(x).name) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo-only genes: 6\n",
      "645463.3.peg.2100 diaminopimelate decarboxylase (EC 4.1.1.20)\n",
      "645463.3.peg.2743 Aspartate--ammonia ligase (EC 6.3.1.1)\n",
      "645463.3.peg.3063 Xylose isomerase (EC 5.3.1.5)\n",
      "645463.3.peg.2170 Homoserine kinase (EC 2.7.1.39)\n",
      "645463.3.peg.2169 Threonine synthase (EC 4.2.3.1)\n",
      "645463.3.peg.1817 diaminopimelate decarboxylase (EC 4.1.1.20)\n",
      "\n",
      "Coculture-only genes: 9\n",
      "645463.3.peg.416 D-Ornithine 4,5-aminomutase E subunit (EC 5.4.3.5) | oraE\n",
      "645463.3.peg.3013 alpha-mannosidase (EC 3.2.1.24)\n",
      "645463.3.peg.414 2-amino-4-ketopentanoate thiolase beta subunit [EC:2.3.1.263] | ortA\n",
      "645463.3.peg.2389 Glycine reductase component B beta subunit\n",
      "645463.3.peg.1421 DNA-3-methyladenine glycosylase II\n",
      "645463.3.peg.2787 Cytosine permease\n",
      "645463.3.peg.413 2-amino-4-ketopentanoate thiolase alpha subunit [EC:2.3.1.263] | ortA\n",
      "645463.3.peg.412 2,4-diaminopentanoate dehydrogenase (EC 1.4.1.12)\n",
      "645463.3.peg.2408 L-aspartate oxidase (EC 1.4.3.16)\n"
     ]
    }
   ],
   "source": [
    "# Distinct genes - BHI\n",
    "genes1 = set([x.id for x in iCdR703_solo.model.genes])\n",
    "genes2 = set([x.id for x in iCdR703_coculture.model.genes])\n",
    "\n",
    "solo_only_genes = genes1.difference(genes2)\n",
    "print('Solo-only genes:', len(solo_only_genes))\n",
    "for x in solo_only_genes: print(x, iCdR703.genes.get_by_id(x).name)\n",
    "\n",
    "coculture_only_genes = genes2.difference(genes1)\n",
    "print('\\nCoculture-only genes:', len(coculture_only_genes))\n",
    "for x in coculture_only_genes: print(x, iCdR703.genes.get_by_id(x).name) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo-only essential genes: 6\n",
      "645463.3.peg.713 Glycine cleavage system H protein\n",
      "645463.3.peg.233 phosphoribosylglycinamide formyltransferase (EC 2.1.2.2)\n",
      "645463.3.peg.2743 Aspartate--ammonia ligase (EC 6.3.1.1)\n",
      "645463.3.peg.1428 Deoxyribose-phosphate aldolase (EC 4.1.2.4)\n",
      "645463.3.peg.2170 Homoserine kinase (EC 2.7.1.39)\n",
      "645463.3.peg.2169 Threonine synthase (EC 4.2.3.1)\n",
      "\n",
      "Coculture-only essential genes: 3\n",
      "645463.3.peg.3413 Pyruvate kinase (EC 2.7.1.40)\n",
      "645463.3.peg.3013 alpha-mannosidase (EC 3.2.1.24)\n",
      "645463.3.peg.1830 Adenine deaminase (EC 3.5.4.2)\n"
     ]
    }
   ],
   "source": [
    "# Differential gene essentiality\n",
    "solo_threshold = iCdR703_solo.model.slim_optimize() * 0.01\n",
    "solo_essential_bhi = find_essential_genes(iCdR703_solo.model, threshold=solo_threshold)\n",
    "solo_essential_bhi = set([x.id for x in solo_essential_bhi])\n",
    "coculture_threshold = iCdR703_coculture.model.slim_optimize() * 0.01\n",
    "coculture_essential_bhi = find_essential_genes(iCdR703_coculture.model, threshold=coculture_threshold)\n",
    "coculture_essential_bhi = set([x.id for x in coculture_essential_bhi])\n",
    "\n",
    "solo_essential_bhi_only = solo_essential_bhi.difference(coculture_essential_bhi)\n",
    "coculture_essential_bhi_only = coculture_essential_bhi.difference(solo_essential_bhi)\n",
    "essential_bhi = solo_essential_bhi.intersection(coculture_essential_bhi)\n",
    "\n",
    "both_contexts = solo_essential_bhi.intersection(coculture_essential_bhi)\n",
    "\n",
    "print('Solo-only essential genes:', len(solo_essential_bhi_only))\n",
    "for x in solo_essential_bhi_only: print(x, iCdR703.genes.get_by_id(x).name)\n",
    "print('\\nCoculture-only essential genes:', len(coculture_essential_bhi_only))\n",
    "for x in coculture_essential_bhi_only: print(x, iCdR703.genes.get_by_id(x).name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo only 22\n",
      "rxn01069_c O-phospho-L-homoserine phosphate-lyase (adding water;L-threonine-forming)\n",
      "rxn00692_c 5,10-Methylenetetrahydrofolate:glycine hydroxymethyltransferase\n",
      "rxn12188_c 2-deoxyglucose-6-phosphatase\n",
      "rxn01138_c Adenosine:phosphate alpha-D-ribosyltransferase\n",
      "EX_cpd00064_e Ornithine exchange\n",
      "rxn01300_c ATP:L-homoserine O-phosphotransferase\n",
      "rxn10840_c L-arginine transport via diffusion (extracellular to periplasm)\n",
      "rxn10883_c ATP phosphohydrolase (polar-amino-acid-importing)\n",
      "rxn01986_c 2-deoxy-D-ribose 1-phosphate 1,5-phosphomutase\n",
      "rxn00223_c D-glucose aldose-ketose-isomerase\n",
      "rxn01302_c L-homoserine:NADP+ oxidoreductase\n",
      "rxn00313_c meso-2,6-diaminoheptanedioate carboxy-lyase (L-lysine-forming)\n",
      "rxn00340_c L-aspartate:ammonia ligase (AMP-forming)\n",
      "rxn05289_c NADPH:oxidized-thioredoxin oxidoreductase\n",
      "EX_cpd00054_e L-Serine exchange\n",
      "rxn00908_c glycine synthase\n",
      "rxn01345_c Sedoheptulose 1,7-bisphosphate 1-phosphohydrolase\n",
      "rxn00784_c 2-deoxy-D-ribose-5-phosphate acetaldehyde-lyase (D-glyceraldehyde-3-phosphate-forming)\n",
      "rxn10131_c arginine/ornithine antiporter (periplasm)\n",
      "rxn00558_c D-glucose-6-phosphate aldose-ketose-isomerase\n",
      "rxn00171_c acetaldehyde:NAD+ oxidoreductase (CoA-acetylating)\n",
      "rxn01334_c Sedoheptulose 1,7-bisphosphate D-glyceraldehyde-3-phosphate-lyase\n",
      "\n",
      "Coculture only 25\n",
      "rxn00022_c maltose glucohydrolase\n",
      "rxn00148_c ATP:pyruvate 2-O-phosphotransferase\n",
      "EX_cpd00277_e Deoxyguanosine exchange\n",
      "rxn10144_c ATP phosphohydrolase (polar-amino-acid-importing)\n",
      "rxn01200_c Sedoheptulose-7-phosphate:D-glyceraldehyde-3-phosphate glycolaldehyde transferase\n",
      "rxn05466_c Ammonia diffusion\n",
      "rxn09688_c deoxyguanosine transport in via proton symport (periplasm)\n",
      "rxn08061_c Acetate transport via proton symport\n",
      "EX_cpd00161_e L-Threonine exchange\n",
      "rxn01116_c D-Ribulose-5-phosphate 3-epimerase\n",
      "EX_cpd00013_e Ammonia exchange\n",
      "rxn05151_c ATP phosphohydrolase (polar-amino-acid-importing)\n",
      "rxn05318_c Adenosine hydrogen symport\n",
      "EX_cpd00182_e Adenosine exchange\n",
      "rxn00785_c D-Fructose 6-phosphate:D-glyceraldehyde-3-phosphate glycolaldehyde transferase\n",
      "rxn13140_c Nucleoside triphosphatase\n",
      "rxn05169_c ATP phosphohydrolase (polar-amino-acid-importing)\n",
      "EX_cpd00132_e L-Asparagine exchange\n",
      "EX_cpd00039_e L-Lysine exchange\n",
      "rxn00777_c D-ribose-5-phosphate aldose-ketose-isomerase\n",
      "rxn01132_c Maltose:phosphate 1-beta-D-glucosyltransferase\n",
      "rxn02380_c beta-D-Glucose 6-phosphate ketol-isomerase\n",
      "rxn00778_c D-Ribose 1,5-phosphomutase\n",
      "rxn00926_c Adenine aminohydrolase\n",
      "EX_cpd00029_e Acetate exchange\n"
     ]
    }
   ],
   "source": [
    "# Differential reaction essentiality\n",
    "solo_essential_bhi = find_essential_reactions(iCdR703_solo.model)\n",
    "solo_essential_bhi = set([x.id for x in solo_essential_bhi])\n",
    "coculture_essential_bhi = find_essential_reactions(iCdR703_coculture.model)\n",
    "coculture_essential_bhi = set([x.id for x in coculture_essential_bhi])\n",
    "solo_essential_bhi_only = solo_essential_bhi.difference(coculture_essential_bhi)\n",
    "coculture_essential_bhi_only = coculture_essential_bhi.difference(solo_essential_bhi)\n",
    "essential_bhi = solo_essential_bhi.intersection(coculture_essential_bhi)\n",
    "\n",
    "print('Solo only', len(solo_essential_bhi_only))\n",
    "for x in solo_essential_bhi_only: print(x, iCdR703.reactions.get_by_id(x).name)   \n",
    "print('\\nCoculture only', len(coculture_essential_bhi_only))\n",
    "for x in coculture_essential_bhi_only: print(x, iCdR703.reactions.get_by_id(x).name)   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo in BHI only 4\n",
      "EX_cpd00438_e Deoxyadenosine exchange\n",
      "EX_cpd03279_e Deoxyinosine exchange\n",
      "EX_cpd00001_e H2O exchange\n",
      "EX_cpd00207_e Guanine exchange\n",
      "\n",
      "Coculture in BHI only 8\n",
      "EX_cpd00161_e L-Threonine exchange\n",
      "EX_cpd00182_e Adenosine exchange\n",
      "EX_cpd19585_e 2-Methylbutyrate exchange\n",
      "EX_cpd00132_e L-Asparagine exchange\n",
      "EX_cpd00039_e L-Lysine exchange\n",
      "EX_cpd00367_e Cytidine exchange\n",
      "EX_cpd00307_e Cytosine exchange\n",
      "EX_cpd00013_e Ammonia exchange\n"
     ]
    }
   ],
   "source": [
    "# Exchanges\n",
    "solo_exch = set([x.id for x in iCdR703_solo.model.exchanges])\n",
    "coculture_exch = set([x.id for x in iCdR703_coculture.model.exchanges])\n",
    "\n",
    "solo_exch_only = solo_exch.difference(coculture_exch)\n",
    "coculture_exch_only = coculture_exch.difference(solo_exch)\n",
    "\n",
    "print('Solo in BHI only', len(solo_exch_only))\n",
    "for x in solo_exch_only: print(x, iCdR703.reactions.get_by_id(x).name)   \n",
    "print('\\nCoculture in BHI only', len(coculture_exch_only))\n",
    "for x in coculture_exch_only: print(x, iCdR703.reactions.get_by_id(x).name)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D-Fructose EX_cpd00082_e -28.664\n",
      "L-Glutamate EX_cpd00023_e -9.127\n",
      "Na+ EX_cpd00971_e -0.047\n",
      "L-Arginine EX_cpd00051_e -2.999\n",
      "Ornithine EX_cpd00064_e -60.834\n",
      "Fe2+ EX_cpd10515_e -0.047\n",
      "Phosphate EX_cpd00009_e -26.822\n",
      "L-Methionine EX_cpd00060_e -2.53\n",
      "L-Alanine EX_cpd00035_e -381.117\n",
      "Riboflavin EX_cpd00220_e -0.052\n",
      "D-Alanine EX_cpd00117_e -49.198\n",
      "Mg EX_cpd00254_e -0.047\n",
      "K+ EX_cpd00205_e -0.047\n",
      "Thiamin EX_cpd00305_e -0.047\n",
      "L-Tryptophan EX_cpd00065_e -0.6\n",
      "Co2+ EX_cpd00149_e -0.047\n",
      "Cl- EX_cpd00099_e -0.047\n",
      "Biotin EX_cpd00104_e -0.047\n",
      "Ca2+ EX_cpd00063_e -0.047\n",
      "L-Valine EX_cpd00156_e -6.147\n",
      "Folate EX_cpd00393_e -0.047\n",
      "L-Isoleucine EX_cpd00322_e -9.334\n",
      "Mn2+ EX_cpd00030_e -0.047\n",
      "L-Cysteine EX_cpd00084_e -1.214\n",
      "Pantothenate EX_cpd00644_e -0.098\n",
      "Pyridoxol EX_cpd00263_e -0.047\n",
      "L-Leucine EX_cpd00107_e -8.528\n",
      "L-Tyrosine EX_cpd00069_e -3.898\n",
      "L-Phenylalanine EX_cpd00066_e -4.105\n",
      "L-Histidine EX_cpd00119_e -1.274\n",
      "L-Serine EX_cpd00054_e -9.678\n",
      "L-Aspartate EX_cpd00041_e -276.455\n",
      "Nicotinamide EX_cpd00133_e -0.234\n",
      "Inosine EX_cpd00246_e -166.134\n",
      "Guanosine EX_cpd00311_e -2.788\n",
      "Deoxyguanosine EX_cpd00277_e -360.036\n",
      "Deoxycytidine EX_cpd00654_e -1.476\n",
      "Deoxyuridine EX_cpd00412_e -229.417\n",
      "D-Glutamate EX_cpd00186_e -28.113\n"
     ]
    }
   ],
   "source": [
    "# Input\n",
    "for (columnName, columnData) in iCdR703_solo.flux_samples.iteritems():\n",
    "    if 'EX_' in columnName:\n",
    "        colMedian = round(numpy.median(columnData.values),3)\n",
    "        if colMedian < -1e-6:\n",
    "            substrate = list(iCdR703.reactions.get_by_id(columnName).reactants)[0].name\n",
    "            print(substrate, columnName, colMedian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H2O EX_cpd00001_e 204.508\n",
      "H+ EX_cpd00067_e 266.805\n",
      "N-Acetyl-D-glucosamine EX_cpd00122_e 433.298\n",
      "Hypoxanthine EX_cpd00226_e 232.652\n",
      "Uracil EX_cpd00092_e 220.173\n",
      "CO2 EX_cpd00011_e 397.416\n",
      "5-Aminopentanoate EX_cpd00339_e 57.182\n",
      "Acetate EX_cpd00029_e 652.122\n",
      "Formate EX_cpd00047_e 307.442\n",
      "Deoxyadenosine EX_cpd00438_e 82.406\n",
      "Guanine EX_cpd00207_e 444.606\n"
     ]
    }
   ],
   "source": [
    "# Output\n",
    "for (columnName, columnData) in iCdR703_solo.flux_samples.iteritems():\n",
    "    if 'EX_' in columnName:\n",
    "        colMedian = round(numpy.median(columnData.values),3)\n",
    "        if colMedian > 1e-6:\n",
    "            substrate = list(iCdR703.reactions.get_by_id(columnName).reactants)[0].name\n",
    "            print(substrate, columnName, colMedian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D-Fructose EX_cpd00082_e -0.016\n",
      "L-Glutamate EX_cpd00023_e -12.706\n",
      "Na+ EX_cpd00971_e -0.065\n",
      "L-Arginine EX_cpd00051_e -4.175\n",
      "Ornithine EX_cpd00064_e -341.616\n",
      "Fe2+ EX_cpd10515_e -0.065\n",
      "Phosphate EX_cpd00009_e -37.338\n",
      "L-Methionine EX_cpd00060_e -3.522\n",
      "H+ EX_cpd00067_e -5.152\n",
      "L-Alanine EX_cpd00035_e -591.005\n",
      "Riboflavin EX_cpd00220_e -0.072\n",
      "D-Alanine EX_cpd00117_e -68.488\n",
      "Mg EX_cpd00254_e -0.065\n",
      "K+ EX_cpd00205_e -0.065\n",
      "Thiamin EX_cpd00305_e -0.065\n",
      "L-Tryptophan EX_cpd00065_e -0.835\n",
      "Co2+ EX_cpd00149_e -0.065\n",
      "Cl- EX_cpd00099_e -0.065\n",
      "Biotin EX_cpd00104_e -0.065\n",
      "Ca2+ EX_cpd00063_e -0.065\n",
      "L-Valine EX_cpd00156_e -8.558\n",
      "L-Lysine EX_cpd00039_e -31.491\n",
      "Folate EX_cpd00393_e -0.065\n",
      "L-Isoleucine EX_cpd00322_e -466.785\n",
      "Mn2+ EX_cpd00030_e -0.065\n",
      "L-Cysteine EX_cpd00084_e -1.689\n",
      "Pantothenate EX_cpd00644_e -0.137\n",
      "Pyridoxol EX_cpd00263_e -0.065\n",
      "L-Leucine EX_cpd00107_e -11.871\n",
      "L-Tyrosine EX_cpd00069_e -5.427\n",
      "L-Phenylalanine EX_cpd00066_e -5.714\n",
      "L-Histidine EX_cpd00119_e -1.774\n",
      "L-Asparagine EX_cpd00132_e -8.401\n",
      "L-Serine EX_cpd00054_e -116.688\n",
      "L-Threonine EX_cpd00161_e -6.47\n",
      "L-Aspartate EX_cpd00041_e -28.563\n",
      "Nicotinamide EX_cpd00133_e -0.326\n",
      "Guanosine EX_cpd00311_e -2.432\n",
      "Cytidine EX_cpd00367_e -483.028\n",
      "Deoxyguanosine EX_cpd00277_e -1.004\n",
      "Adenosine EX_cpd00182_e -4.157\n",
      "Deoxycytidine EX_cpd00654_e -2.055\n",
      "Deoxyuridine EX_cpd00412_e -750.65\n",
      "D-Glutamate EX_cpd00186_e -39.136\n"
     ]
    }
   ],
   "source": [
    "# Input\n",
    "for (columnName, columnData) in iCdR703_coculture.flux_samples.iteritems():\n",
    "    if 'EX_' in columnName:\n",
    "        colMedian = round(numpy.median(columnData.values),3)\n",
    "        if colMedian < -1e-6:\n",
    "            substrate = list(iCdR703.reactions.get_by_id(columnName).reactants)[0].name\n",
    "            print(substrate, columnName, colMedian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N-Acetyl-D-glucosamine EX_cpd00122_e 589.371\n",
      "Cytosine EX_cpd00307_e 483.028\n",
      "Hypoxanthine EX_cpd00226_e 12.234\n",
      "Uracil EX_cpd00092_e 738.355\n",
      "CO2 EX_cpd00011_e 619.981\n",
      "2-Methylbutyrate EX_cpd19585_e 452.114\n",
      "5-Aminopentanoate EX_cpd00339_e 262.868\n",
      "Acetate EX_cpd00029_e 656.005\n",
      "Ammonia EX_cpd00013_e 1000.0\n",
      "Formate EX_cpd00047_e 315.807\n"
     ]
    }
   ],
   "source": [
    "# Output\n",
    "for (columnName, columnData) in iCdR703_coculture.flux_samples.iteritems():\n",
    "    if 'EX_' in columnName:\n",
    "        colMedian = round(numpy.median(columnData.values),3)\n",
    "        if colMedian > 1e-6:\n",
    "            substrate = list(iCdR703.reactions.get_by_id(columnName).reactants)[0].name\n",
    "            print(substrate, columnName, colMedian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import symengine\n",
    "import pandas\n",
    "from scipy import stats\n",
    "from random import sample\n",
    "import cobra.flux_analysis\n",
    "import numpy\n",
    "import warnings\n",
    "\n",
    "# Contribution to biomass\n",
    "def find_contribution(genre):\n",
    "    warnings.filterwarnings('ignore')\n",
    "\n",
    "    # Test each exchange metabolite's contribute to the objective\n",
    "    biomass = genre.slim_optimize()\n",
    "    consume = {}\n",
    "    produce = {}\n",
    "    with genre as g:\n",
    "        for rxn in g.exchanges:\n",
    "            if rxn.lower_bound < 0.0:\n",
    "                substrate = list(rxn.metabolites)[0].name\n",
    "                old_bound = rxn.lower_bound\n",
    "                rxn.lower_bound = old_bound * 0.01\n",
    "                new_biomass = g.slim_optimize()\n",
    "                if str(new_biomass) == 'nan':\n",
    "                    rxn.lower_bound = old_bound\n",
    "                    continue\n",
    "                else:\n",
    "                    effect = (g.slim_optimize() / biomass) * 100.0\n",
    "                    effect = round(100.0 - effect, 3)\n",
    "                    if effect > 1.0: consume[substrate] = effect\n",
    "                    rxn.lower_bound = old_bound\n",
    "                    \n",
    "            elif rxn.upper_bound > 0.0:\n",
    "                substrate = list(rxn.metabolites)[0].name\n",
    "                old_bound = rxn.upper_bound\n",
    "                rxn.upper_bound = old_bound * 0.01\n",
    "                new_biomass = g.slim_optimize()\n",
    "                if str(new_biomass) == 'nan':\n",
    "                    rxn.upper_bound = old_bound\n",
    "                    continue\n",
    "                else:\n",
    "                    effect = (g.slim_optimize() / biomass) * 100.0\n",
    "                    effect = round(100.0 - effect, 3)\n",
    "                    if effect > 1.0: produce[substrate] = effect\n",
    "                    rxn.upper_bound = old_bound\n",
    "    \n",
    "    warnings.filterwarnings('default')\n",
    "    return {'consumes' : consume, 'produces' : produce}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'consumes': {'Ornithine': 10.927,\n",
       "  'H+': 5.383,\n",
       "  'Inosine': 19.485,\n",
       "  'Deoxyadenosine': 4.465},\n",
       " 'produces': {'Acetate': 30.54}}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_contribution(iCdR703_solo.model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'consumes': {'Ornithine': 27.607,\n",
       "  'H+': 7.097,\n",
       "  'L-Serine': 21.015,\n",
       "  'Deoxyuridine': 27.627},\n",
       " 'produces': {'2-Methylbutyrate': 33.498, '5-Aminopentanoate': 2.275}}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_contribution(iCdR703_coculture.model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
