{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/python\n",
    "\n",
    "import os\n",
    "import copy\n",
    "import time\n",
    "import numpy\n",
    "import cobra\n",
    "import bisect\n",
    "import pandas\n",
    "import warnings\n",
    "import symengine\n",
    "import itertools\n",
    "from cobra.util import solver\n",
    "from scipy.stats import spearmanr\n",
    "from numpy.random import permutation\n",
    "from cobra.manipulation.delete import remove_genes\n",
    "from cobra.flux_analysis import flux_variability_analysis\n",
    "\n",
    "\n",
    "# Create a class to house riptide output data\n",
    "class riptideClass:\n",
    "    def __init__(self):\n",
    "        self.model = 'NULL'\n",
    "        self.transcriptome = 'NULL'\n",
    "        self.minimization_coefficients = 'NULL'\n",
    "        self.maximization_coefficients = 'NULL'\n",
    "        self.pruned = 'NULL'\n",
    "        self.flux_samples = 'NULL'\n",
    "        self.flux_variability = 'NULL'\n",
    "        self.fraction_of_optimum = 'NULL'\n",
    "        self.metabolic_tasks = 'NULL'\n",
    "        self.concordance = 'NULL'\n",
    "        self.gpr_integration = 'NULL'\n",
    "        self.percent_of_mapping = 'NULL'\n",
    "        self.defined_coefficients = 'NULL'\n",
    "        self.included_important = 'NULL'\n",
    "        self.additional_parameters = 'NULL'\n",
    "        self.fraction_bounds = 'NULL'\n",
    "        self.fraction_step = 'NULL'\n",
    "\n",
    "\n",
    "# Save the output of RIPTiDe in a newly created directory\n",
    "def save_output(riptide_obj='NULL', path='NULL', file_type='SBML'):\n",
    "    '''Writes RIPTiDe results to files in a new directory\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "\n",
    "    REQUIRED\n",
    "    riptide_obj : RIPTiDe object\n",
    "        Class object creared by riptide.contextualize()\n",
    "\n",
    "    OPTIONAL\n",
    "    path : str\n",
    "        New directory to write output files\n",
    "    file_type : str\n",
    "        Type of output file for RIPTiDe model\n",
    "        Accepts either sbml or json\n",
    "        Default is SBML\n",
    "    '''\n",
    "\n",
    "    if riptide_obj == 'NULL':\n",
    "        raise ValueError('ERROR: Did not provide a RIPTiDe object')\n",
    "    \n",
    "    if path == 'NULL':\n",
    "        print('WARNING: Did not provide an output directory. Using default riptide_files in working directory')\n",
    "        path = 'riptide_files'\n",
    "    try:\n",
    "        os.mkdir(path)\n",
    "    except:\n",
    "        print('WARNING: Output path already exists, overwriting previous files')\n",
    "        pass\n",
    "\n",
    "    # Write model to file\n",
    "    if file_type.upper() == 'JSON':\n",
    "        outFile = path + '/model.json'\n",
    "        cobra.io.save_json_model(riptide_obj.model, outFile)\n",
    "    else:\n",
    "        outFile = path + '/model.sbml'\n",
    "        cobra.io.write_sbml_model(riptide_obj.model, outFile)\n",
    "\n",
    "    # Write flux samples and FVA to a tsv\n",
    "    if isinstance(riptide_obj.flux_samples, str) == False:\n",
    "        outFile = path + '/flux_samples.tsv'\n",
    "        riptide_obj.flux_samples.to_csv(outFile, sep='\\t')\n",
    "    outFile = path + '/flux_variability.tsv'\n",
    "    riptide_obj.flux_variability.to_csv(outFile, sep='\\t')\n",
    "\n",
    "    # Write coefficient dictionaries to tsvs\n",
    "    outFile = path + '/flux_minimization_coefficients.tsv'\n",
    "    with open(outFile, 'w') as min_coefficients:\n",
    "        min_coefficients.write('reaction\\tlinear_coefficient\\n')\n",
    "        for rxn in riptide_obj.minimization_coefficients.keys():\n",
    "            min_coefficients.write(rxn + '\\t' + str(riptide_obj.minimization_coefficients[rxn]) + '\\n')\n",
    "    outFile = path + '/flux_maximization_coefficients.tsv'\n",
    "    with open(outFile, 'w') as max_coefficients:\n",
    "        max_coefficients.write('reaction\\tlinear_coefficient\\n')\n",
    "        for rxn in riptide_obj.maximization_coefficients.keys():\n",
    "            max_coefficients.write(rxn + '\\t' + str(riptide_obj.maximization_coefficients[rxn]) + '\\n')\n",
    "\n",
    "    # Write pruned model component IDs to a tsv\n",
    "    outFile = path + '/pruned_components.tsv'\n",
    "    with open(outFile, 'w') as pruned:\n",
    "    \tline = 'Genes:\\t' + '\\t'.join(riptide_obj.pruned['genes']) + '\\n'\n",
    "    \tline += 'Reactions:\\t' + '\\t'.join(riptide_obj.pruned['reactions']) + '\\n'\n",
    "    \tline += 'Metabolites:\\t' + '\\t'.join(riptide_obj.pruned['metabolites']) + '\\n'\n",
    "    \tpruned.write(line)\n",
    "\n",
    "    # Assemble parameters and output metrics text file\n",
    "    outFile = path + '/parameters.txt'\n",
    "    with open(outFile, 'w') as parameters:\n",
    "        # Parameters\n",
    "        parameters.write('Fraction of optimum objective value: ' + str(riptide_obj.fraction_of_optimum) + '\\n')\n",
    "        parameters.write('Percent of genes in mapping found in model: ' + str(riptide_obj.percent_of_mapping) + '\\n')\n",
    "        parameters.write('Minimum flux to avoid pruning: ' + str(riptide_obj.additional_parameters['threshold']) + '\\n')\n",
    "\n",
    "        if riptide_obj.fraction_bounds == 'NULL':\n",
    "            parameters.write('Run in iterative mode: No\\n')\n",
    "        else:\n",
    "            parameters.write('Run in iterative mode: Yes\\n')\n",
    "        if riptide_obj.gpr_integration == True:\n",
    "            parameters.write('Differential weighting by GPR: Yes\\n')\n",
    "        else:\n",
    "            parameters.write('Differential weighting by GPR: No\\n')\n",
    "        if riptide_obj.additional_parameters['exch_weight'] == True:\n",
    "            parameters.write('Exchanges weighted as adjacent transporters: Yes\\n')\n",
    "        else:\n",
    "            parameters.write('Exchanges weighted as adjacent transporters: No\\n')\n",
    "        if riptide_obj.additional_parameters['set_bounds'] == True:\n",
    "            parameters.write('Set reaction bounds based on FVA results: Yes\\n')\n",
    "        else:\n",
    "            parameters.write('Set reaction bounds based on FVA results: No\\n')\n",
    "        if riptide_obj.additional_parameters['objective'] == True:\n",
    "            parameters.write('Defined cellular objective: Yes\\n')\n",
    "        else:\n",
    "            parameters.write('Defined cellular objective: No\\n')\n",
    "        if riptide_obj.additional_parameters['conservative'] == True:\n",
    "            parameters.write('Conservative pruning based on GPR: Yes\\n')\n",
    "        else:\n",
    "            parameters.write('Conservative pruning based on GPR: No\\n')\n",
    "        if riptide_obj.additional_parameters['additive'] == True:\n",
    "            parameters.write('Pooled transcript based on GPR: Yes\\n')\n",
    "        else:\n",
    "            parameters.write('Pooled transcript based on GPR: No\\n')\n",
    "        if riptide_obj.additional_parameters['open_exchanges'] == True:\n",
    "            parameters.write('Exchange reactions switched open: Yes\\n')\n",
    "        else:\n",
    "            parameters.write('Exchange reactions switched open: No\\n')\n",
    "        if riptide_obj.additional_parameters['silent'] == True:\n",
    "            parameters.write('Run in silent mode: Yes\\n')\n",
    "        else:\n",
    "            parameters.write('Run in silent mode: No\\n\\n')\n",
    "        if riptide_obj.fraction_bounds != 'NULL':\n",
    "             parameters.write('Max fit optimal fraction lower bound: ' + str(riptide_obj.fraction_bounds[0]) + '\\n')\n",
    "             parameters.write('Max fit optimal fraction upper bound: ' + str(riptide_obj.fraction_bounds[1]) + '\\n')\n",
    "             parameters.write('Max fit optimal fraction step: ' + str(riptide_obj.fraction_step) + '\\n')\n",
    "\n",
    "        # Results\n",
    "        parameters.write('Percent pruned reactions: ' + str(riptide_obj.additional_parameters['operation']['reactions']) + '%\\n')\n",
    "        parameters.write('Percent pruned metabolites: ' + str(riptide_obj.additional_parameters['operation']['metabolites']) + '%\\n')\n",
    "        if riptide_obj.additional_parameters['operation']['obj_change'] != 'fails':\n",
    "            parameters.write('Percent change to objective value: ' + str(riptide_obj.additional_parameters['operation']['obj_change']) + '%\\n')\n",
    "        if riptide_obj.concordance != 'Not performed':\n",
    "            if str(riptide_obj.concordance['r']) != 'nan':\n",
    "                r_val = round(riptide_obj.concordance['r'], 4)\n",
    "                p_val = round(riptide_obj.concordance['p'], 4)\n",
    "                parameters.write('Correlation between activity and transcriptome: R=' + str(r_val) + ', p-value=' + str(p_val) + '\\n')\n",
    "        parameters.write('RIPTiDe run time: ' + str(riptide_obj.additional_parameters['operation']['run_time']) + ' seconds\\n')\n",
    "        \n",
    "\n",
    "# Read in transcriptomic read abundances, default is tsv with no header \n",
    "def read_transcription_file(file, header = False, replicates = False, sep = '\\t', rarefy = False, level = 1e6,\n",
    "    binning = False, quant_max = 0.9, quant_min = 0.5, step = 0.125, norm = True, factor = 1e6):\n",
    "    '''Generates dictionary of transcriptomic abundances from a file\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "\n",
    "    REQUIRED\n",
    "    file : string\n",
    "        User-provided file name which contains gene IDs and associated transcription values\n",
    "\n",
    "    OPTIONAL\n",
    "    header : boolean\n",
    "        Defines if read abundance file has a header that needs to be ignored\n",
    "        Default is no header\n",
    "    replicates : boolean\n",
    "        Defines if read abundances contains replicates and medians require calculation\n",
    "        Default is no replicates\n",
    "    sep : string\n",
    "        Defines what character separates entries on each line\n",
    "        Defaults to tab (.tsv)\n",
    "    rarefy : bool\n",
    "        Rarefies rounded transcript abundances to 90% of the smallest replicate\n",
    "        Default is False\n",
    "    level : int\n",
    "        Level by which to rarefy samples\n",
    "        Default is 1000000\n",
    "    binning : boolean\n",
    "        Perform discrete binning of transcript abundances into quantiles\n",
    "        OPTIONAL, not advised\n",
    "        Default is False\n",
    "    quant_max : float\n",
    "        Largest quantile to consider\n",
    "        Default is 0.9\n",
    "    quant_min : float\n",
    "        Largest quantile to consider\n",
    "        Default is 0.5\n",
    "    step : float\n",
    "        Step size for parsing quantiles\n",
    "        Default is 0.125\n",
    "    norm : bool\n",
    "        Normalize transcript abundances using RPM calculation\n",
    "        Performed by default\n",
    "    factor : numeric\n",
    "        Denominator for read normalization calculation\n",
    "        Default is 1e6 (RPM)\n",
    "    '''\n",
    "\n",
    "    # Correct some possible user error\n",
    "    if quant_max >= 1.0 or quant_max <= 0.0: quant_max = 0.99\n",
    "    if quant_min <= 0.0 or quant_min >= 1.0: quant_min = 0.01\n",
    "    if step <= 0.0 or step >= 1.0: step = 0.125\n",
    "    if factor < 1: factor = 1e3\n",
    "    level = int(level)\n",
    "\n",
    "    # Read in file\n",
    "    abund_dict = {}\n",
    "    reps = 1\n",
    "    with open(file, 'r') as transcription:\n",
    "        if header == True: header_line = transcription.readline()\n",
    "\n",
    "        for line in transcription:\n",
    "            line = line.split(sep)\n",
    "            abundances = [float(x) for x in line[1:]]\n",
    "            abund_dict[str(line[0])] = abundances\n",
    "            if reps < len(abundances):\n",
    "                reps = len(abundances)\n",
    "\n",
    "    # Rarefy abundances\n",
    "    if rarefy == True:\n",
    "        genes = list(abund_dict.keys())\n",
    "        for x in range(0, reps):\n",
    "            curr_abund = []\n",
    "            for y in genes:\n",
    "                curr_abund.append(abund_dict[y][x])\n",
    "            curr_abund = _rarefy(curr_abund, level)\n",
    "\n",
    "            for z in range(0, len(genes)):\n",
    "                abund_dict[genes[z]][x] = curr_abund[z]\n",
    "\n",
    "    # Calculate median of replicates if needed\n",
    "    if reps >= 2 and replicates == False:\n",
    "        print('WARNING: Replicates detected, calculating median transcript abundance for each gene')\n",
    "        replicates = True  \n",
    "    if replicates == True:\n",
    "        for gene in abund_dict.keys():\n",
    "            abund_dict[gene] = float(numpy.median(abund_dict[gene]))\n",
    "    else:\n",
    "        for gene in abund_dict.keys():\n",
    "            abund_dict[gene] = abund_dict[gene][0]\n",
    "\n",
    "    # Perform reads per million normalization if specified\n",
    "    if norm == True:\n",
    "        total_transcript = float(sum(abund_dict.values()))\n",
    "        for gene in abund_dict.keys():\n",
    "            new_abund = (abund_dict[gene] / total_transcript) * float(factor) # RPM by default\n",
    "            new_abund = round(new_abund, 3)\n",
    "            abund_dict[gene] = new_abund\n",
    "\n",
    "    # If user-defined, perform abundance binning by quantile\n",
    "    if binning == True:\n",
    "        print('Performing transcript abundance binning by quantile...')\n",
    "        abund_dict = _assign_quantiles(abund_dict, quant_max, quant_min, step)\n",
    "\n",
    "    return abund_dict\n",
    "\n",
    "\n",
    "# Creates transcription abundances catagories based on quantiles - optional\n",
    "def _assign_quantiles(transcription, quant_max, quant_min, step):\n",
    "\n",
    "    if quant_max >= 1.0 or quant_min <= 0.0:\n",
    "        raise ValueError('ERROR: Quantile range values must be between 1.0 and 0.0! Please correct')\n",
    "    elif step >= 1.0 or step <= 0.0:\n",
    "        raise ValueError('ERROR: Quantile step must be between 1.0 and 0.0! Please correct')\n",
    "\n",
    "    abundances = list(transcription.values())\n",
    "    abundances.sort()\n",
    "\n",
    "    thresholds = [max(abundances)]\n",
    "    while quant_max >= quant_min:\n",
    "        if quant_max <= 0.0: quant_max = 0.01\n",
    "        thresholds.append(numpy.quantile(abundances, quant_max))\n",
    "        quant_max -= step\n",
    "    thresholds.sort()\n",
    "\n",
    "    # Identifies which quantile each transcript abundance is in and assigns the largest value in that range instead\n",
    "    for gene in transcription.keys():\n",
    "        abund = transcription[gene]\n",
    "\n",
    "        if abund in thresholds:\n",
    "            index = thresholds.index(abund)\n",
    "            transcription[gene] = thresholds[index]\n",
    "        else:\n",
    "            index = bisect.bisect_right(thresholds, abund)\n",
    "            transcription[gene] = thresholds[index]\n",
    "\n",
    "    return transcription\n",
    "\n",
    "\n",
    "# Rarefies a list of numbers\n",
    "def _rarefy(abunds, n):\n",
    "\n",
    "    counts = numpy.array([round(x) for x in list(abunds)])\n",
    "    if counts.sum() <= n: \n",
    "        return list(abunds)\n",
    "\n",
    "    nz = counts.nonzero()[0]\n",
    "    unpacked = numpy.concatenate([numpy.repeat(numpy.array(i,), counts[i]) for i in nz])\n",
    "    permuted = permutation(unpacked)[:n]\n",
    "    result = numpy.zeros(len(counts))\n",
    "    for p in permuted:\n",
    "        result[p] += 1\n",
    "\n",
    "    return list(result)\n",
    "\n",
    "\n",
    "# Iteratively run RIPTiDe over a range of objective minimum fractions\n",
    "def maxfit_contextualize(model, transcriptome = 'none', frac_min = 0.35, frac_max = 0.95, frac_step = 0.05, first_max = False,\n",
    "\tsamples = 500, exch_weight = False, processes = None, minimum = None, conservative = False, objective = True, additive = False, \n",
    "\timportant = [], set_bounds = True, silent = False, tasks = [], exclude = [], gpr = False, threshold = 1e-6, defined = False, open_exchanges = False):\n",
    "\n",
    "    '''Iterative RIPTiDe for a range of minimum objective fluxes, returns model with best fit to transcriptome\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "\n",
    "    REQUIRED\n",
    "    model : cobra.Model\n",
    "        The model to be contextualized\n",
    "    transcriptome : dictionary\n",
    "        Dictionary of transcript abundances, output of read_transcription_file()\n",
    "    \n",
    "    OPTIONAL\n",
    "    frac_min : float\n",
    "        Lower bound for range of minimal fractions to test\n",
    "        Default is 0.65\n",
    "    frac_max : float\n",
    "        Upper bound for range of minimal fractions to test\n",
    "        Default is 0.85\n",
    "    frac_step : float\n",
    "        Increment to parse input minimal fraction range\n",
    "        Default is 0.02\n",
    "    first_max : bool\n",
    "    \tExits early if next subsequent iteration has a worse correlation\n",
    "    \tDefault is False\n",
    "\n",
    "    ADDITIONAL\n",
    "    All other optional parameters for riptide.contextualize()\n",
    "    '''\n",
    "\n",
    "    iter_start = time.time()\n",
    "\n",
    "    if samples <= 100:\n",
    "        raise ValueError('ERROR: Iterative RIPTiDe requires >100 flux sampling depth')\n",
    "    if transcriptome == 'none':\n",
    "        raise ValueError('ERROR: Iterative RIPTiDe requires an input transcriptome')\n",
    "\n",
    "    if frac_min < 0. or frac_min > 1.:\n",
    "        print('WARNING: Improper minimum fraction provided, setting to default value')\n",
    "        frac_min = 0.65\n",
    "    elif frac_min > frac_max:\n",
    "        print('WARNING: Improper minimum fraction provided, setting to default value')\n",
    "        frac_min = 0.65\n",
    "\n",
    "    if frac_max < 0. or frac_max > 1.:\n",
    "        print('WARNING: Improper maximum fraction provided, setting to default value')\n",
    "        frac_max = 0.85\n",
    "    elif frac_max < frac_min:\n",
    "        print('WARNING: Improper maximum fraction provided, setting to default value')\n",
    "        frac_max = 0.85\n",
    "\n",
    "    if frac_step < 0. or frac_min > 1.:\n",
    "        print('WARNING: Improper fraction step provided, setting to default value')\n",
    "        frac_max = 0.02\n",
    "    \n",
    "    frac_range = [round(x, 3) for x in list(numpy.arange(frac_min, frac_max, frac_step))]\n",
    "    frac_range.append(round(frac_max, 3))\n",
    "    if len(frac_range) == 1:\n",
    "        print('WARNING: Only a single fraction is possible in the input bounds and fraction')\n",
    "\n",
    "    frac_min = round(frac_min, 3)\n",
    "    frac_max = round(frac_max, 3)\n",
    "    frac_step = round(frac_step, 3)\n",
    "\n",
    "    if silent == False:\n",
    "        print('\\nRunning max fit RIPTiDe for objective fraction range:', frac_min, 'to', frac_max, 'with intervals of', frac_step, '\\n')\n",
    "\n",
    "    top_rho = 0.\n",
    "    iters = 0\n",
    "    if silent == False:\n",
    "        print('Testing minimum objective fractions...')\n",
    "    for frac in frac_range:\n",
    "        iters += 1\n",
    "\n",
    "        iter_riptide = contextualize(model, transcriptome, fraction=frac, silent=True, samples=samples, exch_weight=exch_weight, \n",
    "            processes=processes, minimum=minimum, conservative=conservative, \n",
    "            objective=objective, additive=additive, important=important, set_bounds=set_bounds, tasks=tasks, exclude=exclude, \n",
    "            gpr=gpr, threshold=threshold, defined=defined, open_exchanges=open_exchanges)\n",
    "\n",
    "        curr_rho = iter_riptide.concordance['r']\n",
    "        curr_p = iter_riptide.concordance['p']\n",
    "        if silent == False:\n",
    "            print('Fraction =', frac, '| Rho =', round(curr_rho,4), '; p =', round(curr_p,4))\n",
    "        \n",
    "        if curr_rho >= top_rho:\n",
    "            top_fit = copy.deepcopy(iter_riptide)\n",
    "            top_rho = curr_rho\n",
    "        elif first_max == True:\n",
    "            if curr_rho < top_rho:\n",
    "                if silent == False:\n",
    "                    print('High correlation found, exiting search...')\n",
    "                break\n",
    "    \n",
    "    if silent == False:\n",
    "        print('Testing local objective fractions to ' + str(top_fit.fraction_of_optimum) + '...')\n",
    "    new_step = round(frac_step / 2.0, 3)\n",
    "    frac = round(top_fit.fraction_of_optimum, 3)\n",
    "\n",
    "    lower_frac = round(frac - new_step, 3)\n",
    "    iter_riptide = contextualize(model, transcriptome, fraction=lower_frac, silent=True, samples=samples, exch_weight=exch_weight, \n",
    "                    processes=processes, minimum=minimum, conservative=conservative, \n",
    "                    objective=objective, additive=additive, important=important, set_bounds=set_bounds, tasks=tasks, exclude=exclude,\n",
    "                    gpr=gpr, threshold=threshold, defined=defined, open_exchanges=open_exchanges)\n",
    "    curr_rho = iter_riptide.concordance['r']\n",
    "    curr_p = iter_riptide.concordance['p']\n",
    "    if silent == False:\n",
    "        print('Fraction =', lower_frac, '| Rho =', round(curr_rho,4), '; p =', round(curr_p,4))\n",
    "    if curr_rho >= top_rho:\n",
    "        top_fit = copy.deepcopy(iter_riptide)\n",
    "\n",
    "    upper_frac = round(frac + new_step, 3)\n",
    "    iter_riptide = contextualize(model, transcriptome, fraction=upper_frac, silent=True, samples=samples, exch_weight=exch_weight, \n",
    "                    processes=processes, minimum=minimum, conservative=conservative, \n",
    "                    objective=objective, additive=additive, important=important, set_bounds=set_bounds, tasks=tasks, exclude=exclude,\n",
    "                    gpr=gpr, threshold=threshold, defined=defined, open_exchanges=open_exchanges)\n",
    "    curr_rho = iter_riptide.concordance['r']\n",
    "    curr_p = iter_riptide.concordance['p']\n",
    "    if silent == False:\n",
    "        print('Fraction =', upper_frac, '| Rho =', round(curr_rho,4), '; p =', round(curr_p,4))\n",
    "    if curr_rho >= top_rho:\n",
    "        top_fit = copy.deepcopy(iter_riptide)\n",
    "\n",
    "    top_fit.fraction_bounds = [frac_min, frac_max]\n",
    "    top_fit.fraction_step = frac_step\n",
    "\n",
    "    if silent == False:\n",
    "        print('\\nContext-specific metabolism fit with', top_fit.fraction_of_optimum, 'of optimal objective flux')\n",
    "\n",
    "    # Analyze changes introduced by RIPTiDe and return results\n",
    "    report_dict = _operation_report(iter_start, model, top_fit.model, top_fit.concordance, silent, mf=True)\n",
    "    top_fit.additional_parameters['operation'] = report_dict\n",
    "\n",
    "    return top_fit\n",
    "\n",
    "\n",
    "# Create context-specific model based on transcript distribution\n",
    "def contextualize(model, transcriptome = 'none', samples = 500, silent = False, exch_weight = False, processes=None,\n",
    "    fraction = 0.8, minimum = None, conservative = False, objective = True, additive = False, important = [], direct = False,\n",
    "    set_bounds = True, tasks = [], exclude = [], gpr = False, threshold = 1e-6, defined = False, open_exchanges = False):\n",
    "\n",
    "    '''Reaction Inclusion by Parsimony and Transcriptomic Distribution or RIPTiDe\n",
    "    \n",
    "    Creates a contextualized metabolic model based on parsimonious usage of reactions defined\n",
    "    by their associated transcriptomic abundances. Returns a pruned, context-specific cobra.Model \n",
    "    and associated flux analyses along with parameters used in a riptide class object\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "\n",
    "    REQUIRED\n",
    "    model : cobra.Model\n",
    "        The model to be contextualized\n",
    "\n",
    "    OPTIONAL\n",
    "    transcriptome : dictionary\n",
    "        Dictionary of transcript abundances, output of read_transcription_file()\n",
    "        With default, an artifical transcriptome is generated where all abundances equal 1.0\n",
    "    samples : int \n",
    "        Number of flux samples to collect\n",
    "        Default is 500\n",
    "    silent  : bool\n",
    "        Silences std out \n",
    "        Default is False\n",
    "    exch_weight : bool\n",
    "        Weight exchange reactions the same as adjacent transporters\n",
    "        Default is True\n",
    "    processes : int\n",
    "        The number of parallel processes to run for FVA. Optional and if not passed,\n",
    "        will be set to the number of CPUs found. Necessary to change if\n",
    "        your trying to run paralell instance of RIPTiDe on the same machine\n",
    "        Default is none\n",
    "    fraction : float\n",
    "        Minimum percent of optimal objective value during FBA steps\n",
    "        Default is 0.8\n",
    "    minimum : float\n",
    "        Minimum linear coefficient allowed during weight calculation for pFBA\n",
    "        Default is None\n",
    "    conservative : bool\n",
    "        Conservatively remove inactive reactions based on GPR rules (all member reactions must be inactive to prune)\n",
    "        Default is False\n",
    "    objective : bool\n",
    "        Sets previous objective function as a constraint with minimum flux equal to user input fraction\n",
    "        Default is True\n",
    "    additive : bool\n",
    "        Pool transcription abundances for reactions with multiple contributing gene products\n",
    "        Default is False\n",
    "    important : list\n",
    "        List of gene or reaction ID strings for which the highest weights are assigned regardless of transcription\n",
    "        Default is False\n",
    "    direct : bool\n",
    "        Assigns both minimization and maximization step coefficents directly, instead of relying on abundance distribution\n",
    "        Default is False\n",
    "    set_bounds : bool\n",
    "        Uses flux variability analysis results from constrained model to set new bounds for all reactions\n",
    "        Default is True\n",
    "    tasks : list\n",
    "        List of gene or reaction ID strings for forced inclusion in final model (metabolic tasks or essential genes)\n",
    "    exclude : list\n",
    "        List of reaction ID strings for forced exclusion from final model\n",
    "    gpr : bool\n",
    "        Determines if GPR rules will be considered during coefficient assignment\n",
    "        Default is False\n",
    "    threshold : float\n",
    "        Minimum flux a reaction must acheive in order to avoid pruning during flux sum minimization step\n",
    "        Default is 1e-6\n",
    "    defined : False or list\n",
    "        User defined range of linear coeffients, needs to be defined in a list like [1, 0.5, 0.1, 0.01, 0.001]\n",
    "        Works best paired with binned abundance catagories from riptide.read_transcription_file()\n",
    "        Default is False\n",
    "    open_exchanges : bool\n",
    "        Sets all exchange reactions bounds to (-1000., 1000)\n",
    "        Default is False\n",
    "    '''\n",
    "\n",
    "    start_time = time.time()\n",
    "    riptide_object = riptideClass()\n",
    "    riptide_object.additional_parameters = {}\n",
    "    riptide_object.additional_parameters['threshold'] = threshold\n",
    "    riptide_object.additional_parameters['silent'] = silent\n",
    "    riptide_object.additional_parameters['exch_weight'] = exch_weight\n",
    "    riptide_object.additional_parameters['conservative'] = conservative\n",
    "    riptide_object.additional_parameters['objective'] = objective\n",
    "    riptide_object.additional_parameters['additive'] = additive\n",
    "    riptide_object.additional_parameters['set_bounds'] = set_bounds\n",
    "    riptide_object.additional_parameters['open_exchanges'] = open_exchanges\n",
    "\n",
    "    # Correct some possible user error\n",
    "    samples = int(samples)\n",
    "    if samples < 1: samples = 500\n",
    "    if len(important) > 0: samples = 1\n",
    "    riptide_object.additional_parameters['samples'] = samples\n",
    "    fraction = float(fraction)\n",
    "    if fraction <= 0.0: \n",
    "        fraction = 0.01\n",
    "    elif fraction >= 1.0: \n",
    "        fraction = 0.99\n",
    "    if minimum != None:\n",
    "        if minimum <= 0.0: \n",
    "            minimum = 0.0001\n",
    "        elif minimum > 1.0: \n",
    "            minimum = 0.0001\n",
    "    riptide_object.additional_parameters['minimum'] = minimum\n",
    "    solution = model.slim_optimize()\n",
    "    if model.slim_optimize() < 1e-6 or str(model.slim_optimize()) == 'nan':\n",
    "        raise ValueError('ERROR: Provided model objective cannot carry flux! Please correct')\n",
    "    minimum_threshold = threshold\n",
    "    if isinstance(tasks, list) == False: tasks = [tasks]\n",
    "    if defined != False:\n",
    "        if isinstance(defined, list) == False:\n",
    "            defined = [defined]\n",
    "\n",
    "    # Creates artificial transcriptome to identify most parsimonious patterns of metabolism\n",
    "    if transcriptome == 'none':\n",
    "        if silent == False:\n",
    "            print('WARNING: No transcriptome provided. Analyzing most parsimonious state')\n",
    "        transcriptome = {}\n",
    "        for gene in model.genes:\n",
    "            transcriptome[gene.id] = 1.0\n",
    "\n",
    "    # Save parameters as part of the output object\n",
    "    riptide_object.fraction_of_optimum = fraction\n",
    "    riptide_object.transcriptome = transcriptome\n",
    "    riptide_object.gpr_integration = gpr\n",
    "    riptide_object.defined_coefficients = defined\n",
    "\n",
    "    # Check original model functionality\n",
    "    # Partition reactions based on transcription percentile intervals, assign corresponding reaction coefficients\n",
    "    if silent == False: print('\\nInitializing model and integrating transcriptomic data...')\n",
    "    riptide_model = copy.deepcopy(model)\n",
    "    riptide_model.id = str(riptide_model.id) + '_riptide'\n",
    "    riptide_object.metabolic_tasks = tasks\n",
    "\n",
    "    # Open exchange reactions\n",
    "    if open_exchanges == True:\n",
    "        for rxn in riptide_model.exchanges: rxn.bounds = (-1000., 1000.)\n",
    "\n",
    "    # Remove totally blocked reactions to speed up subsequent sections\n",
    "    rm_rxns = list(set(exclude).difference(set(tasks)))\n",
    "    if len(rm_rxns) > 0:\n",
    "        riptide_model = _prune_model(riptide_model, rm_rxns, conservative)\n",
    "\n",
    "    # Define linear coefficients for both steps\n",
    "    min_coefficient_dict, max_coefficient_dict, gene_hits, important_type = _assign_coefficients(transcriptome, riptide_model, minimum, gpr, defined, additive, exch_weight, important, direct)\n",
    "    riptide_object.minimization_coefficients = min_coefficient_dict\n",
    "    riptide_object.maximization_coefficients = max_coefficient_dict\n",
    "    riptide_object.percent_of_mapping = gene_hits\n",
    "    \n",
    "    # Prune now inactive network sections based on coefficients\n",
    "    if silent == False: print('Pruning zero flux subnetworks...')\n",
    "    rm_rxns = _constrain_and_analyze_model(model=riptide_model, coefficient_dict=min_coefficient_dict, fraction=fraction, sampling_depth=0, \n",
    "        objective=objective, tasks=tasks, minimum_threshold=minimum_threshold, cpus=processes)\n",
    "    riptide_model = _prune_model(riptide_model, rm_rxns, conservative)\n",
    "    riptide_object.pruned = _record_pruned_elements(model, riptide_model)\n",
    "\n",
    "    # Find optimal solution space based on transcription and final constraints\n",
    "    if silent == False: print('Analyzing context-specific flux distributions...')\n",
    "    flux_samples, fva_result, concordance = _constrain_and_analyze_model(model=riptide_model, coefficient_dict=max_coefficient_dict, fraction=fraction, \n",
    "        sampling_depth=samples, objective=objective, tasks=tasks, minimum_threshold=minimum_threshold, cpus=processes)\n",
    "    riptide_object.flux_samples = flux_samples\n",
    "    riptide_object.flux_variability = fva_result\n",
    "    riptide_object.concordance = concordance\n",
    "\n",
    "    # Assign new reaction bounds\n",
    "    if set_bounds == True: \n",
    "        for rxn in riptide_model.reactions:\n",
    "            current_fva = list(fva_result.loc[rxn.id])\n",
    "            rxn.bounds = (min(current_fva), max(current_fva))\n",
    "    riptide_object.model = riptide_model\n",
    "\n",
    "    # Report on included subset of user-defined important genes\n",
    "    if len(important) > 0:\n",
    "        if important_type == 'genes':\n",
    "            curr_list = set([x.id for x in riptide_model.genes])\n",
    "        elif important_type == 'reactions':\n",
    "            curr_list = set([y.id for y in riptide_model.reactions])\n",
    "        included_important = set(important).intersection(curr_list)\n",
    "        riptide_object.included_important = included_important\n",
    "    else:\n",
    "        riptide_object.included_important = important\n",
    "\n",
    "    # Analyze changes introduced by RIPTiDe and return results\n",
    "    report_dict = _operation_report(start_time, model, riptide_model, concordance, silent, mf=False)\n",
    "    riptide_object.additional_parameters['operation'] = report_dict\n",
    "\n",
    "    return riptide_object\n",
    "\n",
    "\n",
    "# Use FVA on constrained model to create context-specific model\n",
    "def contextualize_variability(model, transcriptome, fraction=0.8, direct=False, cpus=None):\n",
    "    warnings.filterwarnings('ignore') # Handle UserWarning\n",
    "    \n",
    "    riptide_model = copy.deepcopy(model)\n",
    "    constrained_model = copy.deepcopy(model)\n",
    "\n",
    "    # Defined coefficients\n",
    "    min_coefficient_dict, max_coefficient_dict, gene_hits, important_type = _assign_coefficients(transcriptome, riptide_model, minimum=None, gpr=False, defined_coefficients=False, additive=False, exch_weight=False, important=[], direct=direct)\n",
    "\n",
    "    # Set previous objective as a constraint, allow deviation\n",
    "    prev_obj_val = constrained_model.slim_optimize()\n",
    "    prev_obj_constraint = constrained_model.problem.Constraint(constrained_model.objective.expression, lb=prev_obj_val*fraction, ub=prev_obj_val)\n",
    "    constrained_model.add_cons_vars([prev_obj_constraint])\n",
    "    constrained_model.solver.update()\n",
    "\n",
    "    # Apply weigths to new expression\n",
    "    pfba_expr = symengine.RealDouble(0)\n",
    "    for rxn in constrained_model.reactions:\n",
    "        try:\n",
    "            pfba_expr += min_coefficient_dict[rxn.id] * rxn.forward_variable\n",
    "            pfba_expr += min_coefficient_dict[rxn.id] * rxn.reverse_variable\n",
    "        except KeyError:\n",
    "            continue\n",
    "\n",
    "    # Include metabolic task constraints\n",
    "    #if len(tasks) >= 1:\n",
    "    #    constrained_model = _integrate_tasks(constrained_model, tasks)\n",
    "    \n",
    "    # Integrate pFBA constraint\n",
    "    constrained_model.objective = constrained_model.problem.Objective(pfba_expr, direction='min', sloppy=True)\n",
    "    constrained_model.solver.update()\n",
    "    flux_sum_obj_val = constrained_model.slim_optimize()\n",
    "    flux_sum_constraint = constrained_model.problem.Constraint(pfba_expr, lb=flux_sum_obj_val*fraction, ub=flux_sum_obj_val)\n",
    "    constrained_model.add_cons_vars([flux_sum_constraint])\n",
    "    constrained_model.solver.update()\n",
    "\n",
    "    # Find and set new bounds\n",
    "    fva = flux_variability_analysis(constrained_model, fraction_of_optimum=fraction, processes=cpus)\n",
    "    prune_rxns = []\n",
    "    \n",
    "    for rxn in riptide_model.reactions:\n",
    "        current_fva = list(fva.loc[rxn.id])\n",
    "        rxn.bounds = (min(current_fva), max(current_fva))\n",
    "        if abs(min(current_fva)) + abs(max(current_fva)) <= 1e-6:\n",
    "            prune_rxns.append(rxn.id)\n",
    "\n",
    "    # Prune inactive reactions\n",
    "    prune_rxns = set(prune_rxns)\n",
    "    for rxn in prune_rxns: riptide_model.reactions.get_by_id(rxn).remove_from_model(remove_orphans=True)\n",
    "    warnings.filterwarnings('default')\n",
    "\n",
    "    return riptide_model\n",
    " \n",
    "\n",
    "# Converts a dictionary of transcript abundances to reaction linear coefficients\n",
    "def _assign_coefficients(raw_transcription_dict, model, minimum, gpr, defined_coefficients, additive, exch_weight, important, direct):\n",
    "    \n",
    "    # Screen transcriptomic abundances for genes that are included in model\n",
    "    rxn_transcript_dict = {}\n",
    "    total = 0.0\n",
    "    success = 0.0\n",
    "    fail = 0.0\n",
    "    for gene in model.genes:\n",
    "        total += 1.0\n",
    "        try:\n",
    "            current_abund = float(raw_transcription_dict[gene.id]) + 1.0\n",
    "            current_rxns = list(model.genes.get_by_id(gene.id).reactions)\n",
    "            success += 1.0\n",
    "            for rxn in current_rxns:\n",
    "                try:\n",
    "                    rxn_transcript_dict[rxn.id].append(current_abund)\n",
    "                except KeyError:\n",
    "                    rxn_transcript_dict[rxn.id] = [current_abund]\n",
    "        except KeyError:\n",
    "            fail += 1.0\n",
    "            continue\n",
    "    # Check if any or very few genes were found\n",
    "    if total == fail: \n",
    "        raise LookupError('ERROR: No gene IDs in transcriptome dictionary found in model.')\n",
    "    gene_hits = (float(total - fail) / total) * 100.0\n",
    "    gene_hits = str(round(gene_hits, 2)) + '%'\n",
    "    nogene_abund = numpy.median(list(set(raw_transcription_dict.values())))\n",
    "\n",
    "    # Reduce transcription to single values per reaction based on max/sum or GPR rules\n",
    "    exchanges = {}\n",
    "    all_abundances = set([nogene_abund])\n",
    "    for rxn in model.reactions:\n",
    "        # Identify extracellular exchanges\n",
    "        if set([x.compartment for x in rxn.reactants]) != set([x.compartment for x in rxn.products]):\n",
    "            for cpd in rxn.metabolites: # Transport reactions\n",
    "                for sub_rxn in cpd.reactions:\n",
    "                    if len(sub_rxn.reactants) == 0 or len(sub_rxn.products) == 0:\n",
    "                        exchanges[rxn.id] = sub_rxn.id\n",
    "        try:\n",
    "            transcript_abund = rxn_transcript_dict[rxn.id]\n",
    "            # Parse GPRs if defined by user\n",
    "            if gpr == True:\n",
    "                curr_gpr = str(rxn.gene_reaction_rule).upper()\n",
    "                if ' AND ' in curr_gpr:\n",
    "                    current_abund = float(min(transcript_abund))\n",
    "                    rxn_transcript_dict[rxn.id] = current_abund\n",
    "                    all_abundances |= set([current_abund])\n",
    "                elif ' OR ' in curr_gpr:\n",
    "                    current_abund = float(sum(transcript_abund))\n",
    "                    rxn_transcript_dict[rxn.id] = current_abund\n",
    "                    all_abundances |= set([current_abund])\n",
    "                else:\n",
    "                    current_abund = float(max(transcript_abund))\n",
    "                    rxn_transcript_dict[rxn.id] = current_abund\n",
    "                    all_abundances |= set([current_abund])\n",
    "            else:\n",
    "                if additive == True:\n",
    "                    current_abund = float(sum(transcript_abund))\n",
    "                    rxn_transcript_dict[rxn.id] = current_abund\n",
    "                    all_abundances |= set([current_abund])\n",
    "                else:\n",
    "                    current_abund = float(max(transcript_abund))\n",
    "                    rxn_transcript_dict[rxn.id] = current_abund\n",
    "                    all_abundances |= set([current_abund])\n",
    "        # Coefficient if no gene is associated\n",
    "        except KeyError:\n",
    "            rxn_transcript_dict[rxn.id] = nogene_abund\n",
    "\n",
    "    # Calculate coefficients\n",
    "    all_abundances = list(all_abundances)\n",
    "    denom = max(all_abundances)\n",
    "    all_abundances.sort()\n",
    "    max_coefficients = [x / denom for x in all_abundances]\n",
    "    if direct == True:\n",
    "        fctr = 1.0 + min(max_coefficients)\n",
    "        min_coefficients = [fctr - (x / denom) for x in all_abundances]\n",
    "    else:\n",
    "        min_coefficients = max_coefficients[::-1]\n",
    "    coefficient_dict = {}\n",
    "    for x in range(0, len(all_abundances)):\n",
    "        coefficient_dict[all_abundances[x]] = [min_coefficients[x], max_coefficients[x]]\n",
    "\n",
    "    # Assign coefficients to reactions\n",
    "    rxn_min_coefficient_dict = {}\n",
    "    rxn_max_coefficient_dict = {}\n",
    "    for rxn in rxn_transcript_dict.keys():\n",
    "        rxn_min_coefficient_dict[rxn] = coefficient_dict[rxn_transcript_dict[rxn]][0]\n",
    "        rxn_max_coefficient_dict[rxn] = coefficient_dict[rxn_transcript_dict[rxn]][1]\n",
    "\n",
    "    # Set adjacent exchange reactions to the same coefficients\n",
    "    if exch_weight == True:\n",
    "        for rxn in exchanges.keys():\n",
    "            rxn_max_coefficient_dict[exchanges[rxn]] = rxn_max_coefficient_dict[rxn]\n",
    "            rxn_min_coefficient_dict[exchanges[rxn]] = rxn_min_coefficient_dict[rxn]\n",
    "\n",
    "    # If user has defined important genes/reactions, integrate new weights here\n",
    "    important_type = 'NULL'\n",
    "    if len(important) > 0: \n",
    "        rxn_min_coefficient_dict, important_type = _integrate_important(model, important, rxn_min_coefficient_dict)\n",
    "\n",
    "    return rxn_min_coefficient_dict, rxn_max_coefficient_dict, gene_hits, important_type\n",
    "\n",
    "\n",
    "# Assemble a corrected list of metabolic tasks based on user\n",
    "def _integrate_tasks(model, tasks):\n",
    "    # Check that each task is in the model\n",
    "    screened_tasks = []\n",
    "    for x in set(tasks):\n",
    "        # Genes\n",
    "        try:\n",
    "            rxns = list(model.genes.get_by_id(x).reactions)\n",
    "            rxns = [y.id for y in rxns]\n",
    "            screened_tasks += rxns\n",
    "        except:\n",
    "            pass\n",
    "        # Reactions\n",
    "        try:\n",
    "            rxn = model.reactions.get_by_id(x)\n",
    "            screened_tasks.append(rxn.id)\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "    # Check if any reactions were found in the model that correspond with supplied IDs\n",
    "    if len(screened_tasks) == 0:\n",
    "        print('WARNING: No reactions found associated with provided task IDs')\n",
    "\n",
    "    # Iteratively set each as the objective and find new bounds\n",
    "    for rxn in screened_tasks:\n",
    "        model.objective = rxn\n",
    "        task_obj_val = model.slim_optimize()\n",
    "        # Check sign of objective value, just in case\n",
    "        if task_obj_val > 0.0:\n",
    "            task_constraint = model.problem.Constraint(model.objective.expression, lb=task_obj_val*0.01, ub=task_obj_val)\n",
    "            model.add_cons_vars(task_constraint)\n",
    "            model.solver.update()\n",
    "        elif task_obj_val < 0.0:\n",
    "            task_constraint = model.problem.Constraint(model.objective.expression, lb=task_obj_val, ub=task_obj_val*0.01)\n",
    "            model.add_cons_vars(task_constraint)\n",
    "            model.solver.update()\n",
    "    \n",
    "    return model\n",
    "\n",
    "\n",
    "# Assign heaviest weight during pruning to user-defined important genes\n",
    "def _integrate_important(model, important, coefficient_dict):\n",
    "\n",
    "    #weight = numpy.quantile(list(coefficient_dict.values()), 0.25)\n",
    "    weight = min(list(coefficient_dict.values()))\n",
    "\n",
    "    # Check if reactions or genes, get reaction IDs\n",
    "    rxn_ids = []\n",
    "    for x in important:\n",
    "        # Genes\n",
    "        try:\n",
    "            rxns = list(model.genes.get_by_id(x).reactions)\n",
    "            rxn_ids += [y.id for y in rxns]\n",
    "            important_type = 'genes'\n",
    "        except:\n",
    "            pass\n",
    "        # Reactions\n",
    "        try:\n",
    "            rxn = model.reactions.get_by_id(x)\n",
    "            rxn_ids.append(rxn.id)\n",
    "            important_type = 'reactions'\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "    # Correct weight in coefficient dictionary\n",
    "    for rxn in set(rxn_ids): coefficient_dict[rxn] = weight\n",
    "\n",
    "    return coefficient_dict, important_type\n",
    "\n",
    "\n",
    "# Determine those reactions that carry flux in a pFBA objective set to a threshold of maximum\n",
    "def _constrain_and_analyze_model(model, coefficient_dict, fraction, sampling_depth, objective, tasks, minimum_threshold, cpus):\n",
    "    \n",
    "    constrained_model = copy.deepcopy(model)\n",
    "\n",
    "    # Set previous objective as a constraint, allow deviation\n",
    "    if objective == True:\n",
    "        prev_obj_val = constrained_model.slim_optimize()\n",
    "        prev_obj_constraint = constrained_model.problem.Constraint(constrained_model.objective.expression, lb=prev_obj_val*fraction, ub=prev_obj_val)\n",
    "        constrained_model.add_cons_vars([prev_obj_constraint])\n",
    "        constrained_model.solver.update()\n",
    "\n",
    "    # Apply weigths to new expression\n",
    "    pfba_expr = symengine.RealDouble(0)\n",
    "    for rxn in constrained_model.reactions:\n",
    "        try:\n",
    "            pfba_expr += coefficient_dict[rxn.id] * rxn.forward_variable\n",
    "            pfba_expr += coefficient_dict[rxn.id] * rxn.reverse_variable\n",
    "        except KeyError:\n",
    "            continue\n",
    "\n",
    "    if sampling_depth == 0:\n",
    "        # Include metabolic task constraints\n",
    "        if len(tasks) >= 1:\n",
    "            constrained_model = _integrate_tasks(constrained_model, tasks)\n",
    "\n",
    "        # Determine reactions that do not carry any flux in highly constrained solution space\n",
    "        constrained_model.objective = constrained_model.problem.Objective(pfba_expr, direction='min', sloppy=True)\n",
    "        constrained_model.solver.update()\n",
    "        solution = constrained_model.optimize()\n",
    "        inactive_rxns = set([rxn.id for rxn in constrained_model.reactions if abs(solution.fluxes[rxn.id]) <= minimum_threshold])\n",
    "            \n",
    "        return inactive_rxns\n",
    "        \n",
    "    else:\n",
    "        # Explore solution space of constrained model with flux sampling, allow deviation\n",
    "        constrained_model.objective = constrained_model.problem.Objective(pfba_expr, direction='max', sloppy=True)\n",
    "        constrained_model.solver.update()\n",
    "        flux_sum_obj_val = constrained_model.slim_optimize()\n",
    "        flux_sum_constraint = constrained_model.problem.Constraint(pfba_expr, lb=flux_sum_obj_val*fraction, ub=flux_sum_obj_val)\n",
    "        constrained_model.add_cons_vars([flux_sum_constraint])\n",
    "        constrained_model.solver.update()\n",
    "            \n",
    "        # Analyze flux ranges and calculate concordance\n",
    "        if sampling_depth != 1:\n",
    "            warnings.filterwarnings('ignore') # Handle possible infeasible warnings\n",
    "            flux_samples = _gapsplit(constrained_model, depth=sampling_depth, cpus=cpus)\n",
    "            warnings.filterwarnings('default')\n",
    "            concordance = _calc_concordance(flux_samples, coefficient_dict)\n",
    "        else:\n",
    "            flux_samples = 'Not performed'\n",
    "            concordance = 'Not performed'\n",
    "            \n",
    "        fva = flux_variability_analysis(constrained_model, fraction_of_optimum=fraction, processes=cpus)\n",
    "\n",
    "        return flux_samples, fva, concordance\n",
    "\n",
    "\n",
    "# Find level of concordance between contextualized flux and assigned coefficients\n",
    "def _calc_concordance(flux_samples, coefficient_dict):\n",
    "    warnings.filterwarnings('ignore') # Handle RuntimeWarning\n",
    "\n",
    "    flux_medians = []\n",
    "    coefficients = []\n",
    "\n",
    "    for rxn in coefficient_dict.keys():\n",
    "        try:\n",
    "            flux_medians.append(abs(numpy.median(list(flux_samples[rxn]))))\n",
    "            coefficients.append(coefficient_dict[rxn])\n",
    "        except:\n",
    "            continue\n",
    "        \n",
    "    r_val, p_val = spearmanr(coefficients, flux_medians)\n",
    "    concordance_dict = {'r':r_val, 'p':p_val}\n",
    "    \n",
    "    warnings.filterwarnings('default')\n",
    "    return concordance_dict\n",
    "\n",
    "\n",
    "# Prune model based on blocked reactions from minimization as well as user-defined reactions\n",
    "def _prune_model(new_model, rm_rxns, conserve):\n",
    "    \n",
    "    # Parse elements highlighted for pruning based on GPRs\n",
    "    if conserve == True:\n",
    "        final_rm_rxns = []\n",
    "        for rxn in rm_rxns:\n",
    "            test = 'pass'\n",
    "            current_genes = list(new_model.reactions.get_by_id(rxn).genes)\n",
    "            for gene in current_genes:\n",
    "                for rxn_sub in gene.reactions:\n",
    "                    if rxn_sub.id not in rm_rxns:\n",
    "                        test = 'fail'\n",
    "                    else:\n",
    "                        pass\n",
    "            \n",
    "            if test == 'pass': final_rm_rxns.append(rxn)\n",
    "    else:\n",
    "        final_rm_rxns = rm_rxns\n",
    "    \n",
    "    # Screen for duplicates\n",
    "    final_rm_rxns = list(set(final_rm_rxns))\n",
    "    \n",
    "    # Prune inactive reactions\n",
    "    warnings.filterwarnings('ignore') # Handle UserWarning\n",
    "    for rxn in final_rm_rxns: new_model.reactions.get_by_id(rxn).remove_from_model(remove_orphans=True)\n",
    "    warnings.filterwarnings('default')\n",
    "\n",
    "    # Prune orphaned nodes\n",
    "    new_model = _complete_orphan_prune(new_model)\n",
    "\n",
    "    return new_model\n",
    "\n",
    "\n",
    "# Record IDs of all pruned elements\n",
    "def _record_pruned_elements(old_model, new_model):\n",
    "    prunedDict = {}\n",
    "\n",
    "    # Genes\n",
    "    old_genes = set([x.id for x in old_model.genes])\n",
    "    new_genes = set([y.id for y in new_model.genes])\n",
    "    prunedDict['genes'] = old_genes.difference(new_genes)\n",
    "\n",
    "    # Reactions\n",
    "    old_rxns = set([x.id for x in old_model.reactions])\n",
    "    new_rxns = set([y.id for y in new_model.reactions])\n",
    "    prunedDict['reactions'] = old_rxns.difference(new_rxns)\n",
    "\n",
    "    # Metabolites\n",
    "    old_cpds = set([x.id for x in old_model.metabolites])\n",
    "    new_cpds = set([y.id for y in new_model.metabolites])\n",
    "    prunedDict['metabolites'] = old_cpds.difference(new_cpds)\n",
    "\n",
    "    return prunedDict\n",
    "\n",
    "\n",
    "# Thoroughly remove orphan reactions and metabolites\n",
    "def _complete_orphan_prune(model):\n",
    "\n",
    "    removed = 1\n",
    "    while removed == 1:\n",
    "        removed = 0\n",
    "\n",
    "        # Metabolites\n",
    "        for cpd in model.metabolites:\n",
    "            if len(cpd.reactions) == 0:\n",
    "                cpd.remove_from_model(); removed = 1\n",
    "\n",
    "        # Reactions\n",
    "        for rxn in model.reactions:\n",
    "            if len(rxn.metabolites) == 0: \n",
    "                rxn.remove_from_model(); removed = 1\n",
    "\n",
    "    return model\n",
    "\n",
    "\n",
    "# Reports how long RIPTiDe took to run\n",
    "def _operation_report(start_time, model, riptide, concordance, silent, mf):\n",
    "    report_dict = {}\n",
    "\n",
    "    # Pruning\n",
    "    perc_removal = 100.0 - ((float(len(riptide.reactions)) / float(len(model.reactions))) * 100.0)\n",
    "    perc_removal = round(perc_removal, 2)\n",
    "    if silent == False:\n",
    "        print('\\nReactions pruned to ' + str(len(riptide.reactions)) + ' from ' + str(len(model.reactions)) + ' (' + str(perc_removal) + '% change)')\n",
    "    report_dict['reactions'] = perc_removal\n",
    "    perc_removal = 100.0 - ((float(len(riptide.metabolites)) / float(len(model.metabolites))) * 100.0)\n",
    "    perc_removal = round(perc_removal, 2)\n",
    "    if silent == False:\n",
    "        print('Metabolites pruned to ' + str(len(riptide.metabolites)) + ' from ' + str(len(model.metabolites)) + ' (' + str(perc_removal) + '% change)')\n",
    "    report_dict['metabolites'] = perc_removal\n",
    "\n",
    "    # Flux through objective\n",
    "    model_check = 'works'\n",
    "    # Check that prune model can still achieve flux through the objective (just in case)\n",
    "    try:\n",
    "        if riptide.slim_optimize() < 1e-6 or str(riptide.slim_optimize()) == 'nan':\n",
    "            if silent == False:\n",
    "                print('WARNING: Contextualized model objective can no longer carry flux')\n",
    "            model_check = 'broken'\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "    if model_check == 'works':\n",
    "        new_ov = round(riptide.slim_optimize(), 2)\n",
    "        old_ov = round(model.slim_optimize(), 2)\n",
    "        perc_shift = 100.0 - ((float(new_ov) / float(old_ov)) * 100.0)\n",
    "        report_dict['obj_change'] = round(perc_shift, 2)\n",
    "        if perc_shift == 0.0:\n",
    "            pass\n",
    "        elif perc_shift > 0.0:\n",
    "            perc_shift = round(abs(perc_shift), 2)\n",
    "            if silent == False:\n",
    "                print('Flux through the objective DECREASED to ~' + str(new_ov) + ' from ~' + str(old_ov) + ' (' + str(perc_shift) + '% change)')\n",
    "        elif perc_shift < 0.0:\n",
    "            perc_shift = round(abs(perc_shift), 2)\n",
    "            if silent == False:\n",
    "                print('Flux through the objective INCREASED to ~' + str(new_ov) + ' from ' + str(old_ov) + ' (' + str(perc_shift) + '% change)')\n",
    "    else:\n",
    "        report_dict['obj_change'] = 'fails'\n",
    "\n",
    "    # Report concordance\n",
    "    if concordance != 'Not performed':\n",
    "        rho = 'r=' + str(round(concordance['r'], 3))\n",
    "        if concordance['p'] <= 0.05:\n",
    "            p_val = round(concordance['p'], 3)\n",
    "            if p_val == 0.0:\n",
    "                p_val = 'p<0.001 *'\n",
    "            else:\n",
    "                p_val = 'p=' + str(p_val) + ' *'\n",
    "            if silent == False:\n",
    "                print('Context-specific metabolism correlates with transcriptome (' + rho + ', ' + p_val + ')')\n",
    "        else:\n",
    "            if silent == False:\n",
    "                print('Context-specific metabolism does not correlate with transcriptome (' + rho + ', n.s.)')\n",
    "\n",
    "    # Run time\n",
    "    raw_seconds = int(round(time.time() - start_time))\n",
    "    report_dict['run_time'] = raw_seconds\n",
    "    if silent == False:\n",
    "        if raw_seconds < 60:\n",
    "            print('\\nRIPTiDe completed in ' + str(raw_seconds) + ' seconds\\n')\n",
    "        else:\n",
    "            minutes, seconds = divmod(raw_seconds, 60)\n",
    "            mins = 'minute'\n",
    "            if minutes > 1:\n",
    "                mins = 'minutes'\n",
    "            secs = 'second'\n",
    "            if seconds > 1:\n",
    "                secs = 'seconds'\n",
    "            \n",
    "            if mf == False:\n",
    "                print('\\nRIPTiDe completed in,', str(minutes), mins, 'and', str(int(seconds)), secs, '\\n')\n",
    "            else:\n",
    "                print('\\nMaxfit RIPTiDe completed in,', str(minutes), mins, 'and', str(int(seconds)), secs, '\\n')\n",
    "\n",
    "    return report_dict\n",
    "\n",
    "#-----------------------------------------------------------------#\n",
    "\n",
    "# gapsplit flux sampler\n",
    "# Keaty TC & Jensen PA (2020)\n",
    "\n",
    "def _gapsplit(model, depth, cpus):\n",
    "    fva = flux_variability_analysis(model, model.reactions, fraction_of_optimum=0.001, processes=cpus)\n",
    "\n",
    "    # only split reactions with feasible range >= min_range\n",
    "    idxs = (fva.maximum - fva.minimum >= 1e-5).to_numpy().nonzero()[0]\n",
    "    weights = (1.0 / (fva.maximum - fva.minimum) ** 2).to_numpy()\n",
    "    samples = numpy.zeros((depth, len(model.reactions)))\n",
    "    k = 0\n",
    "    for try_ in range(1000):\n",
    "        relative, target, width = _maxgap(samples[0:k,idxs], fva.iloc[idxs,:])\n",
    "        primary_var = numpy.argmax(relative)\n",
    "        primary_target = target[primary_var]\n",
    "        primary_lb = primary_target - 0.001*width[primary_var]\n",
    "        primary_ub = primary_target + 0.001*width[primary_var]\n",
    "\n",
    "        new_sample = _generate_sample(model, idxs[primary_var], primary_lb, primary_ub)\n",
    "        if new_sample is not None:\n",
    "            new_sample[new_sample > fva.maximum] = fva.maximum[new_sample > fva.maximum]\n",
    "            new_sample[new_sample < fva.minimum] = fva.minimum[new_sample < fva.minimum]\n",
    "            samples[k,:] = new_sample\n",
    "            k += 1\n",
    "        if k >= depth: break\n",
    "\n",
    "    if k < depth:\n",
    "        # max_tries reached; return fewer samples\n",
    "        samples = samples[:k,:]\n",
    "\n",
    "    return pandas.DataFrame(data=samples, columns=fva.maximum.index)\n",
    "\n",
    "def _generate_sample(model, primary_var, primary_lb, primary_ub):\n",
    "    \n",
    "    # Formulate a [MI]QP to find a single solution\n",
    "    with model:\n",
    "        model.reactions[primary_var].lower_bound = primary_lb\n",
    "        model.reactions[primary_var].upper_bound = primary_ub\n",
    "        model.objective = symengine.RealDouble(0)\n",
    "        solution = model.optimize()\n",
    "        if solution.status != 'optimal':\n",
    "            return None\n",
    "        else:\n",
    "            return solution.fluxes\n",
    "\n",
    "def _maxgap(points, fva):\n",
    "    points = points.copy()\n",
    "    points = numpy.vstack((fva.minimum, points, fva.maximum))\n",
    "    points.sort(0)\n",
    "\n",
    "    gaps = points[1:,:] - points[0:-1,:]\n",
    "    width = gaps.max(0)\n",
    "    loc = gaps.argmax(0)\n",
    "    left = numpy.zeros(width.size)\n",
    "    for i in range(width.size):\n",
    "        left[i] = points[loc[i],i]\n",
    "\n",
    "    relative = width / (points[-1,:] - points[0,:])\n",
    "    target = left + width / 2.0\n",
    "\n",
    "    return relative, target, width\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "iCdR700 = cobra.io.load_json_model('/home/mjenior/Desktop/repos/Jenior_Cdifficile_2019/data/reconstructions/iCdR700.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_dict = {}\n",
    "with open('/home/mjenior/Desktop/active_projects/Tamayo/c-di-AMP/CdR20291_genes.tsv', 'r') as gene_catalog:\n",
    "    header = gene_catalog.readline()\n",
    "    for line in gene_catalog:\n",
    "        line = line.split()\n",
    "        refseq_dict[line[1]] = line[0]\n",
    "        \n",
    "# Read in transcription\n",
    "solo = {}\n",
    "with open('/home/mjenior/Desktop/active_projects/Zackular_enterococcus/solo_norm.tsv', 'r') as transcription:\n",
    "    for line in transcription:\n",
    "        line = line.split()\n",
    "        abund = float(line[1])\n",
    "        try:\n",
    "            patric = refseq_dict[line[0]]\n",
    "        except:\n",
    "            continue\n",
    "        solo[patric] = abund\n",
    "        \n",
    "coculture = {}\n",
    "with open('/home/mjenior/Desktop/active_projects/Zackular_enterococcus/coculture_norm.tsv', 'r') as transcription:\n",
    "    for line in transcription:\n",
    "        line = line.split()\n",
    "        abund = float(line[1])\n",
    "        try:\n",
    "            patric = refseq_dict[line[0]]\n",
    "        except:\n",
    "            continue\n",
    "        coculture[patric] = abund"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n"
     ]
    }
   ],
   "source": [
    "test = contextualize_variability(model=iCdR700, transcriptome=solo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n",
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n"
     ]
    }
   ],
   "source": [
    "test2 = contextualize_variability(model=iCdR700, transcriptome=coculture)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Initializing model and integrating transcriptomic data...\n",
      "Pruning zero flux subnetworks...\n",
      "Analyzing context-specific flux distributions...\n",
      "\n",
      "Reactions pruned to 322 from 1310 (75.42% change)\n",
      "Metabolites pruned to 322 from 1243 (74.09% change)\n",
      "Flux through the objective DECREASED to ~129.85 from ~159.18 (18.43% change)\n",
      "Context-specific metabolism does not correlate with transcriptome (r=0.097, n.s.)\n",
      "\n",
      "RIPTiDe completed in 16 seconds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "test3 = contextualize(model=iCdR700, transcriptome=coculture)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCdR700</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07fe4fb46c150</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>314</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>320</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>0.0 + 1.0*biomass - 1.0*biomass_reverse_01e59</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCdR700 at 0x7fe4fb46c150>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "127.34142612878986"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.slim_optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCdR700</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07fe4fb41bb50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>322</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>322</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>0.0 + 1.0*biomass - 1.0*biomass_reverse_01e59</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCdR700 at 0x7fe4fb41bb50>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCdR700_riptide</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07fe4dee00c90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>322</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>322</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>0.0 + 1.0*biomass - 1.0*biomass_reverse_01e59</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCdR700_riptide at 0x7fe4dee00c90>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test3.model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "129.85240974255447"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test3.model.slim_optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "127.34142612878986"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test2.slim_optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo reactions: 10\n",
      "rxn00193_c glutamate racemase\n",
      "rxn05197_c Uracil permease\n",
      "rxn05209_c Sodium hydrogen symporter\n",
      "rxn00704_c alpha-D-Glucose 1-phosphate 1,6-phosphomutase\n",
      "EX_cpd00092_e Uracil exchange\n",
      "rxn00293_c UTP:N-acetyl-alpha-D-glucosamine-1-phosphate uridylyltransferase\n",
      "rxn00283_c alanine racemase\n",
      "EX_cpd00009_e Phosphate exchange\n",
      "rxn00558_c D-glucose-6-phosphate aldose-ketose-isomerase\n",
      "rxn00579_c UDP-glucose:D-fructose 2-alpha-D-glucosyltransferase\n",
      "\n",
      "Coculture-only reactions: 12\n",
      "rxn08168_c L-aspartate:fumarate oxidoreductase (deaminating)\n",
      "EX_cpd00089_e Glucose-1-phosphate exchange\n",
      "rxn09348_c UDPgalactose pyrophosphohydrolase\n",
      "rxn05562_c Glucose-1-phosphate transport via phosphate antiport\n",
      "rxn00355_c UTP:alpha-D-galactose-1-phosphate uridylyltransferase\n",
      "rxn05554_c Fructose-6-phosphate transport via phosphate antiport\n",
      "rxn13251_c Fumarate dependent DHORD\n",
      "rxn05227_c Aspartate oxidase\n",
      "rxn05672_c D-glutamate transport in via proton symport\n",
      "EX_cpd00072_e D-fructose-6-phosphate exchange\n",
      "EX_cpd00186_e D-Glutamate exchange\n",
      "rxn01359_c (S)-dihydroorotate:NADP+ oxidoreductase\n"
     ]
    }
   ],
   "source": [
    "# Distinct reactions - BHI\n",
    "rxns1 = set([x.id for x in test.reactions])\n",
    "rxns2 = set([x.id for x in test2.reactions])\n",
    "\n",
    "solo_only_rxns = rxns1.difference(rxns2)\n",
    "print('Solo reactions:', len(solo_only_rxns))\n",
    "for x in solo_only_rxns: print(x, iCdR700.reactions.get_by_id(x).name)\n",
    "\n",
    "coculture_only_rxns = rxns2.difference(rxns1)\n",
    "print('\\nCoculture-only reactions:', len(coculture_only_rxns))\n",
    "for x in coculture_only_rxns: print(x, iCdR700.reactions.get_by_id(x).name) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
