{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from riptide import *\n",
    "from cobra.flux_analysis.variability import find_essential_genes, find_essential_reactions\n",
    "from cobra.medium.minimal_medium import minimal_medium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "iCdR703 = cobra.io.load_json_model('/home/mjenior/Desktop/repos/Jenior_Cdifficile_2019/data/reconstructions/iCdR703.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCdG703</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f2d4e2dac10</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1243</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1313</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*biomass - 1.0*biomass_reverse_01e59</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCdG703 at 0x7f2d4e2dac10>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iCdR703"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse RefSeq gene ID lookup table\n",
    "refseq_dict = {}\n",
    "with open('/home/mjenior/Desktop/active_projects/Tamayo/c-di-AMP/CdR20291_genes.tsv', 'r') as gene_catalog:\n",
    "    header = gene_catalog.readline()\n",
    "    for line in gene_catalog:\n",
    "        line = line.split()\n",
    "        refseq_dict[line[1]] = line[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes missing in CD196: 0\n"
     ]
    }
   ],
   "source": [
    "cd196_missing = ['CDR20291_0183','CDR20291_1419','CDR20291_1744','CDR20291_1745','CDR20291_1746',\n",
    "                 'CDR20291_1747','CDR20291_1748','CDR20291_1749','CDR20291_1750','CDR20291_1751',\n",
    "                 'CDR20291_1752','CDR20291_1753','CDR20291_1754','CDR20291_1755','CDR20291_1756',\n",
    "                 'CDR20291_1757','CDR20291_1758','CDR20291_1759','CDR20291_1760','CDR20291_1761',\n",
    "                 'CDR20291_1762','CDR20291_1763','CDR20291_1764','CDR20291_1765','CDR20291_1766',\n",
    "                 'CDR20291_1767','CDR20291_1768','CDR20291_1769','CDR20291_1770','CDR20291_1771',\n",
    "                 'CDR20291_1772','CDR20291_1773','CDR20291_1774','CDR20291_1775','CDR20291_1776',\n",
    "                 'CDR20291_1777','CDR20291_1778','CDR20291_1779','CDR20291_1780','CDR20291_1781',\n",
    "                 'CDR20291_1782','CDR20291_1783','CDR20291_1784','CDR20291_1785','CDR20291_1786',\n",
    "                 'CDR20291_1787','CDR20291_1788','CDR20291_1809']\n",
    "\n",
    "check_missing = []\n",
    "for x in cd196_missing:\n",
    "    try:\n",
    "        check_missing.append(refseq_dict[x])\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "check_missing = set(check_missing)\n",
    "included_genes = []\n",
    "for x in iCdR703.genes:\n",
    "    if x.id in check_missing:\n",
    "        included_genes.append(x.id)\n",
    "\n",
    "print('Genes missing in CD196:', len(included_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in transcription\n",
    "solo = {}\n",
    "with open('/home/mjenior/Desktop/active_projects/Zackular_enterococcus/solo_norm.tsv', 'r') as transcription:\n",
    "    for line in transcription:\n",
    "        line = line.split()\n",
    "        abund = float(line[1])\n",
    "        try:\n",
    "            patric = refseq_dict[line[0]]\n",
    "            solo[patric] = abund\n",
    "        except:\n",
    "            continue\n",
    "        \n",
    "coculture = {}\n",
    "with open('/home/mjenior/Desktop/active_projects/Zackular_enterococcus/coculture_norm.tsv', 'r') as transcription:\n",
    "    for line in transcription:\n",
    "        line = line.split()\n",
    "        abund = float(line[1])\n",
    "        try:\n",
    "            patric = refseq_dict[line[0]]\n",
    "            coculture[patric] = abund\n",
    "        except:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "\n",
    "iCdR700 = cobra.io.load_json_model('/home/mjenior/Desktop/repos/Jenior_Cdifficile_2019/data/reconstructions/iCdR700.json')\n",
    "\n",
    "# Ornithine conversion and alanine fermentation\n",
    "\n",
    "cpd00064_e = iCdR700.metabolites.cpd00064_e # ornithine\n",
    "cpd00064_c = iCdR700.metabolites.cpd00064_c # ornithine\n",
    "cpd00001_c = iCdR700.metabolites.cpd00001_c # H20\n",
    "cpd00002_c = iCdR700.metabolites.cpd00002_c # ATP\n",
    "cpd00008_c = iCdR700.metabolites.cpd00008_c # ADP\n",
    "cpd00009_c = iCdR700.metabolites.cpd00009_c # Phosphate\n",
    "cpd00067_c = iCdR700.metabolites.cpd00067_c # H+\n",
    "\n",
    "rxn05156_c = cobra.Reaction('rxn05156_c')\n",
    "rxn05156_c.name = 'Ornithine ABC transport'\n",
    "rxn05156_c.lower_bound = 0.\n",
    "rxn05156_c.upper_bound = 1000.\n",
    "rxn05156_c.add_metabolites({\n",
    "    cpd00001_c: -1.0,\n",
    "    cpd00002_c: -1.0,\n",
    "    cpd00064_e: -1.0,\n",
    "    cpd00008_c: 1.0,\n",
    "    cpd00009_c: 1.0,\n",
    "    cpd00067_c: 1.0,\n",
    "    cpd00064_c: 1.0\n",
    "})\n",
    "rxn05156_c.annotation['seed.reaction'] = 'rxn05156'\n",
    "rxn05156_c.annotation['bigg.reaction'] = ['ORNabc','ORNabcpp']\n",
    "rxn05156_c.gene_reaction_rule = '645463.3.peg.862 or 645463.3.peg.865'\n",
    "# str. 630 = or\n",
    "iCdR700.add_reactions([rxn05156_c])\n",
    "iCdR700.reactions.rxn05156_c.annotation['sbo'] = 'SBO:0000629'\n",
    "\n",
    "cpd00129_c = iCdR700.metabolites.cpd00129_c # proline\n",
    "cpd00013_c = iCdR700.metabolites.cpd00013_c # ammonia\n",
    "\n",
    "rxn00471_c = cobra.Reaction('rxn00471_c')\n",
    "rxn00471_c.name = 'Ornithine cyclodeaminase'\n",
    "rxn00471_c.lower_bound = 0.\n",
    "rxn00471_c.upper_bound = 1000.\n",
    "rxn00471_c.add_metabolites({\n",
    "    cpd00064_c: -1.0,\n",
    "    cpd00129_c: 1.0,\n",
    "    cpd00013_c: 1.0,\n",
    "    cpd00001_c: 1.0\n",
    "})\n",
    "rxn00471_c.annotation['seed.reaction'] = 'rxn00471'\n",
    "rxn00471_c.annotation['kegg.reaction'] = 'R00671'\n",
    "rxn00471_c.gene_reaction_rule = '645463.3.peg.500 or 645463.3.peg.260'\n",
    "# str. 630 = or\n",
    "iCdR700.add_reactions([rxn00471_c])\n",
    "iCdR700.reactions.rxn00471_c.annotation['sbo'] = 'SBO:0000629'\n",
    "\n",
    "iCdR700.reactions.rxn14293_c.gene_reaction_rule = '645463.3.peg.500 or 645463.3.peg.260'\n",
    "iCdR700.reactions.rxn00655_c.gene_reaction_rule= ''\n",
    "\n",
    "#------------------------------------------------------------------------------#\n",
    "\n",
    "# Alanine Stickland fermentation pathway\n",
    "\n",
    "# Alanine --> Pyruvate\n",
    "cpd00035_c = iCdR700.metabolites.cpd00035_c # alanine\n",
    "cpd00020_c = iCdR700.metabolites.cpd00020_c # Pyruvate\n",
    "\n",
    "R00258_c = cobra.Reaction('R00258_c')\n",
    "R00258_c.name = 'Alanine transaminase'\n",
    "R00258_c.lower_bound = -1000.\n",
    "R00258_c.upper_bound = 1000.\n",
    "R00258_c.add_metabolites({\n",
    "    cpd00035_c: -1.0,\n",
    "    cpd00001_c: -1.0,\n",
    "    cpd00020_c: 1.0,\n",
    "    cpd00013_c: 1.0,\n",
    "    cpd00067_c: 1.0\n",
    "})\n",
    "R00258_c.gene_reaction_rule = '645463.3.peg.2419'\n",
    "# str. 630 = 272563.120.peg.2623\n",
    "iCdR700.add_reactions([R00258_c])\n",
    "iCdR700.reactions.R00258_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdR700.reactions.R00258_c.annotation['kegg.reaction'] = 'R00258'\n",
    "\n",
    "# Pyruvate --> Acetyl-CoA (already present: rxn00157_c)\n",
    "cpd00010_c = iCdR700.metabolites.cpd00010_c # CoA\n",
    "cpd00022_c = iCdR700.metabolites.cpd00022_c # Acetyl-CoA\n",
    "cpd00047_c = iCdR700.metabolites.cpd00047_c # Formate\n",
    "\n",
    "# Acetyl-CoA --> Acetylphosphate\n",
    "cpd00196_c = iCdR700.metabolites.cpd00196_c # Acetylphosphate\n",
    "\n",
    "rxn00173_c = cobra.Reaction('rxn00173_c')\n",
    "rxn00173_c.name = 'Phosphate acetyltransferase'\n",
    "rxn00173_c.lower_bound = -1000.\n",
    "rxn00173_c.upper_bound = 1000.\n",
    "rxn00173_c.add_metabolites({\n",
    "    cpd00022_c: -1.0,\n",
    "    cpd00009_c: -1.0,\n",
    "    cpd00196_c: 1.0,\n",
    "    cpd00010_c: 1.0\n",
    "})\n",
    "rxn00173_c.gene_reaction_rule = '645463.3.peg.125'\n",
    "# str. 630 = 272563.120.peg.124\n",
    "iCdR700.add_reactions([rxn00173_c])\n",
    "iCdR700.reactions.rxn00173_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdR700.reactions.rxn00173_c.annotation['seed.reaction'] = 'rxn00173'\n",
    "iCdR700.reactions.rxn00173_c.annotation['kegg.reaction'] = 'R00230'\n",
    "iCdR700.reactions.rxn00173_c.annotation['bigg.reaction'] = ['PTA','PTAr']\n",
    "\n",
    "# Acetylphosphate --> Acetate (already in place: rxn00225_c)\n",
    "\n",
    "#------------------------------------------------------------------------------#\n",
    "\n",
    "# Ornithine conversion to alanine\n",
    "\n",
    "cpd02448_c = cobra.Metabolite(\n",
    "    'cpd02448_c',\n",
    "    formula='C5H13N2O2',\n",
    "    name='2,4-diaminopentanoate',\n",
    "    compartment='cytosol')\n",
    "cpd02448_c.annotation['seed.compound'] = 'cpd02448'\n",
    "cpd02448_c.annotation['kegg.compound'] = 'C03943'\n",
    "\n",
    "gene = cobra.Gene('645463.3.peg.416')\n",
    "\n",
    "# D-ornithine <--> 2,4-diaminopentanoate\n",
    "rxn01780_c = cobra.Reaction('rxn01780_c')\n",
    "rxn01780_c.name = 'Ornithine aminomutase'\n",
    "rxn01780_c.lower_bound = -1000.\n",
    "rxn01780_c.upper_bound = 1000.\n",
    "rxn01780_c.add_metabolites({\n",
    "    cpd00064_c: -1.0,\n",
    "    cpd02448_c: 1.0\n",
    "})\n",
    "rxn01780_c.gene_reaction_rule = '645463.3.peg.416'\n",
    "# str. 630 = 272563.120.peg.471\n",
    "iCdR700.add_reactions([rxn01780_c])\n",
    "iCdR700.reactions.rxn01780_c.annotation['kegg.reaction'] = 'R02461'\n",
    "iCdR700.reactions.rxn01780_c.annotation['sbo'] = 'SBO:0000629'\n",
    "\n",
    "iCdR700.genes.get_by_id('645463.3.peg.416').name = 'D-Ornithine 4,5-aminomutase E subunit (EC 5.4.3.5) | oraE'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.416').annotation['sbo'] = 'SBO:0000243'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.416').annotation['patric'] = 'fig|645463.3.peg.416'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.416').annotation['refseq'] = 'YP_003216896.1'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.416').annotation['refseq.locustag'] = 'CDR20291_0389'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.416').annotation['ccds'] = 'ATGTTAAAGGAAAATAAAAAATTGGATATAGACTATATATTAAAAGATTTGGATAAATATAAACCAAAACGAAGAGGTTGGGTTTGGAGAGAACACTTGGAAAATCTTGAAATGGGACCTTTTAAATATAAAGACTGTACAAAACCTTTAAAGAAGAGTGTAGGGCTACCATCATCTAAGTATTTCAATAATATAGACCCTCAACCAAGTCCAGTTATTACTACTGAAATAGCTTCTGGTAGATTTGAGGATGATATAAGAAGAATGAGAATGGCAGCTCATCATGGAGCTGACCATCTAATGGTTATCAGAACTGCTGGACAATCACATTTTGATGGGCTTATAGAAGGAACTCCACAAGGTATAGGTGGAGTTCCAATAACAAGAAAACAAGTAAGAGCTCAAAGGAAGGCTCTTGATTTTATAGAAGAAGAAGTTGGTAGACCTATTAATTACCATTCATATGTAAGTGGTGTTGCTGGTCCTGAAGTAGCAGTAATGTTTGCTGAGGAAGGAGTCAATGGAGCACATCAAGACCCACAATATAATGTTCTGTATAGGAATATAAACATGGTAAGGTCTTTTGTAGATGCTTGTGAGGCAAAGAAAATAATGGCTTTTGCCAATATAGCTCAAATTGATGGAGCACATAATGCTAATGCTACTGCTAGAGAAGCATGGAAGGTAATGCCTGAACTTATGGTTCAACACGCTTTAAACTCTATATTCTCAGAAAAAATAGGTATAGATAAAAGTAATATATGCTTATCAACAGTTCCACCAACAGCACCTCCAGCACCATGTTTAAGAATAGACTTACCTTACGCTGTTGCACTTAGAGAATTATTTAGTGATTATAAGATGAGAGCTCAAATGAATACTAAATATATGGAATCATCAACTAGAGAAGCAACAGTTACTCATGTATTAAACTTATTAACATCAGTATTAACTAGAGCTGACATACAATCAACTATAACACCTGATGAAGGTAGAAATGTACCTTGGCATATATACAATATAGAAGCATGTGATACAGCAAAGCAAGCTTTAGTTGGTATGGATGGGCTTATGGATATGATAGAACTTAAAGATGTCGGTGAATTAAGAGACAAAGCAAGAGAACTTAAAGAAAGAGCTGTATTATATATGGAAGAAATATCTGAAGTTGGTGGATATTTTGAATCTGTGGAACAAGGATTTTTTGTTGATTCTGGAAATTATCCAGAAAGAAATGGTGATGGAATATCTAGAAAAATAAAAGGTGGAGTTGGTGAAGGTACTGTTTATGAAAGAGAAGAAGATTATTTAGCACCTGTAACTGCTCACTTTGGATACAATAATGTAGCTCAATATGATGAAAATGCTATAGATAATCCATCAATGTTAATAGATGGATGTACATTTGAAAATCCAGACAAAATAATATATATTGATGAGCTTGACGATTTTGATAATGTAGAAAATAGACTAAAAGAATCATATGAGTACAGAAATGGAACTAAAATTAAGCCTGAAATGGAATGGTGTGCAGATGGAATTGTAATGATAACAATGATGCTTCCTACTGATAAAAGAACAGCAGAGTTTGCAGCTTTGGAATTTGTCAAAAAAATGAACTTACAAGAAATTGAGGTCATAAGTAGAGAAGTTATGCATGAATCTGAAGGGACTAGAATAGAAGTGAAAGGAAGAGTTCCGTTTGATATAGATTTAAACAATTTAGTAATACCAGAAGAACCTAAGGTACTAACTGATGAAGAAATAAGAGCAGATATAGAAGAAAAACCTATGAAAATAGTTTCAGCTACGGTTGGAGAAGATGAACATTCAGTTGGTTTAAGAGAAATAATAGATATAAAACATGGAGGTATTGAAAAATATGGCATAGAATGCCACTATCTAGGGACTTCTGTACCTGTAGAAAAATTGGTAGATGCAGCGATTGAACTGAATGCAGATGCAATACTTGCATCTACAATAATAAGTCATGATGATATTCACTATAAAAACATGAAAAAATTACACGATTATTGTGTAGAAAAAGGTATAAGAGATAAAGTAATGATTGCGTGTGGAGGAACACAAGTAACACCTGAAATAGCTGTTGAACAAGGCATCGATGCTGGATTTGGAAGAAACAGCAAAGGCATACATGTAGCCACTTTCTTAGTTGAAAAAAGAAGAGAAATGAGTAATAAATAA'\n",
    "\n",
    "cpd24394_c = cobra.Metabolite(\n",
    "    'cpd24394_c',\n",
    "    formula='C5H9NO3',\n",
    "    name='2-amino-4-ketopentanoate',\n",
    "    compartment='cytosol')\n",
    "cpd24394_c.annotation['sbo'] = 'SBO:0000247'\n",
    "cpd24394_c.annotation['seed.compound'] = 'cpd02448'\n",
    "cpd24394_c.annotation['metacyc'] = 'CPD-299'\n",
    "\n",
    "cpd00006_c = iCdR700.metabolites.cpd00006_c # NADP\n",
    "cpd00005_c = iCdR700.metabolites.cpd00005_c # NADPH\n",
    "\n",
    "# 2,4-diaminopentanoate <--> 2-amino-4-ketopentanoate\n",
    "rxn27496_c = cobra.Reaction('rxn27496_c')\n",
    "rxn27496_c.name = '2,4-diaminopentanoate dehydrogenase'\n",
    "rxn27496_c.lower_bound = -1000.\n",
    "rxn27496_c.upper_bound = 1000.\n",
    "rxn27496_c.add_metabolites({\n",
    "    cpd00001_c: -1.0,\n",
    "    cpd02448_c: -1.0,\n",
    "    cpd00006_c: -1.0,\n",
    "    cpd00013_c: 1.0,\n",
    "    cpd00067_c: 1.0,\n",
    "    cpd24394_c: 1.0,\n",
    "    cpd00005_c: 1.0\n",
    "})\n",
    "rxn27496_c.gene_reaction_rule = '645463.3.peg.412'\n",
    "# str. 630 = 272563.120.peg.467\n",
    "iCdR700.add_reactions([rxn27496_c])\n",
    "iCdR700.reactions.rxn01905_c.gene_reaction_rule = ''\n",
    "iCdR700.reactions.rxn27496_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdR700.reactions.rxn27496_c.annotation['seed.reaction'] = 'rxn27496'\n",
    "\n",
    "gene2 = cobra.Gene('645463.3.peg.413')\n",
    "gene3 = cobra.Gene('645463.3.peg.414')\n",
    "\n",
    "# 2-amino-4-ketopentanoate <--> alanine\n",
    "rxn19080_c = cobra.Reaction('rxn19080_c')\n",
    "rxn19080_c.name = '2-amino-4-oxopentanoate thiolase'\n",
    "rxn19080_c.lower_bound = -1000.\n",
    "rxn19080_c.upper_bound = 1000.\n",
    "rxn19080_c.add_metabolites({\n",
    "    cpd00010_c: -1.0,\n",
    "    cpd24394_c: -1.0,\n",
    "    cpd00022_c: 1.0,\n",
    "    cpd00035_c: 1.0\n",
    "})\n",
    "rxn19080_c.gene_reaction_rule = '645463.3.peg.413 and 645463.3.peg.414'\n",
    "# str. 630 = and\n",
    "iCdR700.add_reactions([rxn19080_c])\n",
    "iCdR700.reactions.rxn19080_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdR700.reactions.rxn19080_c.annotation['seed.reaction'] = 'rxn19080'\n",
    "\n",
    "iCdR700.genes.get_by_id('645463.3.peg.413').name = '2-amino-4-ketopentanoate thiolase alpha subunit [EC:2.3.1.263] | ortA'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.413').annotation['sbo'] = 'SBO:0000243'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.413').annotation['patric'] = 'fig|645463.3.peg.413'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.413').annotation['kegg'] = 'K21399'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.413').annotation['refseq'] = 'YP_003216893.1'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.413').annotation['refseq.locustag'] = 'CDR20291_0386'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.413').annotation['ccds'] = 'ATGGATGCTAAAATTAATGATTGGGTTATTATACACAATATAGTATTGACTCCAGAAGAAAGAGCTCCACAAGTTCCAGAGGATACTAAGAAAGTATCCTTAGAAATGTGGGTTAAAGGTTTTATACAAAGTGATGCTTCTATAGGCGACTTAGTAGAAGTTAAAACAATTACAGGTAGACTTGTAAAAGGTAATTTACTTAAAGTAAATCCATATTATACTCATGATTATGGTAAATGTATACCAGAGTTACTCCAAATAGGAATCCAAGCTAAAGAAATATTATTTGGAGGTGTATATAATGAATAA'\n",
    "\n",
    "iCdR700.genes.get_by_id('645463.3.peg.414').name = '2-amino-4-ketopentanoate thiolase beta subunit [EC:2.3.1.263] | ortA'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.414').annotation['sbo'] = 'SBO:0000243'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.414').annotation['patric'] = 'fig|645463.3.peg.414'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.414').annotation['kegg'] = 'K21400'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.414').annotation['refseq'] = 'YP_003216894.1'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.414').annotation['refseq.locustag'] = 'CDR20291_0387'\n",
    "iCdR700.genes.get_by_id('645463.3.peg.414').annotation['ccds'] = 'ATGAATAATTCAACATCAATAAAAGATATGAGTTATCAAGCTGTTATGGGAAGAAATAACGAAATAATGAAAAATGCAATCGGATTGGATTATTCTTCTTTTGAACAAGAGGGAATAGGGTTTGATTATGAGAAAATGATGAGTGAAACAGGATATACATTACAAGATATTGAAGCAATTCAATCACAGTATGCTGTTGGAAATACTCCTCTAATAGAGCTTAAAAATTTGACTAAATTAGCTAGAAAATGTGCTAAAGAAGGTAAAGGAGCTAAAATATTTGTAAAAGATGAAGCAATGAATGCATCTGGTAGTTTTAAAGCAAGAAGAGCAGCTACTGCAGTATATCATGCCAAACAAATGGGATATAAAGGTGTAATAGCAGCTACCAGTGGTAACTATGGTGCAGCTGTAGCATCACAAGCGGCAATGCAGGGATTAAAATGTATAATAGTTCAAGAATGTTATGATTCTAATGGTGTAGGTCAACCTGAAATTATAGAAAAAGCACGAAAATGTGAGGCATATGGAGCAGAAGTTGTTCAGTTAACTGTTGGACCAGAGTTATTTTATACGTTCTTAGTTTTATTAGAAGAAACAGGCTATTTTAACGCATCATTATATTCACCATTTGGAATTGCAGGTGTAGAAACTTTAGGTTATGAGTTAGCTATCCAATTTAGAGAAAAGTATAAAAAAGACCCTGATATAGTTGTTTGTACAAATGCAGGTGGTGGTAATCTTACTGGAACTGCAAGAGGATTGATAAAAGCTGGTTCCATTAATACTAAAGTAGTAGGAGCAAGTGTAGATTTAAAAGGATTACACATGGCTAGTGATAATCAATTTAATAAAAAATCATTTACAACAGGACATACAGGTTTTGGGATACCATACTGTACATGGCCTGATAGGTCAGATGTACCTAGGTCTGCAGCTAGACCACTTAGATATATGGATAGATATGTATTAGTAAAACAAGGAGAAGTTTTTTATACAACTGAATTATTGGCACAGCTAGAAGGTATAGAAAGAGGTCCAGCAGGAAATACATCTTTAGCAGCAGCATTCTCTTTAGCTCAAGAGTTAGATGAGGATAAAGTTATAGTGGTACAAGAAACAGAATATACAGGCGCAGGTAAACATATATGTCCACAGCTTACTTTTGCAAGAGAAAATGGAATAGATATAAAATTTGGAAATCCAAGAGAAGAAATTGCAGGAGTAAACTTAATATTACCAGAAAGACCAGAATTATTGAAATGTGTTGATATTGATATGAATAAAATAAGAAAGTCATTTATAAAAAATTGCATATCTAATAATCACATTGATAATATTGATAAGCTAAGTAATCGAGATATAGAATTTTTAATGGAAGAAGTAAAAAGTTCAAGAGATTTTGTTGTTGATGTATTAAATAATTTATAG'\n",
    "\n",
    "#------------------------------------------------------------------------------#\n",
    "\n",
    "# Correct p-Cresol metabolism\n",
    "iCdR700.reactions.rxn07124_c.remove_from_model(remove_orphans=True)\n",
    "iCdR700.reactions.ID008_c.remove_from_model(remove_orphans=True)\n",
    "iCdR700.reactions.EX_cpd03170_e.remove_from_model(remove_orphans=True)\n",
    "iCdR700.reactions.ID004_c.remove_from_model(remove_orphans=True)\n",
    "\n",
    "iCdR700.reactions.rxn05061_c.gene_reaction_rule = '645463.3.peg.167 and 645463.3.peg.168 and 645463.3.peg.169'\n",
    "\n",
    "iCdR700.reactions.ID005_c.id = 'rxn29230_c'\n",
    "iCdR700.reactions.rxn29230_c.annotation['seed.reaction'] = 'rxn29230'\n",
    "iCdR700.reactions.ID009_c.id = 'rxn08184_c'\n",
    "iCdR700.reactions.rxn08184_c.annotation['seed.reaction'] = 'rxn08184'\n",
    "\n",
    "# Correct phenylpropanoate transport\n",
    "iCdR700.reactions.ID007_c.remove_from_model(remove_orphans=False)\n",
    "cpd03343_c = iCdR700.metabolites.cpd03343_c # Phenylpropanoate\n",
    "cpd03343_e = iCdR700.metabolites.cpd03343_e # Phenylpropanoate\n",
    "\n",
    "rxn10133_c = cobra.Reaction('rxn10133_c')\n",
    "rxn10133_c.name = 'Phenylpropanoate hydrogen symport'\n",
    "rxn10133_c.lower_bound = -1000.\n",
    "rxn10133_c.upper_bound = 1000.\n",
    "rxn10133_c.add_metabolites({\n",
    "    cpd00067_c: -1.0,\n",
    "    cpd03343_c: -1.0,\n",
    "    cpd00067_c: 1.0,\n",
    "    cpd03343_e: 1.0\n",
    "})\n",
    "iCdR700.add_reactions([rxn10133_c])\n",
    "iCdR700.reactions.rxn10133_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdR700.reactions.rxn10133_c.annotation['seed.reaction'] = 'rxn10133'\n",
    "iCdR700.reactions.rxn10133_c.annotation['bigg.reaction'] = ['PPPNt2','PPPNt2r','PPPNt2rpp']\n",
    "\n",
    "\n",
    "#cobra.io.save_json_model(iCdR700, '/home/mjenior/Desktop/repos/Jenior_Cdifficile_2019/data/reconstructions/iCdR700_v1.1.sbml')\n",
    "#cobra.io.write_sbml_model(iCdR700, '/home/mjenior/Desktop/repos/Jenior_Cdifficile_2019/data/reconstructions/iCdR700_v1.1.json')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "iCdG698 = cobra.io.load_json_model('/home/mjenior/Desktop/repos/Jenior_Cdifficile_2019/data/reconstructions/iCdG698.json')\n",
    "\n",
    "# Ornithine conversion and alanine fermentation\n",
    "\n",
    "cpd00064_e = iCdG698.metabolites.cpd00064_e # ornithine\n",
    "cpd00064_c = iCdG698.metabolites.cpd00064_c # ornithine\n",
    "cpd00001_c = iCdG698.metabolites.cpd00001_c # H20\n",
    "cpd00002_c = iCdG698.metabolites.cpd00002_c # ATP\n",
    "cpd00008_c = iCdG698.metabolites.cpd00008_c # ADP\n",
    "cpd00009_c = iCdG698.metabolites.cpd00009_c # Phosphate\n",
    "cpd00067_c = iCdG698.metabolites.cpd00067_c # H+\n",
    "\n",
    "rxn05156_c = cobra.Reaction('rxn05156_c')\n",
    "rxn05156_c.name = 'Ornithine ABC transport'\n",
    "rxn05156_c.lower_bound = 0.\n",
    "rxn05156_c.upper_bound = 1000.\n",
    "rxn05156_c.add_metabolites({\n",
    "    cpd00001_c: -1.0,\n",
    "    cpd00002_c: -1.0,\n",
    "    cpd00064_e: -1.0,\n",
    "    cpd00008_c: 1.0,\n",
    "    cpd00009_c: 1.0,\n",
    "    cpd00067_c: 1.0,\n",
    "    cpd00064_c: 1.0\n",
    "})\n",
    "rxn05156_c.annotation['seed.reaction'] = 'rxn05156'\n",
    "rxn05156_c.annotation['bigg.reaction'] = ['ORNabc','ORNabcpp']\n",
    "rxn05156_c.gene_reaction_rule = '272563.120.peg.897 or 272563.120.peg.900'\n",
    "iCdG698.add_reactions([rxn05156_c])\n",
    "iCdG698.reactions.rxn05156_c.annotation['sbo'] = 'SBO:0000629'\n",
    "\n",
    "cpd00129_c = iCdG698.metabolites.cpd00129_c # proline\n",
    "cpd00013_c = iCdG698.metabolites.cpd00013_c # ammonia\n",
    "\n",
    "rxn00471_c = cobra.Reaction('rxn00471_c')\n",
    "rxn00471_c.name = 'Ornithine cyclodeaminase'\n",
    "rxn00471_c.lower_bound = 0.\n",
    "rxn00471_c.upper_bound = 1000.\n",
    "rxn00471_c.add_metabolites({\n",
    "    cpd00064_c: -1.0,\n",
    "    cpd00129_c: 1.0,\n",
    "    cpd00013_c: 1.0,\n",
    "    cpd00001_c: 1.0\n",
    "})\n",
    "rxn00471_c.annotation['seed.reaction'] = 'rxn00471'\n",
    "rxn00471_c.annotation['kegg.reaction'] = 'R00671'\n",
    "rxn00471_c.gene_reaction_rule = '272563.120.peg.553'\n",
    "# str. 630 = or\n",
    "iCdG698.add_reactions([rxn00471_c])\n",
    "iCdG698.reactions.rxn00471_c.annotation['sbo'] = 'SBO:0000629'\n",
    "\n",
    "iCdG698.reactions.rxn14293_c.gene_reaction_rule = '272563.120.peg.553'\n",
    "# str. 630 = 272563.120.peg.553 or\n",
    "iCdG698.reactions.rxn00655_c.gene_reaction_rule= ''\n",
    "\n",
    "#------------------------------------------------------------------------------#\n",
    "\n",
    "# Alanine Stickland fermentation pathway\n",
    "\n",
    "# Alanine --> Pyruvate\n",
    "cpd00035_c = iCdG698.metabolites.cpd00035_c # alanine\n",
    "cpd00020_c = iCdG698.metabolites.cpd00020_c # Pyruvate\n",
    "\n",
    "R00258_c = cobra.Reaction('R00258_c')\n",
    "R00258_c.name = 'Alanine transaminase'\n",
    "R00258_c.lower_bound = -1000.\n",
    "R00258_c.upper_bound = 1000.\n",
    "R00258_c.add_metabolites({\n",
    "    cpd00035_c: -1.0,\n",
    "    cpd00001_c: -1.0,\n",
    "    cpd00020_c: 1.0,\n",
    "    cpd00013_c: 1.0,\n",
    "    cpd00067_c: 1.0\n",
    "})\n",
    "R00258_c.gene_reaction_rule = '272563.120.peg.2623'\n",
    "iCdG698.add_reactions([R00258_c])\n",
    "iCdG698.reactions.R00258_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdG698.reactions.R00258_c.annotation['kegg.reaction'] = 'R00258'\n",
    "\n",
    "# Pyruvate --> Acetyl-CoA (already present: rxn00157_c)\n",
    "cpd00010_c = iCdG698.metabolites.cpd00010_c # CoA\n",
    "cpd00022_c = iCdG698.metabolites.cpd00022_c # Acetyl-CoA\n",
    "cpd00047_c = iCdG698.metabolites.cpd00047_c # Formate\n",
    "\n",
    "# Acetyl-CoA --> Acetylphosphate\n",
    "cpd00196_c = iCdG698.metabolites.cpd00196_c # Acetylphosphate\n",
    "\n",
    "rxn00173_c = cobra.Reaction('rxn00173_c')\n",
    "rxn00173_c.name = 'Phosphate acetyltransferase'\n",
    "rxn00173_c.lower_bound = -1000.\n",
    "rxn00173_c.upper_bound = 1000.\n",
    "rxn00173_c.add_metabolites({\n",
    "    cpd00022_c: -1.0,\n",
    "    cpd00009_c: -1.0,\n",
    "    cpd00196_c: 1.0,\n",
    "    cpd00010_c: 1.0\n",
    "})\n",
    "rxn00173_c.gene_reaction_rule = '272563.120.peg.124'\n",
    "iCdG698.add_reactions([rxn00173_c])\n",
    "iCdG698.reactions.rxn00173_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdG698.reactions.rxn00173_c.annotation['seed.reaction'] = 'rxn00173'\n",
    "iCdG698.reactions.rxn00173_c.annotation['kegg.reaction'] = 'R00230'\n",
    "iCdG698.reactions.rxn00173_c.annotation['bigg.reaction'] = ['PTA','PTAr']\n",
    "\n",
    "# Acetylphosphate --> Acetate (already in place: rxn00225_c)\n",
    "\n",
    "#------------------------------------------------------------------------------#\n",
    "\n",
    "# Ornithine conversion to alanine\n",
    "\n",
    "cpd02448_c = cobra.Metabolite(\n",
    "    'cpd02448_c',\n",
    "    formula='C5H13N2O2',\n",
    "    name='2,4-diaminopentanoate',\n",
    "    compartment='cytosol')\n",
    "cpd02448_c.annotation['seed.compound'] = 'cpd02448'\n",
    "cpd02448_c.annotation['kegg.compound'] = 'C03943'\n",
    "\n",
    "gene = cobra.Gene('272563.120.peg.471')\n",
    "\n",
    "# D-ornithine <--> 2,4-diaminopentanoate\n",
    "rxn01780_c = cobra.Reaction('rxn01780_c')\n",
    "rxn01780_c.name = 'Ornithine aminomutase'\n",
    "rxn01780_c.lower_bound = -1000.\n",
    "rxn01780_c.upper_bound = 1000.\n",
    "rxn01780_c.add_metabolites({\n",
    "    cpd00064_c: -1.0,\n",
    "    cpd02448_c: 1.0\n",
    "})\n",
    "rxn01780_c.gene_reaction_rule = '272563.120.peg.471'\n",
    "iCdG698.add_reactions([rxn01780_c])\n",
    "iCdG698.reactions.rxn01780_c.annotation['kegg.reaction'] = 'R02461'\n",
    "iCdG698.reactions.rxn01780_c.annotation['sbo'] = 'SBO:0000629'\n",
    "\n",
    "iCdG698.genes.get_by_id('272563.120.peg.471').name = 'D-Ornithine 4,5-aminomutase E subunit (EC 5.4.3.5) | oraE'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.471').annotation['sbo'] = 'SBO:0000243'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.471').annotation['patric'] = 'fig|272563.120.peg.471'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.471').annotation['refseq'] = 'AJP10141.1'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.471').annotation['refseq.locustag'] = 'CDIF630_00574'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.471').annotation['ccds'] = 'ATGTTAAAGGAAAATAAAAAATTGGATATAGACTATATATTAAAAGATTTGGATAAATATAAACCAAAACGAAGAGGTTGGGTTTGGAGAGAACACTTGGAAAATCTTGAAATGGGACCTTTTAAATATAAAGACTGTACAAAACCTTTAAAGAAGAGTGTAGGGCTACCATCATCTAAGTATTTCAATAATATAGACCCTCAACCAAGTCCAGTTATTACTACTGAAATAGCTTCTGGTAGATTTGAGGATGATATAAGAAGAATGAGAATGGCAGCTCATCATGGAGCTGACCATCTAATGGTTATCAGAACTGCTGGACAATCACATTTTGATGGGCTTATAGAAGGAACTCCACAAGGTATAGGTGGAGTTCCAATAACAAGAAAACAAGTAAGAGCTCAAAGGAAGGCTCTTGATTTTATAGAAGAAGAAGTTGGTAGACCTATTAATTACCATTCATATGTAAGTGGTGTTGCTGGTCCTGAAGTAGCAGTAATGTTTGCTGAGGAAGGAGTCAATGGAGCACATCAAGACCCACAATATAATGTTCTGTATAGGAATATAAACATGGTAAGGTCTTTTGTAGATGCTTGTGAGGCAAAGAAAATAATGGCTTTTGCCAATATAGCTCAAATTGATGGAGCACATAATGCTAATGCTACTGCTAGAGAAGCATGGAAGGTAATGCCTGAACTTATGGTTCAACACGCTTTAAACTCTATATTCTCAGAAAAAATAGGTATAGATAAAAGTAATATATGCTTATCAACAGTTCCACCAACAGCACCTCCAGCACCATGTTTAAGAATAGACTTACCTTACGCTGTTGCACTTAGAGAATTATTTAGTGATTATAAGATGAGAGCTCAAATGAATACTAAATATATGGAATCATCAACTAGAGAAGCAACAGTTACTCATGTATTAAACTTATTAACATCAGTATTAACTAGAGCTGACATACAATCAACTATAACACCTGATGAAGGTAGAAATGTACCTTGGCATATATACAATATAGAAGCATGTGATACAGCAAAGCAAGCTTTAGTTGGTATGGATGGGCTTATGGATATGATAGAACTTAAAGATGTCGGTGAATTAAGAGACAAAGCAAGAGAACTTAAAGAAAGAGCTGTATTATATATGGAAGAAATATCTGAAGTTGGTGGATATTTTGAATCTGTGGAACAAGGATTTTTTGTTGATTCTGGAAATTATCCAGAAAGAAATGGTGATGGAATATCTAGAAAAATAAAAGGTGGAGTTGGTGAAGGTACTGTTTATGAAAGAGAAGAAGATTATTTAGCACCTGTAACTGCTCACTTTGGATACAATAATGTAGCTCAATATGATGAAAATGCTATAGATAATCCATCAATGTTAATAGATGGATGTACATTTGAAAATCCAGACAAAATAATATATATTGATGAGCTTGACGATTTTGATAATGTAGAAAATAGACTAAAAGAATCATATGAGTACAGAAATGGAACTAAAATTAAGCCTGAAATGGAATGGTGTGCAGATGGAATTGTAATGATAACAATGATGCTTCCTACTGATAAAAGAACAGCAGAGTTTGCAGCTTTGGAATTTGTCAAAAAAATGAACTTACAAGAAATTGAGGTCATAAGTAGAGAAGTTATGCATGAATCTGAAGGGACTAGAATAGAAGTGAAAGGAAGAGTTCCGTTTGATATAGATTTAAACAATTTAGTAATACCAGAAGAACCTAAGGTACTAACTGATGAAGAAATAAGAGCAGATATAGAAGAAAAACCTATGAAAATAGTTTCAGCTACGGTTGGAGAAGATGAACATTCAGTTGGTTTAAGAGAAATAATAGATATAAAACATGGAGGTATTGAAAAATATGGCATAGAATGCCACTATCTAGGGACTTCTGTACCTGTAGAAAAATTGGTAGATGCAGCGATTGAACTGAATGCAGATGCAATACTTGCATCTACAATAATAAGTCATGATGATATTCACTATAAAAACATGAAAAAATTACACGATTATTGTGTAGAAAAAGGTATAAGAGATAAAGTAATGATTGCGTGTGGAGGAACACAAGTAACACCTGAAATAGCTGTTGAACAAGGCATCGATGCTGGATTTGGAAGAAACAGCAAAGGCATACATGTAGCCACTTTCTTAGTTGAAAAAAGAAGAGAAATGAGTAATAAATAA'\n",
    "\n",
    "cpd24394_c = cobra.Metabolite(\n",
    "    'cpd24394_c',\n",
    "    formula='C5H9NO3',\n",
    "    name='2-amino-4-ketopentanoate',\n",
    "    compartment='cytosol')\n",
    "cpd24394_c.annotation['sbo'] = 'SBO:0000247'\n",
    "cpd24394_c.annotation['seed.compound'] = 'cpd02448'\n",
    "cpd24394_c.annotation['metacyc'] = 'CPD-299'\n",
    "\n",
    "cpd00006_c = iCdG698.metabolites.cpd00006_c # NADP\n",
    "cpd00005_c = iCdG698.metabolites.cpd00005_c # NADPH\n",
    "\n",
    "# 2,4-diaminopentanoate <--> 2-amino-4-ketopentanoate\n",
    "rxn27496_c = cobra.Reaction('rxn27496_c')\n",
    "rxn27496_c.name = '2,4-diaminopentanoate dehydrogenase'\n",
    "rxn27496_c.lower_bound = -1000.\n",
    "rxn27496_c.upper_bound = 1000.\n",
    "rxn27496_c.add_metabolites({\n",
    "    cpd00001_c: -1.0,\n",
    "    cpd02448_c: -1.0,\n",
    "    cpd00006_c: -1.0,\n",
    "    cpd00013_c: 1.0,\n",
    "    cpd00067_c: 1.0,\n",
    "    cpd24394_c: 1.0,\n",
    "    cpd00005_c: 1.0\n",
    "})\n",
    "rxn27496_c.gene_reaction_rule = '272563.120.peg.467'\n",
    "iCdG698.add_reactions([rxn27496_c])\n",
    "iCdG698.reactions.rxn01905_c.gene_reaction_rule = ''\n",
    "iCdG698.reactions.rxn27496_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdG698.reactions.rxn27496_c.annotation['seed.reaction'] = 'rxn27496'\n",
    "\n",
    "gene2 = cobra.Gene('272563.120.peg.468')\n",
    "gene3 = cobra.Gene('645463.3.peg.414')\n",
    "\n",
    "# 2-amino-4-ketopentanoate <--> alanine\n",
    "rxn19080_c = cobra.Reaction('rxn19080_c')\n",
    "rxn19080_c.name = '2-amino-4-oxopentanoate thiolase'\n",
    "rxn19080_c.lower_bound = -1000.\n",
    "rxn19080_c.upper_bound = 1000.\n",
    "rxn19080_c.add_metabolites({\n",
    "    cpd00010_c: -1.0,\n",
    "    cpd24394_c: -1.0,\n",
    "    cpd00022_c: 1.0,\n",
    "    cpd00035_c: 1.0\n",
    "})\n",
    "rxn19080_c.gene_reaction_rule = '272563.120.peg.468 and 272563.120.peg.469'\n",
    "iCdG698.add_reactions([rxn19080_c])\n",
    "iCdG698.reactions.rxn19080_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdG698.reactions.rxn19080_c.annotation['seed.reaction'] = 'rxn19080'\n",
    "\n",
    "iCdG698.genes.get_by_id('272563.120.peg.468').name = '2-amino-4-ketopentanoate thiolase alpha subunit [EC:2.3.1.263] | ortA'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.468').annotation['sbo'] = 'SBO:0000243'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.468').annotation['patric'] = 'fig|272563.120.peg.468'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.468').annotation['kegg'] = 'K21399'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.468').annotation['refseq'] = 'AJP10139.1'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.468').annotation['refseq.locustag'] = 'CDIF630_00572'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.468').annotation['ccds'] = 'ATGGATGCTAAAATTAATGATTGGGTTATTATACACAATATAGTATTGACTCCAGAAGAAAGAGCTCCACAAGTTCCAGAGGATACTAAGAAAGTATCCTTAGAAATGTGGGTTAAAGGTTTTATACAAAGTGATGCTTCTATAGGCGACTTAGTAGAAGTTAAAACAATTACAGGTAGACTTGTAAAAGGTAATTTACTTAAAGTAAATCCATATTATACTCATGATTATGGTAAATGTATACCAGAGTTACTCCAAATAGGAATCCAAGCTAAAGAAATATTATTTGGAGGTGTATATAATGAATAA'\n",
    "\n",
    "iCdG698.genes.get_by_id('272563.120.peg.469').name = '2-amino-4-ketopentanoate thiolase beta subunit [EC:2.3.1.263] | ortA'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.469').annotation['sbo'] = 'SBO:0000243'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.469').annotation['patric'] = 'fig|272563.120.peg.469'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.469').annotation['kegg'] = 'K21400'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.469').annotation['refseq'] = 'AJP10139.1'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.469').annotation['refseq.locustag'] = 'CDIF630_00572'\n",
    "iCdG698.genes.get_by_id('272563.120.peg.469').annotation['ccds'] = 'ATGAATAATTCAACATCAATAAAAGATATGAGTTATCAAGCTGTTATGGGAAGAAATAACGAAATAATGAAAAATGCAATCGGATTGGATTATTCTTCTTTTGAACAAGAGGGAATAGGGTTTGATTATGAGAAAATGATGAGTGAAACAGGATATACATTACAAGATATTGAAGCAATTCAATCACAGTATGCTGTTGGAAATACTCCTCTAATAGAGCTTAAAAATTTGACTAAATTAGCTAGAAAATGTGCTAAAGAAGGTAAAGGAGCTAAAATATTTGTAAAAGATGAAGCAATGAATGCATCTGGTAGTTTTAAAGCAAGAAGAGCAGCTACTGCAGTATATCATGCCAAACAAATGGGATATAAAGGTGTAATAGCAGCTACCAGTGGTAACTATGGTGCAGCTGTAGCATCACAAGCGGCAATGCAGGGATTAAAATGTATAATAGTTCAAGAATGTTATGATTCTAATGGTGTAGGTCAACCTGAAATTATAGAAAAAGCACGAAAATGTGAGGCATATGGAGCAGAAGTTGTTCAGTTAACTGTTGGACCAGAGTTATTTTATACGTTCTTAGTTTTATTAGAAGAAACAGGCTATTTTAACGCATCATTATATTCACCATTTGGAATTGCAGGTGTAGAAACTTTAGGTTATGAGTTAGCTATCCAATTTAGAGAAAAGTATAAAAAAGACCCTGATATAGTTGTTTGTACAAATGCAGGTGGTGGTAATCTTACTGGAACTGCAAGAGGATTGATAAAAGCTGGTTCCATTAATACTAAAGTAGTAGGAGCAAGTGTAGATTTAAAAGGATTACACATGGCTAGTGATAATCAATTTAATAAAAAATCATTTACAACAGGACATACAGGTTTTGGGATACCATACTGTACATGGCCTGATAGGTCAGATGTACCTAGGTCTGCAGCTAGACCACTTAGATATATGGATAGATATGTATTAGTAAAACAAGGAGAAGTTTTTTATACAACTGAATTATTGGCACAGCTAGAAGGTATAGAAAGAGGTCCAGCAGGAAATACATCTTTAGCAGCAGCATTCTCTTTAGCTCAAGAGTTAGATGAGGATAAAGTTATAGTGGTACAAGAAACAGAATATACAGGCGCAGGTAAACATATATGTCCACAGCTTACTTTTGCAAGAGAAAATGGAATAGATATAAAATTTGGAAATCCAAGAGAAGAAATTGCAGGAGTAAACTTAATATTACCAGAAAGACCAGAATTATTGAAATGTGTTGATATTGATATGAATAAAATAAGAAAGTCATTTATAAAAAATTGCATATCTAATAATCACATTGATAATATTGATAAGCTAAGTAATCGAGATATAGAATTTTTAATGGAAGAAGTAAAAAGTTCAAGAGATTTTGTTGTTGATGTATTAAATAATTTATAG'\n",
    "\n",
    "#------------------------------------------------------------------------------#\n",
    "\n",
    "# Correct p-Cresol metabolism\n",
    "iCdG698.reactions.rxn07124_c.remove_from_model(remove_orphans=True)\n",
    "iCdG698.reactions.ID008_c.remove_from_model(remove_orphans=True)\n",
    "iCdG698.reactions.EX_cpd03170_e.remove_from_model(remove_orphans=True)\n",
    "iCdG698.reactions.ID004_c.remove_from_model(remove_orphans=True)\n",
    "\n",
    "iCdG698.reactions.rxn05061_c.gene_reaction_rule = '272563.120.peg.166 and 272563.120.peg.167 and 272563.120.peg.168'\n",
    "\n",
    "iCdG698.reactions.ID005_c.id = 'rxn29230_c'\n",
    "iCdG698.reactions.rxn29230_c.annotation['seed.reaction'] = 'rxn29230'\n",
    "iCdG698.reactions.ID009_c.id = 'rxn08184_c'\n",
    "iCdG698.reactions.rxn08184_c.annotation['seed.reaction'] = 'rxn08184'\n",
    "\n",
    "# Correct phenylpropanoate transport\n",
    "iCdG698.reactions.ID007_c.remove_from_model(remove_orphans=False)\n",
    "cpd03343_c = iCdG698.metabolites.cpd03343_c # Phenylpropanoate\n",
    "cpd03343_e = iCdG698.metabolites.cpd03343_e # Phenylpropanoate\n",
    "\n",
    "rxn10133_c = cobra.Reaction('rxn10133_c')\n",
    "rxn10133_c.name = 'Phenylpropanoate hydrogen symport'\n",
    "rxn10133_c.lower_bound = -1000.\n",
    "rxn10133_c.upper_bound = 1000.\n",
    "rxn10133_c.add_metabolites({\n",
    "    cpd00067_c: -1.0,\n",
    "    cpd03343_c: -1.0,\n",
    "    cpd00067_c: 1.0,\n",
    "    cpd03343_e: 1.0\n",
    "})\n",
    "iCdG698.add_reactions([rxn10133_c])\n",
    "iCdG698.reactions.rxn10133_c.annotation['sbo'] = 'SBO:0000629'\n",
    "iCdG698.reactions.rxn10133_c.annotation['seed.reaction'] = 'rxn10133'\n",
    "iCdG698.reactions.rxn10133_c.annotation['bigg.reaction'] = ['PPPNt2','PPPNt2r','PPPNt2rpp']\n",
    "\n",
    "\n",
    "cobra.io.save_json_model(iCdG698, '/home/mjenior/Desktop/repos/Jenior_Cdifficile_2019/data/reconstructions/iCdG698_v1.1.sbml')\n",
    "cobra.io.write_sbml_model(iCdG698, '/home/mjenior/Desktop/repos/Jenior_Cdifficile_2019/data/reconstructions/iCdG698_v1.1.json')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "703"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(iCdR700.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BHI\n",
    "bhi = ['cpd00001_e', # water\n",
    "'cpd00035_e', # alanine\n",
    "'cpd00041_e', # aspartic acid\n",
    "'cpd00023_e', # glutamic acid\n",
    "'cpd00119_e', # histidine\n",
    "'cpd00107_e', # leucine\n",
    "'cpd00060_e', # methionine\n",
    "'cpd00161_e', # threonine\n",
    "'cpd00069_e', # tyrosine\n",
    "'cpd00084_e', # cysteine\n",
    "'cpd00033_e', # glycine\n",
    "'cpd00322_e', # isoleucine\n",
    "'cpd00066_e', # phenylalanine\n",
    "'cpd00054_e', # serine\n",
    "'cpd00065_e', # tryptophan\n",
    "'cpd00156_e', # valine\n",
    "'cpd00220_e', # riboflavin\n",
    "'cpd00644_e', # pantothentate(B5)\n",
    "'cpd00393_e', # folate(B9)\n",
    "'cpd00133_e', # niacin(PP)\n",
    "'cpd00263_e', # Pyridoxine\n",
    "'cpd00104_e', # Biotin\n",
    "'cpd00149_e', # Cobalt\n",
    "'cpd00971_e', # sodium\n",
    "'cpd00099_e', # Chloride\n",
    "'cpd00205_e', # potassium\n",
    "'cpd00009_e', # phosphate\n",
    "'cpd00063_e', # calcium\n",
    "'cpd00254_e', # magnesium\n",
    "'cpd10515_e', # Fe2+\n",
    "'cpd00030_e', # Mn2+\n",
    "'cpd00242_e', # Carbonate\n",
    "'cpd00226_e', # hypoxanthine\n",
    "'cpd01242_e', # thyminose\n",
    "'cpd00307_e', # cytosine\n",
    "'cpd00092_e', # uracil\n",
    "'cpd00117_e', # D-Alanine\n",
    "'cpd00067_e', # H+\n",
    "'cpd00567_e', # D-Proline\n",
    "'cpd00132_e', # L-Asparagine\n",
    "'cpd00210_e', # Taurine\n",
    "'cpd00320_e', # D-Aspartate\n",
    "'cpd03279_e', # Deoxyinosine\n",
    "'cpd00246_e', # Inosine\n",
    "'cpd00311_e', # Guanosine\n",
    "'cpd00367_e', # Cytidine\n",
    "'cpd00277_e', # Deoxyguanosine\n",
    "'cpd00182_e', # Adenosine\n",
    "'cpd00654_e', # Deoxycytidine\n",
    "'cpd00412_e', # Deoxyuridine\n",
    "'cpd00438_e', # Deoxyadenosine\n",
    "'cpd00274_e', # Citrulline\n",
    "'cpd00186_e', # D-Glutamate\n",
    "'cpd00637_e', # D-Methionine\n",
    "'cpd00105_e', # D-Ribose\n",
    "'cpd00305_e', # Thiamin\n",
    "'cpd00309_e', # Xanthine\n",
    "'cpd00098_e', # Choline\n",
    "'cpd00207_e'] # Guanine\n",
    "\n",
    "bhi_exch = set(['EX_' + x for x in bhi])\n",
    "for rxn in iCdR700.exchanges:\n",
    "    rxn.lower_bound = 0.\n",
    "    if rxn.id in bhi_exch:\n",
    "        rxn.lower_bound = -1000.\n",
    "        \n",
    "iCdR700.reactions.EX_cpd00053_e.lower_bound = -1000. # L-Glutamine\n",
    "iCdR700.reactions.EX_cpd00039_e.lower_bound = -1000. # Lysine\n",
    "iCdR700.reactions.EX_cpd00079_e.lower_bound = -1000. # D-glucose-6-phosphate\n",
    "iCdR700.reactions.EX_cpd00064_e.lower_bound = -1000. # Ornithine\n",
    "iCdR700.reactions.EX_cpd00051_e.lower_bound = -1000. # D-arginine\n",
    "\n",
    "base_essential_bhi = find_essential_genes(iCdR700)\n",
    "base_essential_bhi = set([x.id for x in base_essential_bhi])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Initializing model and integrating transcriptomic data...\n",
      "Pruning zero flux subnetworks...\n",
      "Analyzing context-specific flux distributions...\n",
      "\n",
      "Reactions pruned to 306 from 1317 (76.77% change)\n",
      "Metabolites pruned to 306 from 1245 (75.42% change)\n",
      "Flux through the objective DECREASED to ~95.0 from ~124.92 (23.95% change)\n",
      "Context-specific metabolism correlates with transcriptome (r=0.21, p<0.001 *)\n",
      "\n",
      "RIPTiDe completed in 17 seconds\n",
      "\n",
      "\n",
      "Initializing model and integrating transcriptomic data...\n",
      "Pruning zero flux subnetworks...\n",
      "Analyzing context-specific flux distributions...\n",
      "\n",
      "Reactions pruned to 309 from 1317 (76.54% change)\n",
      "Metabolites pruned to 309 from 1245 (75.18% change)\n",
      "Flux through the objective DECREASED to ~97.92 from ~124.92 (21.61% change)\n",
      "Context-specific metabolism correlates with transcriptome (r=0.128, p=0.024 *)\n",
      "\n",
      "RIPTiDe completed in 19 seconds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Solo\n",
    "iCdR700_solo = riptide.contextualize(model=iCdR700, transcriptome=solo, fraction=0.75)\n",
    "\n",
    "# Coculture\n",
    "iCdR700_coculture = riptide.contextualize(model=iCdR700, transcriptome=coculture, fraction=0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "riptide.save_output(riptide_obj=iCdR700_solo, path='/home/mjenior/Desktop/active_projects/Zackular_enterococcus/solo_bhi_new')\n",
    "riptide.save_output(riptide_obj=iCdR700_coculture, path='/home/mjenior/Desktop/active_projects/Zackular_enterococcus/coculture_bhi_new')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCdR700_riptide</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07fcd4ca7d290</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>306</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>306</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>0.0 + 1.0*biomass - 1.0*biomass_reverse_01e59</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCdR700_riptide at 0x7fcd4ca7d290>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iCdR700_solo.model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "220"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(iCdR700_solo.model.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iCdR700_riptide</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07fcd31b8abd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>309</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>309</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>0.0 + 1.0*biomass - 1.0*biomass_reverse_01e59</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iCdR700_riptide at 0x7fcd31b8abd0>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iCdR700_coculture.model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "218"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(iCdR700_coculture.model.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo doubling time: 33.68 minutes\n",
      "Coculture doubling time: 32.68 minutes\n",
      "p-value 0.285\n"
     ]
    }
   ],
   "source": [
    "from random import sample\n",
    "import scipy\n",
    "\n",
    "doubling = round((1.0 / iCdR700_solo.model.slim_optimize()) * 3200.0, 2)\n",
    "print('Solo doubling time: ' + str(doubling) + ' minutes')\n",
    "\n",
    "doubling = round((1.0 / iCdR700_coculture.model.slim_optimize()) * 3200.0, 2)\n",
    "print('Coculture doubling time: ' + str(doubling) + ' minutes')\n",
    "\n",
    "pvals = []\n",
    "maxRange = len(list(iCdR700_solo.flux_samples['biomass'])) * 10\n",
    "for x in range(1, maxRange):\n",
    "        solo_sample = sample(list(iCdR700_solo.flux_samples['biomass']), 10)\n",
    "        coculture_sample = sample(list(iCdR700_coculture.flux_samples['biomass']), 10)\n",
    "        pvals.append(scipy.stats.wilcoxon(x=solo_sample, y=coculture_sample)[1])\n",
    "print('p-value', round(numpy.median(pvals),3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo reactions: 15\n",
      "EX_cpd00023_e L-Glutamate exchange\n",
      "rxn05204_c Guanosine hydrogen symport\n",
      "rxn01286_c D-Glyceraldehyde:NAD+ oxidoreductase\n",
      "rxn05298_c Na+/glutamate symport\n",
      "rxn08647_c ATP:(R)-glycerate 2-phosphotransferase\n",
      "EX_cpd00013_e Ammonia exchange\n",
      "rxn00459_c 2-phospho-D-glycerate hydro-lyase (phosphoenolpyruvate-forming)\n",
      "rxn05234_c 2'-Deoxyguanosine 5'-triphosphate:oxydized-thioredoxin 2'-oxidoreductase\n",
      "rxn02380_c beta-D-Glucose 6-phosphate ketol-isomerase\n",
      "EX_cpd00001_e H2O exchange\n",
      "rxn08686_c H2O transport via diffusion\n",
      "rxn05467_c CO2 transport via diffusion\n",
      "rxn05466_c Ammonia diffusion\n",
      "EX_cpd00011_e CO2 exchange\n",
      "EX_cpd00311_e Guanosine exchange\n",
      "\n",
      "Coculture-only reactions: 18\n",
      "EX_cpd00186_e D-Glutamate exchange\n",
      "EX_cpd00067_e H+ exchange\n",
      "rxn00504_c (S)-1-pyrroline-5-carboxylate:NADP+ oxidoreductase\n",
      "EX_cpd00277_e Deoxyguanosine exchange\n",
      "rxn01649_c Xanthosine:orthophosphate ribosyltransferase\n",
      "rxn05203_c Guanine symport\n",
      "rxn00931_c L-Proline:NADP+ 5-oxidoreductase\n",
      "rxn01297_c hypoxanthine:NAD+ oxidoreductase\n",
      "rxn01548_c guanosine:phosphate alpha-D-ribosyltransferase\n",
      "rxn13140_c Nucleoside triphosphatase\n",
      "rxn09688_c deoxyguanosine transport in via proton symport (periplasm)\n",
      "rxn01547_c Guanosine aminohydrolase\n",
      "rxn05703_c Hydroxylamine transport in via proton symport\n",
      "rxn00558_c D-glucose-6-phosphate aldose-ketose-isomerase\n",
      "rxn00109_c ammonia:NAD+ oxidoreductase\n",
      "rxn05672_c D-glutamate transport in via proton symport\n",
      "EX_cpd00165_e Hydroxylamine exchange\n",
      "EX_cpd00207_e Guanine exchange\n"
     ]
    }
   ],
   "source": [
    "# Distinct reactions - BHI\n",
    "rxns1 = set([x.id for x in iCdR700_solo.model.reactions])\n",
    "rxns2 = set([x.id for x in iCdR700_coculture.model.reactions])\n",
    "\n",
    "solo_only_rxns = rxns1.difference(rxns2)\n",
    "print('Solo reactions:', len(solo_only_rxns))\n",
    "for x in solo_only_rxns: print(x, iCdR700.reactions.get_by_id(x).name)\n",
    "\n",
    "coculture_only_rxns = rxns2.difference(rxns1)\n",
    "print('\\nCoculture-only reactions:', len(coculture_only_rxns))\n",
    "for x in coculture_only_rxns: print(x, iCdR700.reactions.get_by_id(x).name)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shared exchange reactions: 46\n",
      "EX_cpd19585_e 2-Methylbutyrate exchange\n",
      "EX_cpd00063_e Ca2+ exchange\n",
      "EX_cpd00039_e L-Lysine exchange\n",
      "EX_cpd00107_e L-Leucine exchange\n",
      "EX_cpd00133_e Nicotinamide exchange\n",
      "EX_cpd00149_e Co2+ exchange\n",
      "EX_cpd00035_e L-Alanine exchange\n",
      "EX_cpd00069_e L-Tyrosine exchange\n",
      "EX_cpd00412_e Deoxyuridine exchange\n",
      "EX_cpd00971_e Na+ exchange\n",
      "EX_cpd00030_e Mn2+ exchange\n",
      "EX_cpd00104_e Biotin exchange\n",
      "EX_cpd00254_e Mg exchange\n",
      "EX_cpd00117_e D-Alanine exchange\n",
      "EX_cpd00065_e L-Tryptophan exchange\n",
      "EX_cpd00644_e Pantothenate exchange\n",
      "EX_cpd00066_e L-Phenylalanine exchange\n",
      "EX_cpd00161_e L-Threonine exchange\n",
      "EX_cpd00099_e Cl- exchange\n",
      "EX_cpd00054_e L-Serine exchange\n",
      "EX_cpd00654_e Deoxycytidine exchange\n",
      "EX_cpd10515_e Fe2+ exchange\n",
      "EX_cpd00246_e Inosine exchange\n",
      "EX_cpd00263_e Pyridoxol exchange\n",
      "EX_cpd00041_e L-Aspartate exchange\n",
      "EX_cpd00060_e L-Methionine exchange\n",
      "EX_cpd00393_e Folate exchange\n",
      "EX_cpd00009_e Phosphate exchange\n",
      "EX_cpd00092_e Uracil exchange\n",
      "EX_cpd00220_e Riboflavin exchange\n",
      "EX_cpd00322_e L-Isoleucine exchange\n",
      "EX_cpd00119_e L-Histidine exchange\n",
      "EX_cpd00205_e K+ exchange\n",
      "EX_cpd03279_e Deoxyinosine exchange\n",
      "EX_cpd05178_e Isovalerate exchange\n",
      "EX_cpd00226_e Hypoxanthine exchange\n",
      "EX_cpd00047_e Formate exchange\n",
      "EX_cpd00051_e L-Arginine exchange\n",
      "EX_cpd00132_e L-Asparagine exchange\n",
      "EX_cpd00029_e Acetate exchange\n",
      "EX_cpd00156_e L-Valine exchange\n",
      "EX_cpd00079_e D-glucose-6-phosphate exchange\n",
      "EX_cpd00305_e Thiamin exchange\n",
      "EX_cpd00084_e L-Cysteine exchange\n",
      "EX_cpd00064_e Ornithine exchange\n",
      "EX_cpd00122_e N-Acetyl-D-glucosamine exchange\n"
     ]
    }
   ],
   "source": [
    "# Shared exchanges\n",
    "rxns1 = set([x.id for x in iCdR700_solo.model.exchanges])\n",
    "rxns2 = set([x.id for x in iCdR700_coculture.model.exchanges])\n",
    "shared_exch = rxns1.intersection(rxns2)\n",
    "\n",
    "print('Shared exchange reactions:', len(shared_exch))\n",
    "for x in shared_exch: print(x, iCdR700.reactions.get_by_id(x).name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo culture-only metabolites: 7\n",
      "cpd00011_e CO2\n",
      "cpd00311_e Guanosine\n",
      "cpd00013_e Ammonia\n",
      "cpd00001_e H2O\n",
      "cpd00482_c 2-Phospho-D-glycerate\n",
      "cpd00223_c Glycerate\n",
      "cpd00023_e L-Glutamate\n",
      "\n",
      "Coculture-only metabolites: 10\n",
      "cpd00277_c Deoxyguanosine\n",
      "cpd00277_e Deoxyguanosine\n",
      "cpd00165_c Hydroxylamine\n",
      "cpd00207_c Guanine\n",
      "cpd00165_e Hydroxylamine\n",
      "cpd00309_c Xanthine\n",
      "cpd00207_e Guanine\n",
      "cpd02431_c 1-Pyrroline-5-carboxylate\n",
      "cpd00186_e D-Glutamate\n",
      "cpd01217_c Xanthosine\n"
     ]
    }
   ],
   "source": [
    "# Distinct metabolites - BHI\n",
    "cpds1 = set([x.id for x in iCdR700_solo.model.metabolites])\n",
    "cpds2 = set([x.id for x in iCdR700_coculture.model.metabolites])\n",
    "\n",
    "solo_only_cpds = cpds1.difference(cpds2)\n",
    "print('Solo culture-only metabolites:', len(solo_only_cpds))\n",
    "for x in solo_only_cpds: print(x, iCdR700.metabolites.get_by_id(x).name)\n",
    "\n",
    "coculture_only_cpds = cpds2.difference(cpds1)\n",
    "print('\\nCoculture-only metabolites:', len(coculture_only_cpds))\n",
    "for x in coculture_only_cpds: print(x, iCdR700.metabolites.get_by_id(x).name) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo-only genes: 6\n",
      "645463.3.peg.2226 Proton/sodium-glutamate symport protein\n",
      "645463.3.peg.1749 Sodium/glutamate symporter\n",
      "645463.3.peg.3215 Enolase (EC 4.2.1.11)\n",
      "645463.3.peg.2261 Aldehyde dehydrogenase (EC 1.2.1.3)\n",
      "645463.3.peg.1272 pyridoxal kinase (EC 2.7.1.35)\n",
      "645463.3.peg.2741 Sodium/glutamate symporter\n",
      "\n",
      "Coculture-only genes: 4\n",
      "645463.3.peg.1483 phosphoribosyl-AMP cyclohydrolase (EC 3.5.4.19)\n",
      "645463.3.peg.1423 Pyrroline-5-carboxylate reductase (EC 1.5.1.2)\n",
      "645463.3.peg.3335 Pyrroline-5-carboxylate reductase (EC 1.5.1.2)\n",
      "645463.3.peg.2222 hydroxylamine reductase (EC 1.7.99.1)\n"
     ]
    }
   ],
   "source": [
    "# Distinct genes - BHI\n",
    "genes1 = set([x.id for x in iCdR700_solo.model.genes])\n",
    "genes2 = set([x.id for x in iCdR700_coculture.model.genes])\n",
    "\n",
    "solo_only_genes = genes1.difference(genes2)\n",
    "print('Solo-only genes:', len(solo_only_genes))\n",
    "for x in solo_only_genes: print(x, iCdR700.genes.get_by_id(x).name)\n",
    "\n",
    "coculture_only_genes = genes2.difference(genes1)\n",
    "print('\\nCoculture-only genes:', len(coculture_only_genes))\n",
    "for x in coculture_only_genes: print(x, iCdR700.genes.get_by_id(x).name) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo-only essential genes: 2\n",
      "645463.3.peg.1079 Acetate kinase (EC 2.7.2.1)\n",
      "645463.3.peg.2376 4-hydroxybutyrate:acetyl-CoA CoA transferase (EC 2.3.1.-)\n",
      "\n",
      "Coculture-only essential genes: 2\n",
      "645463.3.peg.3217 Triosephosphate isomerase (EC 5.3.1.1)\n",
      "645463.3.peg.194 NAD-specific glutamate dehydrogenase (EC 1.4.1.2)\n"
     ]
    }
   ],
   "source": [
    "# Differential gene essentiality\n",
    "solo_threshold = iCdR700_solo.model.slim_optimize() * 0.05\n",
    "solo_essential_bhi = find_essential_genes(iCdR700_solo.model, threshold=solo_threshold)\n",
    "solo_essential_bhi = set([x.id for x in solo_essential_bhi]).difference(base_essential_bhi)\n",
    "\n",
    "coculture_threshold = iCdR700_coculture.model.slim_optimize() * 0.05\n",
    "coculture_essential_bhi = find_essential_genes(iCdR700_coculture.model, threshold=coculture_threshold)\n",
    "coculture_essential_bhi = set([x.id for x in coculture_essential_bhi]).difference(base_essential_bhi)\n",
    "\n",
    "solo_essential_bhi_only = solo_essential_bhi.difference(coculture_essential_bhi)\n",
    "coculture_essential_bhi_only = coculture_essential_bhi.difference(solo_essential_bhi)\n",
    "essential_bhi = solo_essential_bhi.intersection(coculture_essential_bhi)\n",
    "\n",
    "both_contexts = solo_essential_bhi.intersection(coculture_essential_bhi)\n",
    "\n",
    "print('Solo-only essential genes:', len(solo_essential_bhi_only))\n",
    "for x in solo_essential_bhi_only: print(x, iCdR700.genes.get_by_id(x).name)\n",
    "print('\\nCoculture-only essential genes:', len(coculture_essential_bhi_only))\n",
    "for x in coculture_essential_bhi_only: print(x, iCdR700.genes.get_by_id(x).name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Core essential: 72\n",
      "645463.3.peg.48 2-C-methyl-D-erythritol 4-phosphate cytidylyltransferase (EC 2.7.7.60)\n",
      "645463.3.peg.2106 Glutaminyl-tRNA synthetase (EC 6.1.1.18)\n",
      "645463.3.peg.1828 4-hydroxy-3-methylbut-2-enyl diphosphate reductase (EC 1.17.1.2)\n",
      "645463.3.peg.57 thymidylate synthase (EC 2.1.1.45)\n",
      "645463.3.peg.682 Phenylalanyl-tRNA synthetase beta chain (EC 6.1.1.20)\n",
      "645463.3.peg.2802 Phosphoenolpyruvate-protein phosphotransferase of PTS system (EC 2.7.3.9) | ptsI\n",
      "645463.3.peg.3143 PTS system, trehalose-specific IIB component (EC 2.7.1.69)\n",
      "645463.3.peg.2479 Glycyl-tRNA synthetase beta chain (EC 6.1.1.14)\n",
      "645463.3.peg.1842 3-dehydroquinate synthase (EC 4.2.3.4)\n",
      "645463.3.peg.2042 Substrate-specific component TrpP of tryptophan ECF transporter\n",
      "645463.3.peg.2701 UDP-N-acetylmuramoylalanine--D-glutamate ligase (EC 6.3.2.9)\n",
      "645463.3.peg.2703 UDP-N-acetylmuramoyl-tripeptide--D-alanyl-D-alanine ligase\n",
      "645463.3.peg.3609 4-diphosphocytidyl-2-C-methyl-D-erythritol kinase (EC 2.7.1.148)\n",
      "645463.3.peg.2480 Glycyl-tRNA synthetase alpha chain (EC 6.1.1.14)\n",
      "645463.3.peg.1374 aminodeoxychorismate lyase (EC 4.1.3.38)\n",
      "645463.3.peg.360 Manganese-dependent inorganic pyrophosphatase (EC 3.6.1.1)\n",
      "645463.3.peg.2569 Leucyl-tRNA synthetase (EC 6.1.1.4)\n",
      "645463.3.peg.1245 CDP-diacylglycerol--glycerol-3-phosphate 3-phosphatidyltransferase (EC 2.7.8.5)\n",
      "645463.3.peg.398 Fructose-bisphosphate aldolase class II (EC 4.1.2.13)\n",
      "645463.3.peg.1257 alanine transaminase (EC 2.6.1.2)\n",
      "645463.3.peg.120 Ribonucleotide reductase of class III (anaerobic), large subunit (EC 1.17.4.2)\n",
      "645463.3.peg.954 UDP-N-acetylglucosamine 2-epimerase (EC 5.1.3.14)\n",
      "645463.3.peg.1112 1-deoxy-D-xylulose 5-phosphate synthase (EC 2.2.1.7)\n",
      "645463.3.peg.3307 Valyl-tRNA synthetase (EC 6.1.1.9)\n",
      "645463.3.peg.2633 phosphopantothenoylcysteine decarboxylase (EC 4.1.1.36)\n",
      "645463.3.peg.3591 Thymidylate kinase (EC 2.7.4.9)\n",
      "645463.3.peg.2293 Asparaginyl-tRNA synthetase (EC 6.1.1.22)\n",
      "645463.3.peg.530 Threonyl-tRNA synthetase (EC 6.1.1.3)\n",
      "645463.3.peg.2653 Tryptophanyl-tRNA synthetase (EC 6.1.1.2)\n",
      "645463.3.peg.1176 Na(+)-dependent branched-chain amino acid transporter\n",
      "645463.3.peg.2702 Phospho-N-acetylmuramoyl-pentapeptide-transferase (EC 2.7.8.13)\n",
      "645463.3.peg.2662 Isoleucyl-tRNA synthetase (EC 6.1.1.5)\n",
      "645463.3.peg.2971 Substrate-specific component BioY of biotin ECF transporter\n",
      "645463.3.peg.2243 Maltose phosphorylase (EC 2.4.1.8)\n",
      "645463.3.peg.1792 argininosuccinate synthase (EC 6.3.4.5)\n",
      "645463.3.peg.3598 Pantothenate kinase type III, CoaX-like (EC 2.7.1.33)\n",
      "645463.3.peg.1110 (2E,6E)-farnesyl diphosphate synthase (EC 2.5.1.10)\n",
      "645463.3.peg.2774 diglucosyldiacylglycerol synthase (LTA membrane anchor synthesis)\n",
      "645463.3.peg.2273 3-dehydroquinate dehydratase I (EC 4.2.1.10)\n",
      "645463.3.peg.3421 UDP-N-acetylenolpyruvoylglucosamine reductase (EC 1.1.1.158)\n",
      "645463.3.peg.2602 Phosphopantetheine adenylyltransferase (EC 2.7.7.3)\n",
      "645463.3.peg.576 Chloride channel protein\n",
      "645463.3.peg.3276 aspartate-semialdehyde dehydrogenase (EC 1.2.1.11)\n",
      "645463.3.peg.222 PTS system, fructose-specific IIC component (EC 2.7.1.69)\n",
      "645463.3.peg.136 UDP-N-acetylglucosamine 1-carboxyvinyltransferase (EC 2.5.1.7)\n",
      "645463.3.peg.735 ABC transporter, substrate-binding protein\n",
      "645463.3.peg.3595 Lysyl-tRNA synthetase (class II) (EC 6.1.1.6)\n",
      "645463.3.peg.161 Substrate-specific component RibU of riboflavin ECF transporter\n",
      "645463.3.peg.1840 Magnesium and cobalt transport protein CorA\n",
      "645463.3.peg.693 Arginyl-tRNA synthetase (EC 6.1.1.19)\n",
      "645463.3.peg.223 PTS system, fructose-specific IIB component (EC 2.7.1.69)\n",
      "645463.3.peg.2789 Histidyl-tRNA synthetase (EC 6.1.1.21)\n",
      "645463.3.peg.1233 Riboflavin kinase (EC 2.7.1.26)\n",
      "645463.3.peg.1844 Chorismate synthase (EC 4.2.3.5)\n",
      "645463.3.peg.3013 alpha-mannosidase (EC 3.2.1.24)\n",
      "645463.3.peg.971 NAD kinase (EC 2.7.1.23)\n",
      "645463.3.peg.2187 Phosphatidate cytidylyltransferase (EC 2.7.7.41)\n",
      "645463.3.peg.1847 Shikimate kinase I (EC 2.7.1.71)\n",
      "645463.3.peg.1032 Dephospho-CoA kinase (EC 2.7.1.24)\n",
      "645463.3.peg.2788 Aspartyl-tRNA synthetase (EC 6.1.1.12)\n",
      "645463.3.peg.49 2-C-methyl-D-erythritol 2,4-cyclodiphosphate synthase (EC 4.6.1.12)\n",
      "645463.3.peg.1451 Tyrosyl-tRNA synthetase (EC 6.1.1.1)\n",
      "645463.3.peg.54 Cysteinyl-tRNA synthetase (EC 6.1.1.16)\n",
      "645463.3.peg.681 Phenylalanyl-tRNA synthetase alpha chain (EC 6.1.1.20)\n",
      "645463.3.peg.1199 Alanyl-tRNA synthetase (EC 6.1.1.7)\n",
      "645463.3.peg.50 Prolyl-tRNA synthetase (EC 6.1.1.15)\n",
      "645463.3.peg.2699 UDP-N-acetylglucosamine--N-acetylmuramyl-(pentapeptide) pyrophosphoryl-undecaprenol N-acetylglucosamine transferase (EC 2.4.1.227)\n",
      "645463.3.peg.52 Glutamyl-tRNA synthetase (EC 6.1.1.17)\n",
      "645463.3.peg.3495 5-formyltetrahydrofolate cyclo-ligase (EC 6.3.3.2)\n",
      "645463.3.peg.2183 1-deoxy-D-xylulose 5-phosphate reductoisomerase (EC 1.1.1.267)\n",
      "645463.3.peg.770 Multimodular transpeptidase-transglycosylase (EC 2.4.1.129) (EC 3.4.-.-)\n",
      "645463.3.peg.3581 Methionyl-tRNA synthetase (EC 6.1.1.10)\n"
     ]
    }
   ],
   "source": [
    "print('\\nCore essential:', len(base_essential_bhi))\n",
    "for x in base_essential_bhi: print(x, iCdR700.genes.get_by_id(x).name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo only 15\n",
      "EX_cpd00023_e L-Glutamate exchange\n",
      "rxn13713_c 4-hydroxybutyrate CoA-transferase\n",
      "rxn05204_c Guanosine hydrogen symport\n",
      "rxn01236_c ATP:butanoate 1-phosphotransferase\n",
      "rxn05298_c Na+/glutamate symport\n",
      "rxn00225_c ATP:acetate phosphotransferase\n",
      "rxn03641_c acetyl-CoA: 4-hydroxybutanoate CoA transferase\n",
      "EX_cpd00029_e Acetate exchange\n",
      "rxn01345_c Sedoheptulose 1,7-bisphosphate 1-phosphohydrolase\n",
      "rxn05234_c 2'-Deoxyguanosine 5'-triphosphate:oxydized-thioredoxin 2'-oxidoreductase\n",
      "rxn01334_c Sedoheptulose 1,7-bisphosphate D-glyceraldehyde-3-phosphate-lyase\n",
      "rxn08061_c Acetate transport via proton symport\n",
      "rxn00193_c glutamate racemase\n",
      "EX_cpd00311_e Guanosine exchange\n",
      "rxn00871_c Phosphate butyrltransferase\n",
      "\n",
      "Coculture only 6\n",
      "rxn00747_c D-glyceraldehyde-3-phosphate aldose-ketose-isomerase\n",
      "EX_cpd00277_e Deoxyguanosine exchange\n",
      "rxn09688_c deoxyguanosine transport in via proton symport (periplasm)\n",
      "rxn13140_c Nucleoside triphosphatase\n",
      "rxn00182_c L-glutamate:NAD+ oxidoreductase (deaminating)\n",
      "rxn01977_c alpha-D-Glucose 6-phosphate ketol-isomerase\n"
     ]
    }
   ],
   "source": [
    "# Differential reaction essentiality\n",
    "solo_essential_bhi = find_essential_reactions(iCdR700_solo.model)\n",
    "solo_essential_bhi = set([x.id for x in solo_essential_bhi]).difference(base_essential_bhi)\n",
    "coculture_essential_bhi = find_essential_reactions(iCdR700_coculture.model)\n",
    "coculture_essential_bhi = set([x.id for x in coculture_essential_bhi]).difference(base_essential_bhi)\n",
    "solo_essential_bhi_only = solo_essential_bhi.difference(coculture_essential_bhi)\n",
    "coculture_essential_bhi_only = coculture_essential_bhi.difference(solo_essential_bhi)\n",
    "essential_bhi = solo_essential_bhi.intersection(coculture_essential_bhi)\n",
    "\n",
    "print('Solo only', len(solo_essential_bhi_only))\n",
    "for x in solo_essential_bhi_only: print(x, iCdR700.reactions.get_by_id(x).name)   \n",
    "print('\\nCoculture only', len(coculture_essential_bhi_only))\n",
    "for x in coculture_essential_bhi_only: print(x, iCdR700.reactions.get_by_id(x).name)   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo in BHI only 5\n",
      "EX_cpd00023_e L-Glutamate exchange\n",
      "EX_cpd00013_e Ammonia exchange\n",
      "EX_cpd00001_e H2O exchange\n",
      "EX_cpd00011_e CO2 exchange\n",
      "EX_cpd00311_e Guanosine exchange\n",
      "\n",
      "Coculture in BHI only 5\n",
      "EX_cpd00186_e D-Glutamate exchange\n",
      "EX_cpd00067_e H+ exchange\n",
      "EX_cpd00277_e Deoxyguanosine exchange\n",
      "EX_cpd00165_e Hydroxylamine exchange\n",
      "EX_cpd00207_e Guanine exchange\n"
     ]
    }
   ],
   "source": [
    "# Exchanges\n",
    "solo_exch = set([x.id for x in iCdR700_solo.model.exchanges])\n",
    "coculture_exch = set([x.id for x in iCdR700_coculture.model.exchanges])\n",
    "\n",
    "solo_exch_only = solo_exch.difference(coculture_exch)\n",
    "coculture_exch_only = coculture_exch.difference(solo_exch)\n",
    "\n",
    "print('Solo in BHI only', len(solo_exch_only))\n",
    "for x in solo_exch_only: print(x, iCdR700.reactions.get_by_id(x).name)   \n",
    "print('\\nCoculture in BHI only', len(coculture_exch_only))\n",
    "for x in coculture_exch_only: print(x, iCdR700.reactions.get_by_id(x).name)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H2O EX_cpd00001_e -284.401\n",
      "L-Glutamate EX_cpd00023_e -56.63\n",
      "Na+ EX_cpd00971_e -0.071\n",
      "L-Arginine EX_cpd00051_e -4.56\n",
      "Ornithine EX_cpd00064_e -3.905\n",
      "Fe2+ EX_cpd10515_e -0.071\n",
      "L-Methionine EX_cpd00060_e -3.848\n",
      "L-Alanine EX_cpd00035_e -853.638\n",
      "Riboflavin EX_cpd00220_e -0.078\n",
      "D-Alanine EX_cpd00117_e -74.813\n",
      "Mg EX_cpd00254_e -0.071\n",
      "K+ EX_cpd00205_e -0.071\n",
      "Thiamin EX_cpd00305_e -0.071\n",
      "L-Tryptophan EX_cpd00065_e -0.912\n",
      "Co2+ EX_cpd00149_e -0.071\n",
      "Cl- EX_cpd00099_e -0.071\n",
      "Biotin EX_cpd00104_e -0.071\n",
      "Ca2+ EX_cpd00063_e -0.071\n",
      "L-Valine EX_cpd00156_e -9.348\n",
      "L-Lysine EX_cpd00039_e -34.4\n",
      "Folate EX_cpd00393_e -0.071\n",
      "L-Isoleucine EX_cpd00322_e -165.436\n",
      "Mn2+ EX_cpd00030_e -0.071\n",
      "L-Cysteine EX_cpd00084_e -1.845\n",
      "Pantothenate EX_cpd00644_e -0.15\n",
      "Pyridoxol EX_cpd00263_e -0.071\n",
      "L-Leucine EX_cpd00107_e -19.139\n",
      "L-Tyrosine EX_cpd00069_e -5.928\n",
      "L-Phenylalanine EX_cpd00066_e -6.242\n",
      "L-Histidine EX_cpd00119_e -1.938\n",
      "L-Asparagine EX_cpd00132_e -9.177\n",
      "L-Serine EX_cpd00054_e -79.609\n",
      "L-Threonine EX_cpd00161_e -7.068\n",
      "L-Aspartate EX_cpd00041_e -51.551\n",
      "Nicotinamide EX_cpd00133_e -0.071\n",
      "D-glucose-6-phosphate EX_cpd00079_e -101.978\n",
      "Inosine EX_cpd00246_e -743.707\n",
      "Guanosine EX_cpd00311_e -3.754\n",
      "Deoxycytidine EX_cpd00654_e -2.244\n",
      "Deoxyuridine EX_cpd00412_e -915.062\n"
     ]
    }
   ],
   "source": [
    "# Input\n",
    "for (columnName, columnData) in iCdR700_solo.flux_samples.iteritems():\n",
    "    if 'EX_' in columnName:\n",
    "        colMedian = round(numpy.median(columnData.values),3)\n",
    "        if colMedian < -1e-6:\n",
    "            substrate = list(iCdR700.reactions.get_by_id(columnName).reactants)[0].name\n",
    "            print(substrate, columnName, colMedian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phosphate EX_cpd00009_e 58.132\n",
      "N-Acetyl-D-glucosamine EX_cpd00122_e 618.77\n",
      "Hypoxanthine EX_cpd00226_e 683.012\n",
      "Uracil EX_cpd00092_e 915.062\n",
      "CO2 EX_cpd00011_e 404.937\n",
      "Isovalerate EX_cpd05178_e 4.897\n",
      "2-Methylbutyrate EX_cpd19585_e 150.073\n",
      "Acetate EX_cpd00029_e 972.575\n",
      "Ammonia EX_cpd00013_e 387.216\n",
      "Formate EX_cpd00047_e 1000.0\n"
     ]
    }
   ],
   "source": [
    "# Output\n",
    "for (columnName, columnData) in iCdR700_solo.flux_samples.iteritems():\n",
    "    if 'EX_' in columnName:\n",
    "        colMedian = round(numpy.median(columnData.values),3)\n",
    "        if colMedian > 1e-6:\n",
    "            substrate = list(iCdR700.reactions.get_by_id(columnName).reactants)[0].name\n",
    "            print(substrate, columnName, colMedian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Na+ EX_cpd00971_e -0.073\n",
      "L-Arginine EX_cpd00051_e -4.7\n",
      "Ornithine EX_cpd00064_e -103.569\n",
      "Fe2+ EX_cpd10515_e -0.073\n",
      "L-Methionine EX_cpd00060_e -3.966\n",
      "L-Alanine EX_cpd00035_e -746.186\n",
      "Riboflavin EX_cpd00220_e -0.081\n",
      "D-Alanine EX_cpd00117_e -77.109\n",
      "Mg EX_cpd00254_e -0.073\n",
      "K+ EX_cpd00205_e -0.073\n",
      "Thiamin EX_cpd00305_e -0.073\n",
      "L-Tryptophan EX_cpd00065_e -0.94\n",
      "Co2+ EX_cpd00149_e -0.073\n",
      "Cl- EX_cpd00099_e -0.073\n",
      "Biotin EX_cpd00104_e -0.073\n",
      "Ca2+ EX_cpd00063_e -0.073\n",
      "L-Valine EX_cpd00156_e -9.635\n",
      "L-Lysine EX_cpd00039_e -35.456\n",
      "Folate EX_cpd00393_e -0.073\n",
      "L-Isoleucine EX_cpd00322_e -191.142\n",
      "Mn2+ EX_cpd00030_e -0.073\n",
      "L-Cysteine EX_cpd00084_e -1.902\n",
      "Pantothenate EX_cpd00644_e -0.154\n",
      "Pyridoxol EX_cpd00263_e -0.073\n",
      "L-Leucine EX_cpd00107_e -223.464\n",
      "L-Tyrosine EX_cpd00069_e -6.11\n",
      "L-Phenylalanine EX_cpd00066_e -6.433\n",
      "L-Histidine EX_cpd00119_e -1.998\n",
      "L-Asparagine EX_cpd00132_e -9.459\n",
      "L-Serine EX_cpd00054_e -382.682\n",
      "L-Threonine EX_cpd00161_e -7.285\n",
      "L-Aspartate EX_cpd00041_e -76.101\n",
      "Nicotinamide EX_cpd00133_e -0.073\n",
      "D-glucose-6-phosphate EX_cpd00079_e -210.03\n",
      "Inosine EX_cpd00246_e -774.682\n",
      "Deoxyguanosine EX_cpd00277_e -1.131\n",
      "Deoxycytidine EX_cpd00654_e -2.313\n",
      "Deoxyuridine EX_cpd00412_e -595.37\n"
     ]
    }
   ],
   "source": [
    "# Input\n",
    "for (columnName, columnData) in iCdR700_coculture.flux_samples.iteritems():\n",
    "    if 'EX_' in columnName:\n",
    "        colMedian = round(numpy.median(columnData.values),3)\n",
    "        if colMedian < -1e-6:\n",
    "            substrate = list(iCdR700.reactions.get_by_id(columnName).reactants)[0].name\n",
    "            print(substrate, columnName, colMedian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phosphate EX_cpd00009_e 165.666\n",
      "H+ EX_cpd00067_e 85.343\n",
      "N-Acetyl-D-glucosamine EX_cpd00122_e 812.11\n",
      "Hypoxanthine EX_cpd00226_e 432.457\n",
      "Uracil EX_cpd00092_e 595.37\n",
      "Isovalerate EX_cpd05178_e 210.099\n",
      "2-Methylbutyrate EX_cpd19585_e 176.198\n",
      "Acetate EX_cpd00029_e 862.527\n",
      "Formate EX_cpd00047_e 1000.0\n",
      "Hydroxylamine EX_cpd00165_e 490.404\n",
      "D-Glutamate EX_cpd00186_e 36.994\n",
      "Guanine EX_cpd00207_e 336.706\n"
     ]
    }
   ],
   "source": [
    "# Output\n",
    "for (columnName, columnData) in iCdR700_coculture.flux_samples.iteritems():\n",
    "    if 'EX_' in columnName:\n",
    "        colMedian = round(numpy.median(columnData.values),3)\n",
    "        if colMedian > 1e-6:\n",
    "            substrate = list(iCdR700.reactions.get_by_id(columnName).reactants)[0].name\n",
    "            print(substrate, columnName, colMedian)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solo in BHI only 5\n",
      "EX_cpd00311_e Guanosine exchange\n",
      "EX_cpd00035_e L-Alanine exchange\n",
      "EX_cpd00412_e Deoxyuridine exchange\n",
      "EX_cpd00079_e D-glucose-6-phosphate exchange\n",
      "EX_cpd00277_e Deoxyguanosine exchange\n",
      "\n",
      "Coculture in BHI only 4\n",
      "EX_cpd00064_e Ornithine exchange\n",
      "EX_cpd00246_e Inosine exchange\n",
      "EX_cpd00186_e D-Glutamate exchange\n",
      "EX_cpd00009_e Phosphate exchange\n"
     ]
    }
   ],
   "source": [
    "# Define minimal medium for each model\n",
    "\n",
    "# constrained to media condition\n",
    "minmedia_solo = minimal_medium(iCdR700_solo.model, minimize_components=True, min_objective_value=0.8)\n",
    "minmedia_coculture = minimal_medium(iCdR700_coculture.model, minimize_components=True, min_objective_value=0.8)\n",
    "\n",
    "coculture_min_only = set(list(minmedia_coculture.index)).difference(set(list(minmedia_solo.index)))\n",
    "solo_min_only = set(list(minmedia_solo.index)).difference(set(list(minmedia_coculture.index)))\n",
    "\n",
    "print('Solo in BHI only', len(solo_min_only))\n",
    "for x in solo_min_only: print(x, iCdR700.reactions.get_by_id(x).name)   \n",
    "print('\\nCoculture in BHI only', len(coculture_min_only))\n",
    "for x in coculture_min_only: print(x, iCdR700.reactions.get_by_id(x).name)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
